<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper plugin-docs plugin-id-default docs-version-current docs-doc-page docs-doc-id-modules/filter_genotypes" data-has-hydrated="false">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v3.8.1">
<title data-rh="true">FilterGenotypes | GATK-SV</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/gatk-sv/docs/modules/fg"><meta data-rh="true" property="og:locale" content="en"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="FilterGenotypes | GATK-SV"><meta data-rh="true" name="description" content="Recalibrates qualities and filters genotypes"><meta data-rh="true" property="og:description" content="Recalibrates qualities and filters genotypes"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/gatk-sv/docs/modules/fg"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/gatk-sv/docs/modules/fg" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/gatk-sv/docs/modules/fg" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://FPHFCFLYVM-dsn.algolia.net" crossorigin="anonymous"><script data-rh="true" type="application/ld+json">{"@context":"https://schema.org","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"name":"Modules","item":"https://broadinstitute.github.io/gatk-sv/docs/category/modules"},{"@type":"ListItem","position":2,"name":"FilterGenotypes","item":"https://broadinstitute.github.io/gatk-sv/docs/modules/fg"}]}</script><link rel="preconnect" href="https://www.google-analytics.com">
<link rel="preconnect" href="https://www.googletagmanager.com">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-9DQEYKHD1M"></script>
<script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-9DQEYKHD1M",{anonymize_ip:!0})</script>



<link rel="search" type="application/opensearchdescription+xml" title="GATK-SV" href="/gatk-sv/opensearch.xml"><link rel="stylesheet" href="/gatk-sv/assets/css/styles.82e231fb.css">
<script src="/gatk-sv/assets/js/runtime~main.1bb5af84.js" defer="defer"></script>
<script src="/gatk-sv/assets/js/main.1ed6e5ce.js" defer="defer"></script>
</head>
<body class="navigation-with-keyboard">
<svg xmlns="http://www.w3.org/2000/svg" style="display: none;"><defs>
<symbol id="theme-svg-external-link" viewBox="0 0 24 24"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"/></symbol>
</defs></svg>
<script>!function(){var t="light";var e=function(){try{return new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}}()||function(){try{return window.localStorage.getItem("theme")}catch(t){}}();document.documentElement.setAttribute("data-theme",e||t),document.documentElement.setAttribute("data-theme-choice",e||t)}(),function(){try{const c=new URLSearchParams(window.location.search).entries();for(var[t,e]of c)if(t.startsWith("docusaurus-data-")){var a=t.replace("docusaurus-data-","data-");document.documentElement.setAttribute(a,e)}}catch(t){}}()</script><div id="__docusaurus"><div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="theme-layout-navbar navbar navbar--fixed-top"><div class="navbar__inner"><div class="theme-layout-navbar-left navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/gatk-sv/"><b class="navbar__title text--truncate">GATK-SV</b></a></div><div class="theme-layout-navbar-right navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/gatk-sv/docs/intro">Documentation</a><a href="https://github.com/broadinstitute/gatk-sv/issues" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">Questions<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a><a href="https://github.com/broadinstitute/gatk-sv" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="system mode" aria-label="Switch between dark and light mode (currently system mode)"><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP systemToggleIcon_QzmC"><path fill="currentColor" d="m12 21c4.971 0 9-4.029 9-9s-4.029-9-9-9-9 4.029-9 9 4.029 9 9 9zm4.95-13.95c1.313 1.313 2.05 3.093 2.05 4.95s-0.738 3.637-2.05 4.95c-1.313 1.313-3.093 2.05-4.95 2.05v-14c1.857 0 3.637 0.737 4.95 2.05z"></path></svg></button></div><div class="navbarSearchContainer_Bca1"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search (Command+K)"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20" aria-hidden="true"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="theme-layout-main main-wrapper mainWrapper_z2l0"><div class="docsWrapper_hBAB"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docRoot_UBD9"><aside class="theme-doc-sidebar-container docSidebarContainer_YfHR"><div class="sidebarViewport_aRkj"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/intro">About</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/getting-started">Getting Started</a><button aria-label="Expand sidebar category &#x27;Getting Started&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/best_practices">Best Practices Guide</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/execution">Execution</a><button aria-label="Expand sidebar category &#x27;Execution&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" href="/gatk-sv/docs/category/modules">Modules</a><button aria-label="Collapse sidebar category &#x27;Modules&#x27;" aria-expanded="true" type="button" class="clean-btn menu__caret"></button></div><ul class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/">Overview</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/gse">GatherSampleEvidence</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/eqc">EvidenceQC</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/gcnv">TrainGCNV</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/gbe">GatherBatchEvidence</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/cb">ClusterBatch</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/gbm">GenerateBatchMetrics</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/fb">FilterBatch</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/msites">MergeBatchSites</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/gb">GenotypeBatch</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/rgcnvs">RegenotypeCNVs</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/cmb">CombineBatches</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/rcv">ResolveComplexVariants</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/gcv">GenotypeComplexVariants</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/cvcf">CleanVcf</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/refcv">RefineComplexVariants</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/jrc">JoinRawCalls</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/svc">SVConcordance</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/gatk-sv/docs/modules/fg">FilterGenotypes</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/av">AnnotateVcf</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/mvqc">MainVcfQC</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/viz">VisualizeCnvs</a></li></ul></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/runtime_attr">Runtime attributes</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/resources">Resource files</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/troubleshooting">Troubleshooting</a><button aria-label="Expand sidebar category &#x27;Troubleshooting&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/advanced-guides">Advanced Guides</a><button aria-label="Expand sidebar category &#x27;Advanced Guides&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/acknowledgements">Acknowledgements</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/references">Citation and references</a></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_TBSr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/gatk-sv/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><a class="breadcrumbs__link" href="/gatk-sv/docs/category/modules"><span>Modules</span></a></li><li class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link">FilterGenotypes</span></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><header><h1>FilterGenotypes</h1></header><p><a href="https://github.com/broadinstitute/gatk-sv/blob/main/wdl/FilterGenotypes.wdl" target="_blank" rel="noopener noreferrer">WDL source code</a></p>
<p>Performs genotype quality recalibration using a machine learning model based on <a href="https://github.com/dmlc/xgboost" target="_blank" rel="noopener noreferrer">xgboost</a>
and filters genotypes. The output VCF contains the following updated fields:</p>
<ul>
<li><code>SL</code> : Scaled logit scores (see <a href="#sl-scores">here</a>)</li>
<li><code>GQ</code> : Updated genotype quality rescaled using <code>SL</code></li>
<li><code>OGQ</code> : Original <code>GQ</code> score before recalibration</li>
<li><code>HIGH_NCR</code> : Filter status assigned to variants exceeding a <a href="#optional-no_call_rate_cutoff">threshold proportion</a>
of no-call genotypes. This will also be applied to variants with genotypes that have already been filtered in the input VCF.</li>
</ul>
<p>The following diagram illustrates the recommended invocation order:</p>
<!-- -->
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="model-features">Model features<a href="#model-features" class="hash-link" aria-label="Direct link to Model features" title="Direct link to Model features">​</a></h3>
<p>The model uses the following features:</p>
<ul>
<li>Genotype properties:<!-- -->
<ul>
<li>Non-reference and no-call allele counts</li>
<li>Genotype quality (<code>GQ</code>)</li>
<li>Supporting evidence types (<code>EV</code>) and respective genotype qualities (<code>PE_GQ</code>, <code>SR_GQ</code>, <code>RD_GQ</code>)</li>
<li>Raw call concordance (<code>CONC_ST</code>)</li>
</ul>
</li>
<li>Variant properties:<!-- -->
<ul>
<li>Variant type (<code>SVTYPE</code>) and size (<code>SVLEN</code>)</li>
<li>Calling algorithms (<code>ALGORITHMS</code>)</li>
<li>Supporting evidence types (<code>EVIDENCE</code>)</li>
<li>Two-sided SR support flag (<code>BOTHSIDES_SUPPORT</code>)</li>
<li>Evidence overdispersion flag (<code>PESR_GT_OVERDISPERSION</code>)</li>
<li>SR noise flag (<code>HIGH_SR_BACKGROUND</code>)</li>
<li>Raw call concordance (<code>STATUS</code>, <code>NON_REF_GENOTYPE_CONCORDANCE</code>, <code>VAR_PPV</code>, <code>VAR_SENSITIVITY</code>, <code>TRUTH_AF</code>)</li>
</ul>
</li>
<li>Reference context with respect to UCSC Genome Browser tracks:<!-- -->
<ul>
<li>RepeatMasker</li>
<li>Segmental duplications</li>
<li>Simple repeats</li>
<li>K-mer mappability (umap_s100 and umap_s24)</li>
</ul>
</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="model-availability">Model availability<a href="#model-availability" class="hash-link" aria-label="Direct link to Model availability" title="Direct link to Model availability">​</a></h3>
<p>For ease of use, we provide a model pre-trained on high-quality data with truth data derived from long-read calls:</p>
<div class="language-text codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_QJqH"><pre tabindex="0" class="prism-code language-text codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">gs://gatk-sv-resources-public/hg38/v0/sv-resources/resources/v1/gatk-sv-recalibrator.aou_phase_1.v1.model</span><br></span></code></pre></div></div>
<p>See the SV &quot;Genotype Filter&quot; section on page 34 of the <a href="https://support.researchallofus.org/hc/en-us/articles/4617899955092-All-of-Us-Genomic-Quality-Report-ARCHIVED-C2022Q4R9-CDR-v7" target="_blank" rel="noopener noreferrer">All of Us Genomic Quality Report C2022Q4R9 CDR v7</a> for further details on model training. The generation and release of this model was made possible by the All of Us program (see <a href="/gatk-sv/docs/acknowledgements">here</a>).</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="sl-scores">SL scores<a href="#sl-scores" class="hash-link" aria-label="Direct link to SL scores" title="Direct link to SL scores">​</a></h3>
<p>All valid genotypes are annotated with a &quot;scaled logit&quot; (<code>SL</code>) score, which is rescaled to non-negative adjusted <code>GQ</code> values on [1, 99]. Note that the rescaled <code>GQ</code> values should <em>not</em> be interpreted as probabilities. Original genotype qualities are retained in the <code>OGQ</code> field.</p>
<p>A more positive <code>SL</code> score indicates higher probability that the given genotype is not homozygous for the reference allele. Genotypes are therefore filtered using <code>SL</code> thresholds that depend on SV type and size. This workflow also generates QC plots using the <a href="https://github.com/broadinstitute/gatk-sv/blob/main/wdl/MainVcfQc.wdl" target="_blank" rel="noopener noreferrer">MainVcfQc</a> workflow to review call set quality (see below for recommended practices).</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="modes">Modes<a href="#modes" class="hash-link" aria-label="Direct link to Modes" title="Direct link to Modes">​</a></h3>
<p>This workflow can be run in one of two modes:</p>
<ol>
<li>
<p>(Recommended) The user explicitly provides a set of <code>SL</code> cutoffs through the <code>sl_filter_args</code> parameter, e.g.</p>
<div class="language-text codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_QJqH"><pre tabindex="0" class="prism-code language-text codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">&quot;--small-del-threshold 93 --medium-del-threshold 150 --small-dup-threshold -51 --medium-dup-threshold -4 --ins-threshold -13 --inv-threshold -19&quot;</span><br></span></code></pre></div></div>
<p>Genotypes with <code>SL</code> scores less than the cutoffs are set to no-call (<code>./.</code>). The above values were taken directly from Appendix N of the <a href="https://support.researchallofus.org/hc/en-us/articles/4617899955092-All-of-Us-Genomic-Quality-Report-ARCHIVED-C2022Q4R9-CDR-v7" target="_blank" rel="noopener noreferrer">All of Us Genomic Quality Report C2022Q4R9 CDR v7 </a>. Users should adjust the thresholds depending on data quality and desired accuracy. Please see the arguments in <a href="https://github.com/broadinstitute/gatk-sv/blob/main/src/sv-pipeline/scripts/apply_sl_filter.py" target="_blank" rel="noopener noreferrer">this script</a> for all available options.</p>
</li>
<li>
<p>(Advanced) The user provides truth labels for a subset of non-reference calls, and <code>SL</code> cutoffs are automatically optimized. These truth labels should be provided as a json file in the following format:</p>
<div class="language-json codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_QJqH"><pre tabindex="0" class="prism-code language-json codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"> </span><span class="token punctuation" style="color:#393A34">{</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   </span><span class="token property" style="color:#36acaa">&quot;sample_1&quot;</span><span class="token operator" style="color:#393A34">:</span><span class="token plain"> </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   </span><span class="token punctuation" style="color:#393A34">{</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">     </span><span class="token property" style="color:#36acaa">&quot;good_variant_ids&quot;</span><span class="token operator" style="color:#393A34">:</span><span class="token plain"> </span><span class="token punctuation" style="color:#393A34">[</span><span class="token string" style="color:#e3116c">&quot;variant_1&quot;</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;variant_3&quot;</span><span class="token punctuation" style="color:#393A34">]</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"> </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">     </span><span class="token property" style="color:#36acaa">&quot;bad_variant_ids&quot;</span><span class="token operator" style="color:#393A34">:</span><span class="token plain"> </span><span class="token punctuation" style="color:#393A34">[</span><span class="token string" style="color:#e3116c">&quot;variant_5&quot;</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;variant_10&quot;</span><span class="token punctuation" style="color:#393A34">]</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   </span><span class="token punctuation" style="color:#393A34">}</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   </span><span class="token property" style="color:#36acaa">&quot;sample_2&quot;</span><span class="token operator" style="color:#393A34">:</span><span class="token plain"> </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   </span><span class="token punctuation" style="color:#393A34">{</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">     </span><span class="token property" style="color:#36acaa">&quot;good_variant_ids&quot;</span><span class="token operator" style="color:#393A34">:</span><span class="token plain"> </span><span class="token punctuation" style="color:#393A34">[</span><span class="token string" style="color:#e3116c">&quot;variant_2&quot;</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;variant_13&quot;</span><span class="token punctuation" style="color:#393A34">]</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"> </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">     </span><span class="token property" style="color:#36acaa">&quot;bad_variant_ids&quot;</span><span class="token operator" style="color:#393A34">:</span><span class="token plain"> </span><span class="token punctuation" style="color:#393A34">[</span><span class="token string" style="color:#e3116c">&quot;variant_8&quot;</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;variant_11&quot;</span><span class="token punctuation" style="color:#393A34">]</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   </span><span class="token punctuation" style="color:#393A34">}</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"> </span><span class="token punctuation" style="color:#393A34">}</span><br></span></code></pre></div></div>
<p>where &quot;good_variant_ids&quot; and &quot;bad_variant_ids&quot; are lists of variant IDs corresponding to non-reference (i.e. het or hom-var) sample genotypes that are true positives and false positives, respectively. <code>SL</code> cutoffs are optimized by maximizing the <a href="https://en.wikipedia.org/wiki/F-score" target="_blank" rel="noopener noreferrer">F-score</a> with &quot;beta&quot; parameter <code>fmax_beta</code>, which modulates the weight given to precision over recall (lower values give higher precision).</p>
</li>
</ol>
<p>In both modes, the workflow additionally filters variants based on the &quot;no-call rate&quot;, the proportion of genotypes that were filtered in a given variant. Variants exceeding the <code>no_call_rate_cutoff</code> are assigned a <code>HIGH_NCR</code> filter status.</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="qc-recommendations">QC recommendations<a href="#qc-recommendations" class="hash-link" aria-label="Direct link to QC recommendations" title="Direct link to QC recommendations">​</a></h3>
<p>We strongly recommend performing call set QC after this module. By default, QC plotting is enabled with the <a href="#optional-run_qc">run_qc</a>
argument. Users should carefully inspect the main plots from the <a href="#optional-main_vcf_qc_tarball">main_vcf_qc_tarball</a>.
Please see the <a href="/gatk-sv/docs/modules/mvqc">MainVcfQc</a> module documentation for more information on interpreting these plots and recommended
QC criteria.</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="inputs">Inputs<a href="#inputs" class="hash-link" aria-label="Direct link to Inputs" title="Direct link to Inputs">​</a></h3>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-vcf"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>vcf</code><a href="#optional-vcf" class="hash-link" aria-label="Direct link to optional-vcf" title="Direct link to optional-vcf">​</a></h4>
<p>Input VCF generated from <a href="/gatk-sv/docs/modules/svc#concordance_vcf">SVConcordance</a>.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-output_prefix"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>output_prefix</code><a href="#optional-output_prefix" class="hash-link" aria-label="Direct link to optional-output_prefix" title="Direct link to optional-output_prefix">​</a></h4>
<p>Default: use input VCF filename. Prefix for the output VCF, such as the cohort name. May be alphanumeric with underscores.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="ploidy_table"><code>ploidy_table</code><a href="#ploidy_table" class="hash-link" aria-label="Direct link to ploidy_table" title="Direct link to ploidy_table">​</a></h4>
<p>Table of sample ploidies generated in <a href="/gatk-sv/docs/modules/jrc#ploidy_table">JoinRawCalls</a>.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="gq_recalibrator_model_file"><code>gq_recalibrator_model_file</code><a href="#gq_recalibrator_model_file" class="hash-link" aria-label="Direct link to gq_recalibrator_model_file" title="Direct link to gq_recalibrator_model_file">​</a></h4>
<p>GQ-Recalibrator model. A public model is listed as <code>aou_recalibrate_gq_model_file</code> <a href="https://github.com/broadinstitute/gatk-sv/blob/main/inputs/values/resources_hg38.json" target="_blank" rel="noopener noreferrer">here</a>.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="recalibrate_gq_args"><code>recalibrate_gq_args</code><a href="#recalibrate_gq_args" class="hash-link" aria-label="Direct link to recalibrate_gq_args" title="Direct link to recalibrate_gq_args">​</a></h4>
<p>Arguments to pass to the <code>GQ</code> recalibration tool. Users should leave this with the default configuration in Terra.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="genome_tracks"><code>genome_tracks</code><a href="#genome_tracks" class="hash-link" aria-label="Direct link to genome_tracks" title="Direct link to genome_tracks">​</a></h4>
<p>Genome tracks for sequence context annotation. Users should leave this with the default configuration in Terra.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-no_call_rate_cutoff"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>no_call_rate_cutoff</code><a href="#optional-no_call_rate_cutoff" class="hash-link" aria-label="Direct link to optional-no_call_rate_cutoff" title="Direct link to optional-no_call_rate_cutoff">​</a></h4>
<p>Default: <code>0.05</code>. Threshold fraction of samples that must have no-call genotypes in order to filter a variant. Set to 1 to disable.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-fmax_beta"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>fmax_beta</code><a href="#optional-fmax_beta" class="hash-link" aria-label="Direct link to optional-fmax_beta" title="Direct link to optional-fmax_beta">​</a></h4>
<p>Default: <code>0.4</code>. If providing a truth set, defines the beta parameter for F-score optimization.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-truth_json"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>truth_json</code><a href="#optional-truth_json" class="hash-link" aria-label="Direct link to optional-truth_json" title="Direct link to optional-truth_json">​</a></h4>
<p>Truth labels for input variants. If provided, the workflow will attempt to optimize filtering cutoffs automatically
using the F-score. If provided, <a href="#optional-sl_filter_args">sl_filter_args</a> is ignored.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-sl_filter_args"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>sl_filter_args</code><a href="#optional-sl_filter_args" class="hash-link" aria-label="Direct link to optional-sl_filter_args" title="Direct link to optional-sl_filter_args">​</a></h4>
<p>Arguments for the <a href="https://github.com/broadinstitute/gatk-sv/blob/main/src/sv-pipeline/scripts/apply_sl_filter.py" target="_blank" rel="noopener noreferrer">SL filtering script</a>.
This should be used to set <code>SL</code> cutoffs for filtering (refer to description above). Overridden by <a href="#optional-truth_json">truth_json</a>.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-run_qc"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>run_qc</code><a href="#optional-run_qc" class="hash-link" aria-label="Direct link to optional-run_qc" title="Direct link to optional-run_qc">​</a></h4>
<p>Default: <code>true</code>. Enable running <a href="/gatk-sv/docs/modules/mvqc">MainVcfQc</a> automatically. By default, filtered variants will be excluded from
the plots.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-optimize_vcf_records_per_shard"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>optimize_vcf_records_per_shard</code><a href="#optional-optimize_vcf_records_per_shard" class="hash-link" aria-label="Direct link to optional-optimize_vcf_records_per_shard" title="Direct link to optional-optimize_vcf_records_per_shard">​</a></h4>
<p>Default: <code>50000</code>. Shard size for scattered cutoff optimization tasks. Decrease this if those steps are running slowly.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-filter_vcf_records_per_shard"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>filter_vcf_records_per_shard</code><a href="#optional-filter_vcf_records_per_shard" class="hash-link" aria-label="Direct link to optional-filter_vcf_records_per_shard" title="Direct link to optional-filter_vcf_records_per_shard">​</a></h4>
<p>Default: <code>20000</code>. Shard size for scattered <code>GQ</code> recalibration tasks. Decrease this if those steps are running slowly.</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="outputs">Outputs<a href="#outputs" class="hash-link" aria-label="Direct link to Outputs" title="Direct link to Outputs">​</a></h3>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="filtered_vcf"><code>filtered_vcf</code><a href="#filtered_vcf" class="hash-link" aria-label="Direct link to filtered_vcf" title="Direct link to filtered_vcf">​</a></h4>
<p>Filtered VCF.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-main_vcf_qc_tarball"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>main_vcf_qc_tarball</code><a href="#optional-main_vcf_qc_tarball" class="hash-link" aria-label="Direct link to optional-main_vcf_qc_tarball" title="Direct link to optional-main_vcf_qc_tarball">​</a></h4>
<p>QC plots generated with <a href="/gatk-sv/docs/modules/mvqc">MainVcfQc</a>. Only generated if using <a href="#optional-run_qc">run_qc</a>.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-vcf_optimization_table"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>vcf_optimization_table</code><a href="#optional-vcf_optimization_table" class="hash-link" aria-label="Direct link to optional-vcf_optimization_table" title="Direct link to optional-vcf_optimization_table">​</a></h4>
<p>Table of cutoff optimization metrics. Only generated if <a href="#optional-truth_json">truth_json</a> is provided.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-sl_cutoff_qc_tarball"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>sl_cutoff_qc_tarball</code><a href="#optional-sl_cutoff_qc_tarball" class="hash-link" aria-label="Direct link to optional-sl_cutoff_qc_tarball" title="Direct link to optional-sl_cutoff_qc_tarball">​</a></h4>
<p>Cutoff optimization and QC plots. Only generated if <a href="#optional-truth_json">truth_json</a> is provided.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="unfiltered_recalibrated_vcf"><code>unfiltered_recalibrated_vcf</code><a href="#unfiltered_recalibrated_vcf" class="hash-link" aria-label="Direct link to unfiltered_recalibrated_vcf" title="Direct link to unfiltered_recalibrated_vcf">​</a></h4>
<p>Supplemental output of the VCF after assigning <code>SL</code> genotype scores but before applying filtering.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="row margin-top--sm theme-doc-footer-edit-meta-row"><div class="col"><a href="https://github.com/broadinstitute/gatk-sv/tree/master/website/docs/modules/filter_genotypes.md" target="_blank" rel="noopener noreferrer" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_JAkA"></div></div></footer></article><nav class="docusaurus-mt-lg pagination-nav" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/gatk-sv/docs/modules/svc"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">SVConcordance</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/gatk-sv/docs/modules/av"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">AnnotateVcf</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#model-features" class="table-of-contents__link toc-highlight">Model features</a></li><li><a href="#model-availability" class="table-of-contents__link toc-highlight">Model availability</a></li><li><a href="#sl-scores" class="table-of-contents__link toc-highlight">SL scores</a></li><li><a href="#modes" class="table-of-contents__link toc-highlight">Modes</a></li><li><a href="#qc-recommendations" class="table-of-contents__link toc-highlight">QC recommendations</a></li><li><a href="#inputs" class="table-of-contents__link toc-highlight">Inputs</a></li><li><a href="#outputs" class="table-of-contents__link toc-highlight">Outputs</a></li></ul></div></div></div></div></main></div></div></div><footer class="theme-layout-footer footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="theme-layout-footer-column col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/gatk-sv/docs/intro">About</a></li></ul></div><div class="theme-layout-footer-column col footer__col"><div class="footer__title">Community</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/broadinstitute/gatk-sv/discussions" target="_blank" rel="noopener noreferrer" class="footer__link-item">Github<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a></li></ul></div><div class="theme-layout-footer-column col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://talkowski.mgh.harvard.edu" target="_blank" rel="noopener noreferrer" class="footer__link-item">Talkowski lab</a></li><li class="footer__item"><a href="https://www.broadinstitute.org" target="_blank" rel="noopener noreferrer" class="footer__link-item">Broad Institute</a></li><li class="footer__item"><a href="https://www.massgeneral.org/research/cgm" target="_blank" rel="noopener noreferrer" class="footer__link-item">Center for Genomic Medicine</a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2025 BroadInstitute, Built with Docusaurus.</div></div></div></footer></div>
</body>
</html>