"use strict";(self.webpackChunkGATK_SV=self.webpackChunkGATK_SV||[]).push([[3976],{2053:(e,t,n)=>{n.r(t),n.d(t,{assets:()=>c,contentTitle:()=>a,default:()=>u,frontMatter:()=>r,metadata:()=>i,toc:()=>d});const i=JSON.parse('{"id":"intro","title":"About","description":"GATK-SV","source":"@site/docs/intro.md","sourceDirName":".","slug":"/intro","permalink":"/gatk-sv/docs/intro","draft":false,"unlisted":false,"editUrl":"https://github.com/broadinstitute/gatk-sv/tree/master/website/docs/intro.md","tags":[],"version":"current","sidebarPosition":1,"frontMatter":{"title":"About","description":"GATK-SV","sidebar_position":1},"sidebar":"tutorialSidebar","next":{"title":"Getting Started","permalink":"/gatk-sv/docs/category/getting-started"}}');var s=n(4848),o=n(8453);const r={title:"About",description:"GATK-SV",sidebar_position:1},a=void 0,c={},d=[{value:"Methods",id:"methods",level:3},{value:"GATK Best Practices",id:"gatk-best-practices",level:3},{value:"Where to go from here",id:"where-to-go-from-here",level:3}];function l(e){const t={a:"a",h3:"h3",p:"p",...(0,o.R)(),...e.components};return(0,s.jsxs)(s.Fragment,{children:[(0,s.jsxs)(t.p,{children:["GATK-SV is a comprehensive, cloud-based ensemble pipeline for discovering and annotating all\nclasses of structural variants (SV) from short-read whole genome sequencing (WGS) data. It can detect\ndeletions, duplications, multi-allelic copy number variants, balanced inversions,\ninsertions, translocations, and a diverse spectrum of complex SV. Briefly, GATK-SV\nmaximizes the sensitivity of SV discovery by harmonizing output from five tools:\nManta, Wham, Scramble, cn.MOPS, and GATK-gCNV. To minimize false positives, raw SVs\nare adjudicated and re-genotyped, considering all potential\nsequencing evidence including anomalous paired-end (PE) reads, split reads (SR),\nread-depth (RD), and B-allele frequencies (BAF). It also fully resolves 11 classes of complex\nSVs composed of multiple breakpoints. GATK-SV is intended for use on the ",(0,s.jsx)(t.a,{href:"https://app.terra.bio/",children:"Terra"}),"\nplatform."]}),"\n",(0,s.jsx)(t.h3,{id:"methods",children:"Methods"}),"\n",(0,s.jsxs)(t.p,{children:["Further details about GATK-SV methods can be found in ",(0,s.jsx)(t.a,{href:"https://www.nature.com/articles/s41586-020-2287-8",children:"Collins et al. 2020"}),"."]}),"\n",(0,s.jsx)(t.h3,{id:"gatk-best-practices",children:"GATK Best Practices"}),"\n",(0,s.jsxs)(t.p,{children:["Additional guidance on running GATK-SV is also available ",(0,s.jsx)(t.a,{href:"https://gatk.broadinstitute.org/hc/en-us/articles/9022653744283-GATK-Best-Practices-for-Structural-Variation-Discovery-on-Single-Samples",children:"here"}),"."]}),"\n",(0,s.jsx)(t.h3,{id:"where-to-go-from-here",children:"Where to go from here"}),"\n",(0,s.jsx)(t.p,{children:"This documentation includes instructions for running the pipeline, technical implementation details, troubleshooting\ninformation, and guides for advanced users who wish to work with the source code or rebuild the project."}),"\n",(0,s.jsxs)(t.p,{children:["We recommend new users continue to the ",(0,s.jsx)(t.a,{href:"/gatk-sv/docs/gs/overview",children:"Getting Started overview"}),"."]})]})}function u(e={}){const{wrapper:t}={...(0,o.R)(),...e.components};return t?(0,s.jsx)(t,{...e,children:(0,s.jsx)(l,{...e})}):l(e)}},8453:(e,t,n)=>{n.d(t,{R:()=>r,x:()=>a});var i=n(6540);const s={},o=i.createContext(s);function r(e){const t=i.useContext(o);return i.useMemo((function(){return"function"==typeof e?e(t):{...t,...e}}),[t,e])}function a(e){let t;return t=e.disableParentContext?"function"==typeof e.components?e.components(s):e.components||s:r(e.components),i.createElement(o.Provider,{value:t},e.children)}}}]);