"use strict";(self.webpackChunkGATK_SV=self.webpackChunkGATK_SV||[]).push([[510],{3905:(e,t,n)=>{n.d(t,{Zo:()=>u,kt:()=>f});var a=n(7294);function r(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function i(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var a=Object.getOwnPropertySymbols(e);t&&(a=a.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,a)}return n}function l(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?i(Object(n),!0).forEach((function(t){r(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):i(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function o(e,t){if(null==e)return{};var n,a,r=function(e,t){if(null==e)return{};var n,a,r={},i=Object.keys(e);for(a=0;a<i.length;a++)n=i[a],t.indexOf(n)>=0||(r[n]=e[n]);return r}(e,t);if(Object.getOwnPropertySymbols){var i=Object.getOwnPropertySymbols(e);for(a=0;a<i.length;a++)n=i[a],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(r[n]=e[n])}return r}var s=a.createContext({}),p=function(e){var t=a.useContext(s),n=t;return e&&(n="function"==typeof e?e(t):l(l({},t),e)),n},u=function(e){var t=p(e.components);return a.createElement(s.Provider,{value:t},e.children)},m="mdxType",c={inlineCode:"code",wrapper:function(e){var t=e.children;return a.createElement(a.Fragment,{},t)}},d=a.forwardRef((function(e,t){var n=e.components,r=e.mdxType,i=e.originalType,s=e.parentName,u=o(e,["components","mdxType","originalType","parentName"]),m=p(n),d=r,f=m["".concat(s,".").concat(d)]||m[d]||c[d]||i;return n?a.createElement(f,l(l({ref:t},u),{},{components:n})):a.createElement(f,l({ref:t},u))}));function f(e,t){var n=arguments,r=t&&t.mdxType;if("string"==typeof e||r){var i=n.length,l=new Array(i);l[0]=d;var o={};for(var s in t)hasOwnProperty.call(t,s)&&(o[s]=t[s]);o.originalType=e,o[m]="string"==typeof e?e:r,l[1]=o;for(var p=2;p<i;p++)l[p]=n[p];return a.createElement.apply(null,l)}return a.createElement.apply(null,n)}d.displayName="MDXCreateElement"},3674:(e,t,n)=>{n.r(t),n.d(t,{assets:()=>s,contentTitle:()=>l,default:()=>c,frontMatter:()=>i,metadata:()=>o,toc:()=>p});var a=n(7462),r=(n(7294),n(3905));const i={title:"Input Data",description:"Supported input and reference data.",sidebar_position:3,slug:"./inputs"},l=void 0,o={unversionedId:"gs/input_files",id:"gs/input_files",title:"Input Data",description:"Supported input and reference data.",source:"@site/docs/gs/input_files.md",sourceDirName:"gs",slug:"/gs/inputs",permalink:"/gatk-sv/docs/gs/inputs",draft:!1,editUrl:"https://github.com/broadinstitute/gatk-sv/tree/master/website/docs/gs/input_files.md",tags:[],version:"current",sidebarPosition:3,frontMatter:{title:"Input Data",description:"Supported input and reference data.",sidebar_position:3,slug:"./inputs"},sidebar:"tutorialSidebar",previous:{title:"Quick Start",permalink:"/gatk-sv/docs/gs/qs"},next:{title:"Runtime Environments",permalink:"/gatk-sv/docs/gs/runtime-env"}},s={},p=[{value:"Sample Exclusion",id:"sample-exclusion",level:3},{value:"Sample ID requirements",id:"sample-id-requirements",level:3},{value:"Sample IDs must",id:"sample-ids-must",level:4},{value:"Sample IDs should not",id:"sample-ids-should-not",level:4},{value:"Generating a reference panel (single-sample mode only)",id:"generating-a-reference-panel-single-sample-mode-only",level:3}],u={toc:p},m="wrapper";function c(e){let{components:t,...n}=e;return(0,r.kt)(m,(0,a.Z)({},u,n,{components:t,mdxType:"MDXLayout"}),(0,r.kt)("p",null,"GATK-SV requires the following input data:"),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},(0,r.kt)("p",{parentName:"li"},"Illumina short-read whole-genome CRAMs or BAMs, aligned to hg38 with ",(0,r.kt)("a",{parentName:"p",href:"https://github.com/lh3/bwa"},"bwa-mem"),".\nBAMs must also be indexed.")),(0,r.kt)("li",{parentName:"ul"},(0,r.kt)("p",{parentName:"li"},"Family structure definitions file in\n",(0,r.kt)("a",{parentName:"p",href:"https://gatk.broadinstitute.org/hc/en-us/articles/360035531972-PED-Pedigree-format"},"PED format"),".",(0,r.kt)("br",{parentName:"p"}),"\n",(0,r.kt)("strong",{parentName:"p"},"Note"),": Sex aneuploidies (detected in EvidenceQC) should be entered as sex = 0."))),(0,r.kt)("h3",{id:"sample-exclusion"},"Sample Exclusion"),(0,r.kt)("p",null,"We recommend filtering out samples with a high percentage\nof improperly paired reads (>10% or an outlier for your data)\nas technical outliers prior to running ",(0,r.kt)("a",{parentName:"p",href:"/docs/modules/gse"},"GatherSampleEvidence"),".\nA high percentage of improperly paired reads may indicate issues\nwith library prep, degradation, or contamination. Artifactual\nimproperly paired reads could cause incorrect SV calls, and\nthese samples have been observed to have longer runtimes and\nhigher compute costs for ",(0,r.kt)("a",{parentName:"p",href:"/docs/modules/gse"},"GatherSampleEvidence"),"."),(0,r.kt)("h3",{id:"sample-id-requirements"},"Sample ID requirements"),(0,r.kt)("h4",{id:"sample-ids-must"},"Sample IDs must"),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},"Be unique within the cohort"),(0,r.kt)("li",{parentName:"ul"},"Contain only alphanumeric characters and underscores (no dashes, whitespace, or special characters)")),(0,r.kt)("h4",{id:"sample-ids-should-not"},"Sample IDs should not"),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},"Contain only numeric characters"),(0,r.kt)("li",{parentName:"ul"},"Be a substring of another sample ID in the same cohort"),(0,r.kt)("li",{parentName:"ul"},"Contain any of the following substrings: ",(0,r.kt)("inlineCode",{parentName:"li"},"chr"),", ",(0,r.kt)("inlineCode",{parentName:"li"},"name"),", ",(0,r.kt)("inlineCode",{parentName:"li"},"DEL"),", ",(0,r.kt)("inlineCode",{parentName:"li"},"DUP"),", ",(0,r.kt)("inlineCode",{parentName:"li"},"CPX"),", ",(0,r.kt)("inlineCode",{parentName:"li"},"CHROM"))),(0,r.kt)("p",null,"The same requirements apply to family IDs in the PED file,\nas well as batch IDs and the cohort ID provided as workflow inputs."),(0,r.kt)("p",null,"Sample IDs are provided to GatherSampleEvidence directly and\nneed not match sample names from the BAM/CRAM headers.\n",(0,r.kt)("inlineCode",{parentName:"p"},"GetSampleID.wdl")," can be used to fetch BAM sample IDs and\nalso generates a set of alternate IDs that are considered\nsafe for this pipeline; alternatively, ",(0,r.kt)("a",{parentName:"p",href:"https://github.com/talkowski-lab/gnomad_sv_v3/blob/master/sample_id/convert_sample_ids.py"},"this script"),"\ntransforms a list of sample IDs to fit these requirements.\nCurrently, sample IDs can be replaced again in ",(0,r.kt)("a",{parentName:"p",href:"/docs/modules/gbe"},"GatherBatchEvidence"),"."),(0,r.kt)("p",null,"The following inputs will need to be updated with the transformed sample IDs:"),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},"Sample ID list for ",(0,r.kt)("a",{parentName:"li",href:"/docs/modules/gse"},"GatherSampleEvidence")," or ",(0,r.kt)("a",{parentName:"li",href:"/docs/modules/gbe"},"GatherBatchEvidence")),(0,r.kt)("li",{parentName:"ul"},"PED file")),(0,r.kt)("h3",{id:"generating-a-reference-panel-single-sample-mode-only"},"Generating a reference panel (single-sample mode only)"),(0,r.kt)("p",null,"New reference panels can be generated easily from a single run of the ",(0,r.kt)("inlineCode",{parentName:"p"},"GATKSVPipelineBatch")," workflow. If using a Cromwell server, we recommend copying the outputs to a permanent location by adding the following option to the workflow configuration file:"),(0,r.kt)("pre",null,(0,r.kt)("code",{parentName:"pre"},'  "final_workflow_outputs_dir" : "gs://my-outputs-bucket",\n  "use_relative_output_paths": false,\n')),(0,r.kt)("p",null,"Here is an example of how to generate workflow input jsons from ",(0,r.kt)("inlineCode",{parentName:"p"},"GATKSVPipelineBatch")," workflow metadata:"),(0,r.kt)("pre",null,(0,r.kt)("code",{parentName:"pre"},'> cromshell -t60 metadata 38c65ca4-2a07-4805-86b6-214696075fef > metadata.json\n> python scripts/inputs/create_test_batch.py \\\n    --execution-bucket gs://my-exec-bucket \\\n    --final-workflow-outputs-dir gs://my-outputs-bucket \\\n    metadata.json \\\n    > inputs/values/my_ref_panel.json\n> # Define your google project id (for Cromwell inputs) and Terra billing project (for workspace inputs)\n> echo \'{ "google_project_id": "my-google-project-id", "terra_billing_project_id": "my-terra-billing-project" }\' > inputs/values/google_cloud.my_project.json\n> # Build test files for batched workflows (google cloud project id required)\n> python scripts/inputs/build_inputs.py \\\n    inputs/values \\\n    inputs/templates/test \\\n    inputs/build/my_ref_panel/test \\\n    -a \'{ "test_batch" : "ref_panel_1kg", "cloud_env": "google_cloud.my_project" }\'\n> # Build test files for the single-sample workflow\n> python scripts/inputs/build_inputs.py \\\n    inputs/values \\\n    inputs/templates/test/GATKSVPipelineSingleSample \\\n    inputs/build/NA19240/test_my_ref_panel \\\n    -a \'{ "single_sample" : "test_single_sample_NA19240", "ref_panel" : "my_ref_panel" }\'\n> # Build files for a Terra workspace\n> python scripts/inputs/build_inputs.py \\\n    inputs/values \\\n    inputs/templates/terra_workspaces/single_sample \\\n    inputs/build/NA12878/terra_my_ref_panel \\\n    -a \'{ "single_sample" : "test_single_sample_NA12878", "ref_panel" : "my_ref_panel" }\'\n')),(0,r.kt)("p",null,"Note that the inputs to ",(0,r.kt)("inlineCode",{parentName:"p"},"GATKSVPipelineBatch")," may be used as resources for the reference panel and therefore should also be in a permanent location."))}c.isMDXComponent=!0}}]);