"use strict";(self.webpackChunkGATK_SV=self.webpackChunkGATK_SV||[]).push([[7868],{34:(e,n,i)=>{i.r(n),i.d(n,{assets:()=>l,contentTitle:()=>a,default:()=>h,frontMatter:()=>o,metadata:()=>t,toc:()=>c});const t=JSON.parse('{"id":"references","title":"Citation and references","description":"Citation and references","source":"@site/docs/references.md","sourceDirName":".","slug":"/references","permalink":"/gatk-sv/docs/references","draft":false,"unlisted":false,"editUrl":"https://github.com/broadinstitute/gatk-sv/tree/master/website/docs/references.md","tags":[],"version":"current","sidebarPosition":11,"frontMatter":{"title":"Citation and references","description":"Citation and references","sidebar_position":11},"sidebar":"tutorialSidebar","previous":{"title":"Acknowledgements","permalink":"/gatk-sv/docs/acknowledgements"}}');var r=i(4848),s=i(8453);const o={title:"Citation and references",description:"Citation and references",sidebar_position:11},a=void 0,l={},c=[{value:"How to cite",id:"how-to-cite",level:3},{value:"Additional references",id:"additional-references",level:3}];function d(e){const n={a:"a",h3:"h3",li:"li",p:"p",ul:"ul",...(0,s.R)(),...e.components};return(0,r.jsxs)(r.Fragment,{children:[(0,r.jsx)(n.h3,{id:"how-to-cite",children:"How to cite"}),"\n",(0,r.jsx)(n.p,{children:"To credit GATK-SV, please cite the following publication:"}),"\n",(0,r.jsxs)(n.ul,{children:["\n",(0,r.jsx)(n.li,{children:(0,r.jsx)(n.a,{href:"https://doi.org/10.1038/s41586-020-2287-8",children:"Collins RL et al. A structural variation reference for medical and population genetics. Nature. 2020 May;581(7809):444-451."})}),"\n"]}),"\n",(0,r.jsx)(n.h3,{id:"additional-references",children:"Additional references"}),"\n",(0,r.jsx)(n.p,{children:"The following is a selected list of studies that have used GATK-SV:"}),"\n",(0,r.jsxs)(n.ul,{children:["\n",(0,r.jsx)(n.li,{children:(0,r.jsx)(n.a,{href:"https://doi.org/10.1016/j.ajhg.2021.02.012",children:"Belyeu JR et al. De novo structural mutation rates and gamete-of-origin biases revealed through genome sequencing of 2,396 families. Am J Hum Genet. 2021 Apr 1;108(4):597-607."})}),"\n",(0,r.jsx)(n.li,{children:(0,r.jsx)(n.a,{href:"https://doi.org/10.1002/ana.26608",children:"Billingsley KJ et al. Genome-Wide Analysis of Structural Variants in Parkinson Disease. Ann Neurol. 2023 May;93(5):1012-1022."})}),"\n",(0,r.jsx)(n.li,{children:(0,r.jsx)(n.a,{href:"https://doi.org/10.1016/j.cell.2022.08.004",children:"Byrska-Bishop M et al. High-coverage whole-genome sequencing of the expanded 1000 Genomes Project cohort including 602 trios. Cell. 2022 Sep 1;185(18):3426-3440.e19."})}),"\n",(0,r.jsx)(n.li,{children:(0,r.jsx)(n.a,{href:"https://doi.org/10.1016/j.gim.2023.100918",children:"Diaz Perez KK et al. Rare variants found in clinical gene panels illuminate the genetic and allelic architecture of orofacial clefting. Genet Med. 2023 Oct;25(10):100918."})}),"\n",(0,r.jsx)(n.li,{children:(0,r.jsx)(n.a,{href:"https://doi.org/10.1101/2024.04.27.591484",children:"Gillani R et al. Rare germline structural variants increase risk for pediatric solid tumors. bioRxiv [Preprint]. 2024 Apr 29:2024.04.27.591484."})}),"\n",(0,r.jsx)(n.li,{children:(0,r.jsx)(n.a,{href:"https://doi.org/10.1038/s41467-020-16736-1",children:"Han L et al. Functional annotation of rare structural variation in the human brain. Nat Commun. 2020 Jun 12;11(1):2990."})}),"\n",(0,r.jsx)(n.li,{children:(0,r.jsx)(n.a,{href:"https://doi.org/10.1016/j.gim.2024.101216",children:"Jurgens JA et al. Expanding the genetics and phenotypes of ocular congenital cranial dysinnervation disorders. Genet Med. 2024 Jul 17:101216."})}),"\n",(0,r.jsx)(n.li,{children:(0,r.jsx)(n.a,{href:"https://doi.org/10.1016/j.xgen.2023.100316",children:"Kaivola K et al. Genome-wide structural variant analysis identifies risk loci for non-Alzheimer's dementias. Cell Genom. 2023 May 4;3(6):100316."})}),"\n",(0,r.jsx)(n.li,{children:(0,r.jsx)(n.a,{href:"https://doi.org/10.1101/gr.278378.123",children:"Koenig Z et al. A harmonized public resource of deeply sequenced diverse human genomes. Genome Res. 2024 Jun 25;34(5):796-809."})}),"\n",(0,r.jsx)(n.li,{children:(0,r.jsx)(n.a,{href:"https://doi.org/10.1038/s41467-024-52463-7",children:"Lee AS et al. A cell type-aware framework for nominating non-coding variants in Mendelian regulatory disorders. Nat Commun. 2024 Sep 27;15(1):8268."})}),"\n",(0,r.jsx)(n.li,{children:(0,r.jsx)(n.a,{href:"https://doi.org/10.1016/j.ajhg.2023.07.010",children:"Lowther C et al. Systematic evaluation of genome sequencing for the diagnostic assessment of autism spectrum disorder and fetal structural anomalies. Am J Hum Genet. 2023 Sep 7;110(9):1454-1469."})}),"\n",(0,r.jsx)(n.li,{children:(0,r.jsx)(n.a,{href:"https://doi.org/10.1038/s41588-018-0107-y",children:"Werling DM et al. An analytical framework for whole-genome sequence association studies and its implications for autism spectrum disorder. Nat Genet. 2018 Apr 26;50(5):727-736."})}),"\n",(0,r.jsx)(n.li,{children:(0,r.jsx)(n.a,{href:"https://doi.org/10.1056/nejmoa2314761",children:"Wojcik MH et al. Genome Sequencing for Diagnosing Rare Diseases. N Engl J Med. 2024 Jun 6;390(21):1985-1997."})}),"\n",(0,r.jsx)(n.li,{children:(0,r.jsx)(n.a,{href:"https://doi.org/10.1038/s41431-022-01137-3",children:"Zhao B et al. A neurodevelopmental disorder caused by a novel de novo SVA insertion in exon 13 of the SRCAP gene. Eur J Hum Genet. 2022 Sep;30(9):1083-1087."})}),"\n"]})]})}function h(e={}){const{wrapper:n}={...(0,s.R)(),...e.components};return n?(0,r.jsx)(n,{...e,children:(0,r.jsx)(d,{...e})}):d(e)}},8453:(e,n,i)=>{i.d(n,{R:()=>o,x:()=>a});var t=i(6540);const r={},s=t.createContext(r);function o(e){const n=t.useContext(s);return t.useMemo((function(){return"function"==typeof e?e(n):{...n,...e}}),[n,e])}function a(e){let n;return n=e.disableParentContext?"function"==typeof e.components?e.components(r):e.components||r:o(e.components),t.createElement(s.Provider,{value:n},e.children)}}}]);