"use strict";(globalThis.webpackChunkGATK_SV=globalThis.webpackChunkGATK_SV||[]).push([[5684],{1944(e,t,i){i.d(t,{$:()=>n});var l=i(4848);const n=({children:e})=>(0,l.jsx)("span",{style:{backgroundColor:"var(--highlight-optional-arg-background-color)",borderRadius:"2px",color:"var(--highlight-optional-arg-text-color)",padding:"0.2rem"},children:e})},3447(e,t,i){i.r(t),i.d(t,{assets:()=>c,contentTitle:()=>a,default:()=>p,frontMatter:()=>s,metadata:()=>l,toc:()=>d});const l=JSON.parse('{"id":"modules/filter_genotypes","title":"FilterGenotypes","description":"Filters genotypes using GQ recalibrator model","source":"@site/docs/modules/filter_genotypes.md","sourceDirName":"modules","slug":"/modules/fg","permalink":"/gatk-sv/docs/modules/fg","draft":false,"unlisted":false,"editUrl":"https://github.com/broadinstitute/gatk-sv/tree/master/website/docs/modules/filter_genotypes.md","tags":[],"version":"current","sidebarPosition":20,"frontMatter":{"title":"FilterGenotypes","description":"Filters genotypes using GQ recalibrator model","sidebar_position":20,"slug":"fg"},"sidebar":"tutorialSidebar","previous":{"title":"ScoreGenotypes","permalink":"/gatk-sv/docs/modules/sg"},"next":{"title":"AnnotateVcf","permalink":"/gatk-sv/docs/modules/av"}}');var n=i(4848),o=i(8453),r=i(1944);const s={title:"FilterGenotypes",description:"Filters genotypes using GQ recalibrator model",sidebar_position:20,slug:"fg"},a=void 0,c={},d=[{value:"QC recommendations",id:"qc-recommendations",level:3},{value:"Inputs",id:"inputs",level:3},{value:"<code>vcf</code>",id:"vcf",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>output_prefix</code>",id:"optional-output_prefix",level:4},{value:"<code>ploidy_table</code>",id:"ploidy_table",level:4},{value:"<code>sl_cutoff_table</code>",id:"sl_cutoff_table",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>optimized_sl_cutoff_table</code>",id:"optional-optimized_sl_cutoff_table",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>no_call_rate_cutoff</code>",id:"optional-no_call_rate_cutoff",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>sl_filter_args</code>",id:"optional-sl_filter_args",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>run_qc</code>",id:"optional-run_qc",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>filter_vcf_records_per_shard</code>",id:"optional-filter_vcf_records_per_shard",level:4},{value:"Outputs",id:"outputs",level:3},{value:"<code>filtered_vcf</code>",id:"filtered_vcf",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>main_vcf_qc_tarball</code>",id:"optional-main_vcf_qc_tarball",level:4}];function h(e){const t={a:"a",code:"code",em:"em",h3:"h3",h4:"h4",mermaid:"mermaid",p:"p",...(0,o.R)(),...e.components};return(0,n.jsxs)(n.Fragment,{children:[(0,n.jsx)(t.p,{children:(0,n.jsx)(t.a,{href:"https://github.com/broadinstitute/gatk-sv/blob/main/wdl/FilterGenotypes.wdl",children:"WDL source code"})}),"\n",(0,n.jsxs)(t.p,{children:["Filter genotypes using the GQ model with recalibrated quality scores. The output VCF contains the ",(0,n.jsx)(t.code,{children:"HIGH_NCR"}),"  field, which is a filter status assigned to variants exceeding a ",(0,n.jsx)(t.a,{href:"#optional-no_call_rate_cutoff",children:"threshold proportion"}),"\nof no-call genotypes. This will also be applied to variants with genotypes that have already been filtered in the input VCF."]}),"\n",(0,n.jsx)(t.p,{children:"The following diagram illustrates the recommended invocation order:"}),"\n",(0,n.jsx)(t.mermaid,{value:"\nstateDiagram\n  direction LR\n  \n  classDef inModules stroke-width:0px,fill:#caf0f8,color:#00509d\n  classDef thisModule font-weight:bold,stroke-width:0px,fill:#ff9900,color:white\n  classDef outModules stroke-width:0px,fill:#caf0f8,color:#00509d\n\n  sg: ScoreGenotypes\n  fg: FilterGenotypes\n  avcf: AnnotateVcf\n  sg --\x3e fg\n  fg --\x3e avcf\n  \n  class fg thisModule\n  class sg inModules\n  class avcf outModules"}),"\n",(0,n.jsx)(t.h3,{id:"qc-recommendations",children:"QC recommendations"}),"\n",(0,n.jsxs)(t.p,{children:["We strongly recommend performing call set QC after this module. By default, QC plotting is enabled with the ",(0,n.jsx)(t.a,{href:"#optional-run_qc",children:"run_qc"}),"\nargument. Users should carefully inspect the main plots from the ",(0,n.jsx)(t.a,{href:"#optional-main_vcf_qc_tarball",children:"main_vcf_qc_tarball"}),".\nPlease see the ",(0,n.jsx)(t.a,{href:"./mvqc",children:"MainVcfQc"})," module documentation for more information on interpreting these plots and recommended\nQC criteria."]}),"\n",(0,n.jsx)(t.h3,{id:"inputs",children:"Inputs"}),"\n",(0,n.jsx)(t.h4,{id:"vcf",children:(0,n.jsx)(t.code,{children:"vcf"})}),"\n",(0,n.jsxs)(t.p,{children:["Input VCF with recalibrated scores generated from ",(0,n.jsx)(t.a,{href:"./sg#score_genotype",children:"ScoreGenotypes"}),"."]}),"\n",(0,n.jsxs)(t.h4,{id:"optional-output_prefix",children:[(0,n.jsx)(r.$,{children:"Optional"})," ",(0,n.jsx)(t.code,{children:"output_prefix"})]}),"\n",(0,n.jsx)(t.p,{children:"Default: use input VCF filename. Prefix for the output VCF, such as the cohort name. May be alphanumeric with underscores."}),"\n",(0,n.jsx)(t.h4,{id:"ploidy_table",children:(0,n.jsx)(t.code,{children:"ploidy_table"})}),"\n",(0,n.jsxs)(t.p,{children:["Table of sample ploidies generated in ",(0,n.jsx)(t.a,{href:"./jrc#ploidy_table",children:"JoinRawCalls"}),"."]}),"\n",(0,n.jsx)(t.h4,{id:"sl_cutoff_table",children:(0,n.jsx)(t.code,{children:"sl_cutoff_table"})}),"\n",(0,n.jsxs)(t.p,{children:["An argument for the ",(0,n.jsx)(t.a,{href:"https://github.com/broadinstitute/gatk-sv/blob/main/src/sv-pipeline/scripts/apply_sl_filter.py",children:"SL filtering script"})," which is used to set ",(0,n.jsx)(t.a,{href:"./sg#sl-scores",children:(0,n.jsx)(t.em,{children:"scaled logit"})})," (",(0,n.jsx)(t.code,{children:"SL"}),") cutoffs for filtering. Overridden by ",(0,n.jsx)(t.a,{href:"#optimized_sl_cutoff_table",children:"optimized_sl_cutoff_table"}),"."]}),"\n",(0,n.jsxs)(t.h4,{id:"optional-optimized_sl_cutoff_table",children:[(0,n.jsx)(r.$,{children:"Optional"})," ",(0,n.jsx)(t.code,{children:"optimized_sl_cutoff_table"})]}),"\n",(0,n.jsxs)(t.p,{children:["This is an output from the ",(0,n.jsx)(t.a,{href:"https://github.com/broadinstitute/gatk-sv/blob/main/src/sv-pipeline/scripts/optimize_sl_filter.py",children:"SL optimization script"}),". This can be used to set ",(0,n.jsx)(t.code,{children:"SL"})," cutoffs for filtering in a more truth-aware manner. Overrides ",(0,n.jsx)(t.a,{href:"#sl_cutoff_table",children:"sl_cutoff_table"})," if passed."]}),"\n",(0,n.jsxs)(t.h4,{id:"optional-no_call_rate_cutoff",children:[(0,n.jsx)(r.$,{children:"Optional"})," ",(0,n.jsx)(t.code,{children:"no_call_rate_cutoff"})]}),"\n",(0,n.jsxs)(t.p,{children:["Default: ",(0,n.jsx)(t.code,{children:"0.05"}),". Threshold fraction of samples that must have no-call genotypes in order to filter a variant. Set to 1 to disable."]}),"\n",(0,n.jsxs)(t.h4,{id:"optional-sl_filter_args",children:[(0,n.jsx)(r.$,{children:"Optional"})," ",(0,n.jsx)(t.code,{children:"sl_filter_args"})]}),"\n",(0,n.jsxs)(t.p,{children:["Arguments for the ",(0,n.jsx)(t.a,{href:"https://github.com/broadinstitute/gatk-sv/blob/main/src/sv-pipeline/scripts/apply_sl_filter.py",children:"SL filtering script"}),"."]}),"\n",(0,n.jsxs)(t.h4,{id:"optional-run_qc",children:[(0,n.jsx)(r.$,{children:"Optional"})," ",(0,n.jsx)(t.code,{children:"run_qc"})]}),"\n",(0,n.jsxs)(t.p,{children:["Default: ",(0,n.jsx)(t.code,{children:"true"}),". Enable running ",(0,n.jsx)(t.a,{href:"./mvqc",children:"MainVcfQc"})," automatically. By default, filtered variants will be excluded from\nthe plots."]}),"\n",(0,n.jsxs)(t.h4,{id:"optional-filter_vcf_records_per_shard",children:[(0,n.jsx)(r.$,{children:"Optional"})," ",(0,n.jsx)(t.code,{children:"filter_vcf_records_per_shard"})]}),"\n",(0,n.jsxs)(t.p,{children:["Default: ",(0,n.jsx)(t.code,{children:"20000"}),". Shard size for scattered ",(0,n.jsx)(t.code,{children:"GQ"})," recalibration tasks. Decrease this if those steps are running slowly."]}),"\n",(0,n.jsx)(t.h3,{id:"outputs",children:"Outputs"}),"\n",(0,n.jsx)(t.h4,{id:"filtered_vcf",children:(0,n.jsx)(t.code,{children:"filtered_vcf"})}),"\n",(0,n.jsx)(t.p,{children:"Filtered VCF."}),"\n",(0,n.jsxs)(t.h4,{id:"optional-main_vcf_qc_tarball",children:[(0,n.jsx)(r.$,{children:"Optional"})," ",(0,n.jsx)(t.code,{children:"main_vcf_qc_tarball"})]}),"\n",(0,n.jsxs)(t.p,{children:["QC plots generated with ",(0,n.jsx)(t.a,{href:"./mvqc",children:"MainVcfQc"}),". Only generated if using ",(0,n.jsx)(t.a,{href:"#optional-run_qc",children:"run_qc"}),"."]})]})}function p(e={}){const{wrapper:t}={...(0,o.R)(),...e.components};return t?(0,n.jsx)(t,{...e,children:(0,n.jsx)(h,{...e})}):h(e)}},8453(e,t,i){i.d(t,{R:()=>r,x:()=>s});var l=i(6540);const n={},o=l.createContext(n);function r(e){const t=l.useContext(o);return l.useMemo(function(){return"function"==typeof e?e(t):{...t,...e}},[t,e])}function s(e){let t;return t=e.disableParentContext?"function"==typeof e.components?e.components(n):e.components||n:r(e.components),l.createElement(o.Provider,{value:t},e.children)}}}]);