"use strict";(self.webpackChunkGATK_SV=self.webpackChunkGATK_SV||[]).push([[9783],{6473:(e,t,s)=>{s.d(t,{$:()=>l});var r=s(4848);const l=e=>{let{children:t}=e;return(0,r.jsx)("span",{style:{backgroundColor:"var(--highlight-optional-arg-background-color)",borderRadius:"2px",color:"var(--highlight-optional-arg-text-color)",padding:"0.2rem"},children:t})}},8453:(e,t,s)=>{s.d(t,{R:()=>i,x:()=>c});var r=s(6540);const l={},n=r.createContext(l);function i(e){const t=r.useContext(n);return r.useMemo((function(){return"function"==typeof e?e(t):{...t,...e}}),[t,e])}function c(e){let t;return t=e.disableParentContext?"function"==typeof e.components?e.components(l):e.components||l:i(e.components),r.createElement(n.Provider,{value:t},e.children)}},9965:(e,t,s)=>{s.r(t),s.d(t,{assets:()=>o,contentTitle:()=>d,default:()=>h,frontMatter:()=>c,metadata:()=>r,toc:()=>a});const r=JSON.parse('{"id":"modules/cluster_batch","title":"ClusterBatch","description":"Cluster Batch","source":"@site/docs/modules/cluster_batch.md","sourceDirName":"modules","slug":"/modules/cb","permalink":"/gatk-sv/docs/modules/cb","draft":false,"unlisted":false,"editUrl":"https://github.com/broadinstitute/gatk-sv/tree/master/website/docs/modules/cluster_batch.md","tags":[],"version":"current","sidebarPosition":5,"frontMatter":{"title":"ClusterBatch","description":"Cluster Batch","sidebar_position":5,"slug":"cb"},"sidebar":"tutorialSidebar","previous":{"title":"GatherBatchEvidence","permalink":"/gatk-sv/docs/modules/gbe"},"next":{"title":"GenerateBatchMetrics","permalink":"/gatk-sv/docs/modules/gbm"}}');var l=s(4848),n=s(8453),i=s(6473);const c={title:"ClusterBatch",description:"Cluster Batch",sidebar_position:5,slug:"cb"},d=void 0,o={},a=[{value:"Inputs",id:"inputs",level:2},{value:"<code>batch</code>",id:"batch",level:4},{value:"<code>*_vcf_tar</code>",id:"_vcf_tar",level:4},{value:"<code>del_bed</code>, <code>dup_bed</code>",id:"del_bed-dup_bed",level:4},{value:"<code>ped_file</code>",id:"ped_file",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg>  <code>N_IQR_cutoff_plotting</code>",id:"optional--n_iqr_cutoff_plotting",level:4},{value:"Outputs",id:"outputs",level:2},{value:"<code>clustered_*_vcf</code>",id:"clustered__vcf",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg>  <code>clustered_sv_counts</code>, <code>clustered_sv_count_plots</code>, <code>clustered_outlier_samples_preview</code>, <code>clustered_outlier_samples_with_reason</code>, <code>clustered_num_outlier_samples</code>",id:"optional--clustered_sv_counts-clustered_sv_count_plots-clustered_outlier_samples_preview-clustered_outlier_samples_with_reason-clustered_num_outlier_samples",level:4}];function u(e){const t={a:"a",admonition:"admonition",code:"code",h2:"h2",h4:"h4",mermaid:"mermaid",p:"p",...(0,n.R)(),...e.components};return(0,l.jsxs)(l.Fragment,{children:[(0,l.jsx)(t.p,{children:(0,l.jsx)(t.a,{href:"https://github.com/broadinstitute/gatk-sv/blob/main/wdl/ClusterBatch.wdl",children:"WDL source code"})}),"\n",(0,l.jsxs)(t.p,{children:["Clusters SV calls across a batch. For each caller, redundant variants are merged across samples\ninto representative variant records based on interval overlap criteria. Some variants will be hard-filtered\nif they overlap with predefined intervals known to pose challenges to SV and CNV callers (e.g. centromeres).\n",(0,l.jsx)(t.a,{href:"https://gatk.broadinstitute.org/hc/en-us/articles/27007962371099-SVCluster-BETA",children:"GATK-SVCluster"}),"\nis the primary tool used in for variant clustering."]}),"\n",(0,l.jsx)(t.p,{children:"The following diagram illustrates the recommended invocation order:"}),"\n",(0,l.jsx)(t.mermaid,{value:"\nstateDiagram\n  direction LR\n  \n  classDef inModules stroke-width:0px,fill:#caf0f8,color:#00509d\n  classDef thisModule font-weight:bold,stroke-width:0px,fill:#ff9900,color:white\n  classDef outModules stroke-width:0px,fill:#caf0f8,color:#00509d\n\n  gbe: GatherBatchEvidence\n  cb: ClusterBatch\n  gbm: GenerateBatchMetrics\n  jrc: JoinRawCalls\n  \n  gbe --\x3e cb\n  cb --\x3e gbm\n  cb --\x3e jrc\n  \n  class cb thisModule\n  class gbe inModules\n  class gbm outModules\n  class jrc outModules"}),"\n",(0,l.jsx)(t.admonition,{type:"note",children:(0,l.jsxs)(t.p,{children:[(0,l.jsx)(t.a,{href:"./gbm",children:"GenerateBatchMetrics"})," is the primary downstream module in batch processing. ",(0,l.jsx)(t.a,{href:"./jrc",children:"JoinRawCalls"})," is\nrequired for genotype filtering but does not need to be run until later in the pipeline."]})}),"\n",(0,l.jsx)(t.h2,{id:"inputs",children:"Inputs"}),"\n",(0,l.jsx)(t.h4,{id:"batch",children:(0,l.jsx)(t.code,{children:"batch"})}),"\n",(0,l.jsxs)(t.p,{children:["An identifier for the batch. Should match the name used in ",(0,l.jsx)(t.a,{href:"./gbe#batch",children:"GatherBatchEvidence"}),"."]}),"\n",(0,l.jsx)(t.h4,{id:"_vcf_tar",children:(0,l.jsx)(t.code,{children:"*_vcf_tar"})}),"\n",(0,l.jsxs)(t.p,{children:["Standardized VCF tarballs from ",(0,l.jsx)(t.a,{href:"./gbe#std__vcf_tar",children:"GatherBatchEvidence"})]}),"\n",(0,l.jsxs)(t.h4,{id:"del_bed-dup_bed",children:[(0,l.jsx)(t.code,{children:"del_bed"}),", ",(0,l.jsx)(t.code,{children:"dup_bed"})]}),"\n",(0,l.jsxs)(t.p,{children:["Merged CNV call files (",(0,l.jsx)(t.code,{children:".bed.gz"}),") from ",(0,l.jsx)(t.a,{href:"./gbe#merged_dels-merged_dups",children:"GatherBatchEvidence"})]}),"\n",(0,l.jsx)(t.h4,{id:"ped_file",children:(0,l.jsx)(t.code,{children:"ped_file"})}),"\n",(0,l.jsxs)(t.p,{children:["Family structures and sex assignments determined in ",(0,l.jsx)(t.a,{href:"./eqc",children:"EvidenceQC"}),". See ",(0,l.jsx)(t.a,{href:"/docs/gs/inputs#ped-format",children:"PED file format"}),"."]}),"\n",(0,l.jsxs)(t.h4,{id:"optional--n_iqr_cutoff_plotting",children:[(0,l.jsx)(i.$,{children:"Optional"}),"  ",(0,l.jsx)(t.code,{children:"N_IQR_cutoff_plotting"})]}),"\n",(0,l.jsx)(t.p,{children:"If provided, plot SV counts per sample. This number is used as the cutoff of interquartile range multiples for flagging\noutlier samples. Example value: 4."}),"\n",(0,l.jsx)(t.h2,{id:"outputs",children:"Outputs"}),"\n",(0,l.jsx)(t.h4,{id:"clustered__vcf",children:(0,l.jsx)(t.code,{children:"clustered_*_vcf"})}),"\n",(0,l.jsxs)(t.p,{children:["Clustered variants for each caller (",(0,l.jsx)(t.code,{children:"depth"})," corresponds to depth-based CNV callers ",(0,l.jsx)(t.code,{children:"cnMOPS"})," and ",(0,l.jsx)(t.code,{children:"GATK-gCNV"}),") in VCF format."]}),"\n",(0,l.jsxs)(t.h4,{id:"optional--clustered_sv_counts-clustered_sv_count_plots-clustered_outlier_samples_preview-clustered_outlier_samples_with_reason-clustered_num_outlier_samples",children:[(0,l.jsx)(i.$,{children:"Optional"}),"  ",(0,l.jsx)(t.code,{children:"clustered_sv_counts"}),", ",(0,l.jsx)(t.code,{children:"clustered_sv_count_plots"}),", ",(0,l.jsx)(t.code,{children:"clustered_outlier_samples_preview"}),", ",(0,l.jsx)(t.code,{children:"clustered_outlier_samples_with_reason"}),", ",(0,l.jsx)(t.code,{children:"clustered_num_outlier_samples"})]}),"\n",(0,l.jsxs)(t.p,{children:["SV count QC tables and plots. Enable by providing ",(0,l.jsx)(t.a,{href:"#optional--n_iqr_cutoff_plotting",children:"N_IQR_cutoff_plotting"})]})]})}function h(e={}){const{wrapper:t}={...(0,n.R)(),...e.components};return t?(0,l.jsx)(t,{...e,children:(0,l.jsx)(u,{...e})}):u(e)}}}]);