"use strict";(self.webpackChunkGATK_SV=self.webpackChunkGATK_SV||[]).push([[7575],{6473:(e,n,i)=>{i.d(n,{$:()=>o});var t=i(4848);const o=e=>{let{children:n}=e;return(0,t.jsx)("span",{style:{backgroundColor:"var(--highlight-optional-arg-background-color)",borderRadius:"2px",color:"var(--highlight-optional-arg-text-color)",padding:"0.2rem"},children:n})}},7702:(e,n,i)=>{i.r(n),i.d(n,{assets:()=>d,contentTitle:()=>s,default:()=>p,frontMatter:()=>r,metadata:()=>t,toc:()=>c});const t=JSON.parse('{"id":"modules/annotate_vcf","title":"AnnotateVcf","description":"Annotate VCF","source":"@site/docs/modules/annotate_vcf.md","sourceDirName":"modules","slug":"/modules/av","permalink":"/gatk-sv/docs/modules/av","draft":false,"unlisted":false,"editUrl":"https://github.com/broadinstitute/gatk-sv/tree/master/website/docs/modules/annotate_vcf.md","tags":[],"version":"current","sidebarPosition":20,"frontMatter":{"title":"AnnotateVcf","description":"Annotate VCF","sidebar_position":20,"slug":"av"},"sidebar":"tutorialSidebar","previous":{"title":"FilterGenotypes","permalink":"/gatk-sv/docs/modules/fg"},"next":{"title":"MainVcfQC","permalink":"/gatk-sv/docs/modules/mvqc"}}');var o=i(4848),l=i(8453),a=i(6473);const r={title:"AnnotateVcf",description:"Annotate VCF",sidebar_position:20,slug:"av"},s=void 0,d={},c=[{value:"Inputs",id:"inputs",level:3},{value:"<code>vcf</code>",id:"vcf",level:4},{value:"<code>prefix</code>",id:"prefix",level:4},{value:"<code>protein_coding_gtf</code>",id:"protein_coding_gtf",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>noncoding_bed</code>",id:"optional-noncoding_bed",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>promoter_window</code>",id:"optional-promoter_window",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>max_breakend_as_cnv_length</code>",id:"optional-max_breakend_as_cnv_length",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>svannotate_additional_args</code>",id:"optional-svannotate_additional_args",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>sample_pop_assignments</code>",id:"optional-sample_pop_assignments",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>sample_keep_list</code>",id:"optional-sample_keep_list",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>ped_file</code>",id:"optional-ped_file",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>par_bed</code>",id:"optional-par_bed",level:4},{value:"<code>sv_per_shard</code>",id:"sv_per_shard",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>external_af_ref_bed</code>",id:"optional-external_af_ref_bed",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>external_af_ref_prefix</code>",id:"optional-external_af_ref_prefix",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>external_af_population</code>",id:"optional-external_af_population",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>use_hail</code>",id:"optional-use_hail",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>gcs_project</code>",id:"optional-gcs_project",level:4},{value:"Outputs",id:"outputs",level:3},{value:"<code>annotated_vcf</code>",id:"annotated_vcf",level:4}];function h(e){const n={a:"a",code:"code",h3:"h3",h4:"h4",li:"li",mermaid:"mermaid",p:"p",ul:"ul",...(0,l.R)(),...e.components};return(0,o.jsxs)(o.Fragment,{children:[(0,o.jsx)(n.p,{children:"Adds annotations, such as the inferred function and allele frequencies of variants, to a VCF."}),"\n",(0,o.jsx)(n.p,{children:"Annotations methods include:"}),"\n",(0,o.jsxs)(n.ul,{children:["\n",(0,o.jsxs)(n.li,{children:["Functional annotation - The GATK tool ",(0,o.jsx)(n.a,{href:"https://gatk.broadinstitute.org/hc/en-us/articles/13832752531355-SVAnnotate",children:"SVAnnotate"}),"\nis used to annotate SVs with inferred functional consequence on protein-coding regions, regulatory regions such as\nUTR and promoters, and other non-coding elements."]}),"\n",(0,o.jsxs)(n.li,{children:["Allele Frequency (",(0,o.jsx)(n.code,{children:"AF"}),") annotation - annotate SVs with their allele frequencies across all samples, and samples of\nspecific sex, as well as specific subpopulations."]}),"\n",(0,o.jsx)(n.li,{children:"Allele Frequency annotation with external callset - annotate SVs with the allele frequencies of their overlapping SVs\nin another callset, eg. the gnomAD-SV reference callset."}),"\n"]}),"\n",(0,o.jsx)(n.p,{children:"The following diagram illustrates the recommended invocation order:"}),"\n",(0,o.jsx)(n.mermaid,{value:"\nstateDiagram\n  direction LR\n  \n  classDef inModules stroke-width:0px,fill:#caf0f8,color:#00509d\n  classDef thisModule font-weight:bold,stroke-width:0px,fill:#ff9900,color:white\n  classDef outModules stroke-width:0px,fill:#caf0f8,color:#00509d\n\n  fg: FilterGenotypes\n  avcf: AnnotateVcf\n  fg --\x3e avcf\n  \n  class avcf thisModule\n  class fg inModules"}),"\n",(0,o.jsx)(n.h3,{id:"inputs",children:"Inputs"}),"\n",(0,o.jsx)(n.h4,{id:"vcf",children:(0,o.jsx)(n.code,{children:"vcf"})}),"\n",(0,o.jsxs)(n.p,{children:["Any SV VCF. Running on the ",(0,o.jsx)(n.a,{href:"./fg#filtered_vcf",children:"genotype filtered VCF"})," is recommended."]}),"\n",(0,o.jsx)(n.h4,{id:"prefix",children:(0,o.jsx)(n.code,{children:"prefix"})}),"\n",(0,o.jsx)(n.p,{children:"Prefix for the output VCF, such as the cohort name. May be alphanumeric with underscores."}),"\n",(0,o.jsx)(n.h4,{id:"protein_coding_gtf",children:(0,o.jsx)(n.code,{children:"protein_coding_gtf"})}),"\n",(0,o.jsxs)(n.p,{children:["Coding transcript definitions, see ",(0,o.jsx)(n.a,{href:"/docs/resources#protein_coding_gtf",children:"here"}),"."]}),"\n",(0,o.jsxs)(n.h4,{id:"optional-noncoding_bed",children:[(0,o.jsx)(a.$,{children:"Optional"})," ",(0,o.jsx)(n.code,{children:"noncoding_bed"})]}),"\n",(0,o.jsxs)(n.p,{children:["Non-coding reference intervals, see ",(0,o.jsx)(n.a,{href:"//docs/resources#noncoding_bed",children:"here"}),"."]}),"\n",(0,o.jsxs)(n.h4,{id:"optional-promoter_window",children:[(0,o.jsx)(a.$,{children:"Optional"})," ",(0,o.jsx)(n.code,{children:"promoter_window"})]}),"\n",(0,o.jsxs)(n.p,{children:["Promoter window size. See ",(0,o.jsx)(n.a,{href:"https://gatk.broadinstitute.org/hc/en-us/articles/27007964610331-SVAnnotate#--promoter-window-length",children:"here"}),"."]}),"\n",(0,o.jsxs)(n.h4,{id:"optional-max_breakend_as_cnv_length",children:[(0,o.jsx)(a.$,{children:"Optional"})," ",(0,o.jsx)(n.code,{children:"max_breakend_as_cnv_length"})]}),"\n",(0,o.jsxs)(n.p,{children:["Max size for treating ",(0,o.jsx)(n.code,{children:"BND"})," records as CNVs. See ",(0,o.jsx)(n.a,{href:"https://gatk.broadinstitute.org/hc/en-us/articles/27007964610331-SVAnnotate#--max-breakend-as-cnv-length",children:"here"}),"."]}),"\n",(0,o.jsxs)(n.h4,{id:"optional-svannotate_additional_args",children:[(0,o.jsx)(a.$,{children:"Optional"})," ",(0,o.jsx)(n.code,{children:"svannotate_additional_args"})]}),"\n",(0,o.jsxs)(n.p,{children:["Additional arguments for ",(0,o.jsx)(n.a,{href:"https://gatk.broadinstitute.org/hc/en-us/articles/27007964610331-SVAnnotate",children:"GATK-SVAnnotate"}),"."]}),"\n",(0,o.jsxs)(n.h4,{id:"optional-sample_pop_assignments",children:[(0,o.jsx)(a.$,{children:"Optional"})," ",(0,o.jsx)(n.code,{children:"sample_pop_assignments"})]}),"\n",(0,o.jsx)(n.p,{children:'Two-column file with sample ID & population assignment. "." for population will ignore the sample. If provided,\nannotates population-specific allele frequencies.'}),"\n",(0,o.jsxs)(n.h4,{id:"optional-sample_keep_list",children:[(0,o.jsx)(a.$,{children:"Optional"})," ",(0,o.jsx)(n.code,{children:"sample_keep_list"})]}),"\n",(0,o.jsx)(n.p,{children:"If provided, subset samples to this list in the output VCF."}),"\n",(0,o.jsxs)(n.h4,{id:"optional-ped_file",children:[(0,o.jsx)(a.$,{children:"Optional"})," ",(0,o.jsx)(n.code,{children:"ped_file"})]}),"\n",(0,o.jsxs)(n.p,{children:["Family structures and sex assignments determined in ",(0,o.jsx)(n.a,{href:"./eqc",children:"EvidenceQC"}),". See ",(0,o.jsx)(n.a,{href:"/docs/gs/inputs#ped-format",children:"PED file format"}),".\nIf provided, sex-specific allele frequencies will be annotated."]}),"\n",(0,o.jsxs)(n.h4,{id:"optional-par_bed",children:[(0,o.jsx)(a.$,{children:"Optional"})," ",(0,o.jsx)(n.code,{children:"par_bed"})]}),"\n",(0,o.jsxs)(n.p,{children:["Pseudo-autosomal region (PAR) bed file. If provided, variants overlapping PARs will be annotated with the ",(0,o.jsx)(n.code,{children:"PAR"})," field."]}),"\n",(0,o.jsx)(n.h4,{id:"sv_per_shard",children:(0,o.jsx)(n.code,{children:"sv_per_shard"})}),"\n",(0,o.jsx)(n.p,{children:"Shard size for parallel processing. Decreasing this may help if the workflow is running too slowly."}),"\n",(0,o.jsxs)(n.h4,{id:"optional-external_af_ref_bed",children:[(0,o.jsx)(a.$,{children:"Optional"})," ",(0,o.jsx)(n.code,{children:"external_af_ref_bed"})]}),"\n",(0,o.jsxs)(n.p,{children:["Reference SV set (see ",(0,o.jsx)(n.a,{href:"/docs/resources#external_af_ref_bed",children:"here"}),"). If provided, annotates variants with allele frequencies\nfrom the reference population."]}),"\n",(0,o.jsxs)(n.h4,{id:"optional-external_af_ref_prefix",children:[(0,o.jsx)(a.$,{children:"Optional"})," ",(0,o.jsx)(n.code,{children:"external_af_ref_prefix"})]}),"\n",(0,o.jsxs)(n.p,{children:["External ",(0,o.jsx)(n.code,{children:"AF"})," annotation prefix. Required if providing ",(0,o.jsx)(n.a,{href:"#optional-external_af_ref_bed",children:"external_af_ref_bed"}),"."]}),"\n",(0,o.jsxs)(n.h4,{id:"optional-external_af_population",children:[(0,o.jsx)(a.$,{children:"Optional"})," ",(0,o.jsx)(n.code,{children:"external_af_population"})]}),"\n",(0,o.jsxs)(n.p,{children:['Population names in the external SV reference set, e.g. "ALL", "AFR", "AMR", "EAS", "EUR". Required if providing\n',(0,o.jsx)(n.a,{href:"#optional-external_af_ref_bed",children:"external_af_ref_bed"})," and must match the populations in the bed file."]}),"\n",(0,o.jsxs)(n.h4,{id:"optional-use_hail",children:[(0,o.jsx)(a.$,{children:"Optional"})," ",(0,o.jsx)(n.code,{children:"use_hail"})]}),"\n",(0,o.jsxs)(n.p,{children:["Default: ",(0,o.jsx)(n.code,{children:"false"}),". Use Hail for VCF concatenation. This should only be used for projects with over 50k samples. If enabled, the\n",(0,o.jsx)(n.a,{href:"#optional-gcs_project",children:"gcs_project"})," must also be provided. Does not work on Terra."]}),"\n",(0,o.jsxs)(n.h4,{id:"optional-gcs_project",children:[(0,o.jsx)(a.$,{children:"Optional"})," ",(0,o.jsx)(n.code,{children:"gcs_project"})]}),"\n",(0,o.jsxs)(n.p,{children:["Google Cloud project ID. Required only if enabling ",(0,o.jsx)(n.a,{href:"#optional-use_hail",children:"use_hail"}),"."]}),"\n",(0,o.jsx)(n.h3,{id:"outputs",children:"Outputs"}),"\n",(0,o.jsx)(n.h4,{id:"annotated_vcf",children:(0,o.jsx)(n.code,{children:"annotated_vcf"})}),"\n",(0,o.jsx)(n.p,{children:"Output VCF."})]})}function p(e={}){const{wrapper:n}={...(0,l.R)(),...e.components};return n?(0,o.jsx)(n,{...e,children:(0,o.jsx)(h,{...e})}):h(e)}},8453:(e,n,i)=>{i.d(n,{R:()=>a,x:()=>r});var t=i(6540);const o={},l=t.createContext(o);function a(e){const n=t.useContext(l);return t.useMemo((function(){return"function"==typeof e?e(n):{...n,...e}}),[n,e])}function r(e){let n;return n=e.disableParentContext?"function"==typeof e.components?e.components(o):e.components||o:a(e.components),t.createElement(l.Provider,{value:n},e.children)}}}]);