"use strict";(self.webpackChunkGATK_SV=self.webpackChunkGATK_SV||[]).push([[6790],{2496:(e,t,i)=>{i.r(t),i.d(t,{assets:()=>d,contentTitle:()=>c,default:()=>u,frontMatter:()=>o,metadata:()=>l,toc:()=>a});const l=JSON.parse('{"id":"modules/filter_batch","title":"FilterBatch","description":"Filter Batch","source":"@site/docs/modules/filter_batch.md","sourceDirName":"modules","slug":"/modules/fb","permalink":"/gatk-sv/docs/modules/fb","draft":false,"unlisted":false,"editUrl":"https://github.com/broadinstitute/gatk-sv/tree/master/website/docs/modules/filter_batch.md","tags":[],"version":"current","sidebarPosition":7,"frontMatter":{"title":"FilterBatch","description":"Filter Batch","sidebar_position":7,"slug":"fb"},"sidebar":"tutorialSidebar","previous":{"title":"GenerateBatchMetrics","permalink":"/gatk-sv/docs/modules/gbm"},"next":{"title":"MergeBatchSites","permalink":"/gatk-sv/docs/modules/msites"}}');var s=i(4848),n=i(8453),r=i(6473);const o={title:"FilterBatch",description:"Filter Batch",sidebar_position:7,slug:"fb"},c=void 0,d={},a=[{value:"Inputs",id:"inputs",level:3},{value:"<code>batch</code>",id:"batch",level:4},{value:"<code>*_vcf</code>",id:"_vcf",level:4},{value:"<code>evidence_metrics</code>",id:"evidence_metrics",level:4},{value:"<code>evidence_metrics_common</code>",id:"evidence_metrics_common",level:4},{value:"<code>outlier_cutoff_nIQR</code>",id:"outlier_cutoff_niqr",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg>  <code>outlier_cutoff_table</code>",id:"optional--outlier_cutoff_table",level:4},{value:"Outputs",id:"outputs",level:3},{value:"<code>filtered_depth_vcf</code>",id:"filtered_depth_vcf",level:4},{value:"<code>filtered_pesr_vcf</code>",id:"filtered_pesr_vcf",level:4},{value:"<code>cutoffs</code>",id:"cutoffs",level:4},{value:"<code>sv_counts</code>",id:"sv_counts",level:4},{value:"<code>sv_count_plots</code>",id:"sv_count_plots",level:4},{value:"<code>outlier_samples_excluded</code>",id:"outlier_samples_excluded",level:4},{value:"<code>outlier_samples_excluded_file</code>",id:"outlier_samples_excluded_file",level:4},{value:"<code>batch_samples_postOutlierExclusion</code>",id:"batch_samples_postoutlierexclusion",level:4},{value:"<code>batch_samples_postOutlierExclusion_file</code>",id:"batch_samples_postoutlierexclusion_file",level:4}];function h(e){const t={a:"a",code:"code",h3:"h3",h4:"h4",li:"li",mermaid:"mermaid",ol:"ol",p:"p",...(0,n.R)(),...e.components};return(0,s.jsxs)(s.Fragment,{children:[(0,s.jsx)(t.p,{children:(0,s.jsx)(t.a,{href:"https://github.com/broadinstitute/gatk-sv/blob/main/wdl/FilterBatch.wdl",children:"WDL source code"})}),"\n",(0,s.jsx)(t.p,{children:"Filters poor quality variants and outlier samples.\nThis workflow can be run all at once with the top-level WDL,\nor it can be run in two steps to enable tuning of outlier\nfiltration cutoffs. The two subworkflows are:"}),"\n",(0,s.jsxs)(t.ol,{children:["\n",(0,s.jsxs)(t.li,{children:["\n",(0,s.jsxs)(t.p,{children:[(0,s.jsx)(t.a,{href:"https://github.com/broadinstitute/gatk-sv/blob/main/wdl/FilterBatchSites.wdl",children:"FilterBatchSites"}),": Per-batch variant filtration.\nVisualize filtered SV counts per sample per type to help choose an IQR cutoff for outlier sample filtering, and preview\noutlier samples for a given cutoff."]}),"\n"]}),"\n",(0,s.jsxs)(t.li,{children:["\n",(0,s.jsxs)(t.p,{children:[(0,s.jsx)(t.a,{href:"https://github.com/broadinstitute/gatk-sv/blob/main/wdl/FilterBatchSamples.wdl",children:"FilterBatchSamples"}),": Per-batch outlier sample filtration;\nprovide an appropriate ",(0,s.jsx)(t.a,{href:"#outlier_cutoff_niqr",children:"outlier_cutoff_nIQR"})," based on the\nSV count plots and outlier previews from step 2. Note\nthat not removing high outliers can result in increased\ncompute cost and a higher false positive rate in later steps."]}),"\n"]}),"\n"]}),"\n",(0,s.jsx)(t.p,{children:"The following diagram illustrates the recommended invocation order:"}),"\n",(0,s.jsx)(t.mermaid,{value:"\nstateDiagram\n  direction LR\n  \n  classDef inModules stroke-width:0px,fill:#caf0f8,color:#00509d\n  classDef thisModule font-weight:bold,stroke-width:0px,fill:#ff9900,color:white\n  classDef outModules stroke-width:0px,fill:#caf0f8,color:#00509d\n\n  gbm: GenerateBatchMetrics\n  fb: FilterBatch\n  mbs: MergeBatchSites\n  gbm --\x3e fb\n  fb --\x3e mbs\n  \n  class fb thisModule\n  class gbm inModules\n  class mbs outModules"}),"\n",(0,s.jsx)(t.h3,{id:"inputs",children:"Inputs"}),"\n",(0,s.jsx)(t.h4,{id:"batch",children:(0,s.jsx)(t.code,{children:"batch"})}),"\n",(0,s.jsxs)(t.p,{children:["An identifier for the batch. Should match the name used in ",(0,s.jsx)(t.a,{href:"./gbe#batch",children:"GatherBatchEvidence"}),"."]}),"\n",(0,s.jsx)(t.h4,{id:"_vcf",children:(0,s.jsx)(t.code,{children:"*_vcf"})}),"\n",(0,s.jsxs)(t.p,{children:["Clustered VCFs from ",(0,s.jsx)(t.a,{href:"./cb#clustered__vcf",children:"ClusterBatch"})]}),"\n",(0,s.jsx)(t.h4,{id:"evidence_metrics",children:(0,s.jsx)(t.code,{children:"evidence_metrics"})}),"\n",(0,s.jsxs)(t.p,{children:["Metrics table ",(0,s.jsx)(t.a,{href:"./gbm#metrics",children:"GenerateBatchMetrics"})]}),"\n",(0,s.jsx)(t.h4,{id:"evidence_metrics_common",children:(0,s.jsx)(t.code,{children:"evidence_metrics_common"})}),"\n",(0,s.jsxs)(t.p,{children:["Common variant metrics table ",(0,s.jsx)(t.a,{href:"./gbm#metrics_common",children:"GenerateBatchMetrics"})]}),"\n",(0,s.jsx)(t.h4,{id:"outlier_cutoff_niqr",children:(0,s.jsx)(t.code,{children:"outlier_cutoff_nIQR"})}),"\n",(0,s.jsxs)(t.p,{children:["Defines outlier sample cutoffs based on variant counts. Samples deviating from the batch median count by more than\nthe given multiple of the interquartile range are hard filtered from the VCF. Recommended range is between ",(0,s.jsx)(t.code,{children:"3"})," and ",(0,s.jsx)(t.code,{children:"9"}),"\ndepending on desired sensitivity (higher is less stringent), or disable with ",(0,s.jsx)(t.code,{children:"10000"}),"."]}),"\n",(0,s.jsxs)(t.h4,{id:"optional--outlier_cutoff_table",children:[(0,s.jsx)(r.$,{children:"Optional"}),"  ",(0,s.jsx)(t.code,{children:"outlier_cutoff_table"})]}),"\n",(0,s.jsxs)(t.p,{children:["A cutoff table to set permissible nIQR ranges for each SVTYPE. If provided, overrides ",(0,s.jsx)(t.code,{children:"outlier_cutoff_nIQR"}),". Expected\ncolumns are: ",(0,s.jsx)(t.code,{children:"algorithm"}),", ",(0,s.jsx)(t.code,{children:"svtype"}),", ",(0,s.jsx)(t.code,{children:"lower_cuff"}),", ",(0,s.jsx)(t.code,{children:"higher_cff"}),". See the ",(0,s.jsx)(t.code,{children:"outlier_cutoff_table"})," resource in\n",(0,s.jsx)(t.a,{href:"https://github.com/broadinstitute/gatk-sv/blob/main/inputs/values/ref_panel_1kg.json",children:"this json"})," for an example table."]}),"\n",(0,s.jsx)(t.h3,{id:"outputs",children:"Outputs"}),"\n",(0,s.jsx)(t.h4,{id:"filtered_depth_vcf",children:(0,s.jsx)(t.code,{children:"filtered_depth_vcf"})}),"\n",(0,s.jsx)(t.p,{children:"Depth-based CNV caller VCFs after variant and sample filtering."}),"\n",(0,s.jsx)(t.h4,{id:"filtered_pesr_vcf",children:(0,s.jsx)(t.code,{children:"filtered_pesr_vcf"})}),"\n",(0,s.jsx)(t.p,{children:"PE/SR (non-depth) caller VCFs after variant and sample filtering."}),"\n",(0,s.jsx)(t.h4,{id:"cutoffs",children:(0,s.jsx)(t.code,{children:"cutoffs"})}),"\n",(0,s.jsx)(t.p,{children:"Variant metric cutoffs for genotyping."}),"\n",(0,s.jsx)(t.h4,{id:"sv_counts",children:(0,s.jsx)(t.code,{children:"sv_counts"})}),"\n",(0,s.jsxs)(t.p,{children:["Array of TSVs containing SV counts for each sample, i.e. ",(0,s.jsx)(t.code,{children:"sample-svtype-count"})," triplets. Each file corresponds to\na different SV caller."]}),"\n",(0,s.jsx)(t.h4,{id:"sv_count_plots",children:(0,s.jsx)(t.code,{children:"sv_count_plots"})}),"\n",(0,s.jsx)(t.p,{children:"Array of images plotting SV counts stratified by SV type. Each file corresponds to a different SV caller."}),"\n",(0,s.jsx)(t.h4,{id:"outlier_samples_excluded",children:(0,s.jsx)(t.code,{children:"outlier_samples_excluded"})}),"\n",(0,s.jsx)(t.p,{children:"Array of sample IDs excluded by outlier analysis."}),"\n",(0,s.jsx)(t.h4,{id:"outlier_samples_excluded_file",children:(0,s.jsx)(t.code,{children:"outlier_samples_excluded_file"})}),"\n",(0,s.jsx)(t.p,{children:"Text file of sample IDs excluded by outlier analysis."}),"\n",(0,s.jsx)(t.h4,{id:"batch_samples_postoutlierexclusion",children:(0,s.jsx)(t.code,{children:"batch_samples_postOutlierExclusion"})}),"\n",(0,s.jsx)(t.p,{children:"Array of remaining sample IDs after outlier exclusion."}),"\n",(0,s.jsx)(t.h4,{id:"batch_samples_postoutlierexclusion_file",children:(0,s.jsx)(t.code,{children:"batch_samples_postOutlierExclusion_file"})}),"\n",(0,s.jsx)(t.p,{children:"Text file of remaining sample IDs after outlier exclusion."})]})}function u(e={}){const{wrapper:t}={...(0,n.R)(),...e.components};return t?(0,s.jsx)(t,{...e,children:(0,s.jsx)(h,{...e})}):h(e)}},6473:(e,t,i)=>{i.d(t,{$:()=>s});var l=i(4848);const s=e=>{let{children:t}=e;return(0,l.jsx)("span",{style:{backgroundColor:"var(--highlight-optional-arg-background-color)",borderRadius:"2px",color:"var(--highlight-optional-arg-text-color)",padding:"0.2rem"},children:t})}},8453:(e,t,i)=>{i.d(t,{R:()=>r,x:()=>o});var l=i(6540);const s={},n=l.createContext(s);function r(e){const t=l.useContext(n);return l.useMemo((function(){return"function"==typeof e?e(t):{...t,...e}}),[t,e])}function o(e){let t;return t=e.disableParentContext?"function"==typeof e.components?e.components(s):e.components||s:r(e.components),l.createElement(n.Provider,{value:t},e.children)}}}]);