"use strict";(self.webpackChunkGATK_SV=self.webpackChunkGATK_SV||[]).push([[8864],{782:(e,n,i)=>{i.r(n),i.d(n,{assets:()=>d,contentTitle:()=>t,default:()=>_,frontMatter:()=>a,metadata:()=>s,toc:()=>c});const s=JSON.parse('{"id":"resources","title":"Resource files","description":"Resource files manifest","source":"@site/docs/resources.md","sourceDirName":".","slug":"/resources","permalink":"/gatk-sv/docs/resources","draft":false,"unlisted":false,"editUrl":"https://github.com/broadinstitute/gatk-sv/tree/master/website/docs/resources.md","tags":[],"version":"current","sidebarPosition":7,"frontMatter":{"title":"Resource files","description":"Resource files manifest","sidebar_position":7},"sidebar":"tutorialSidebar","previous":{"title":"Runtime attributes","permalink":"/gatk-sv/docs/runtime_attr"},"next":{"title":"Troubleshooting","permalink":"/gatk-sv/docs/category/troubleshooting"}}');var l=i(4848),r=i(8453);const a={title:"Resource files",description:"Resource files manifest",sidebar_position:7},t=void 0,d={},c=[{value:"Reference resources",id:"reference-resources",level:3},{value:"allosome_file",id:"allosome_file",level:4},{value:"autosome_file",id:"autosome_file",level:4},{value:"bin_exclude",id:"bin_exclude",level:4},{value:"cnmops_exclude_list",id:"cnmops_exclude_list",level:4},{value:"contig_ploidy_priors",id:"contig_ploidy_priors",level:4},{value:"cytobands",id:"cytobands",level:4},{value:"sd_locs_vcf",id:"sd_locs_vcf",level:4},{value:"depth_exclude_list",id:"depth_exclude_list",level:4},{value:"empty_file",id:"empty_file",level:4},{value:"exclude_intervals_for_gcnv_filter_intervals",id:"exclude_intervals_for_gcnv_filter_intervals",level:4},{value:"external_af_ref_bed",id:"external_af_ref_bed",level:4},{value:"genome_file",id:"genome_file",level:4},{value:"manta_region_bed",id:"manta_region_bed",level:4},{value:"mei_bed",id:"mei_bed",level:4},{value:"melt_std_vcf_header",id:"melt_std_vcf_header",level:4},{value:"noncoding_bed",id:"noncoding_bed",level:4},{value:"par_bed",id:"par_bed",level:4},{value:"pesr_exclude_list",id:"pesr_exclude_list",level:4},{value:"preprocessed_intervals",id:"preprocessed_intervals",level:4},{value:"primary_contigs_fai",id:"primary_contigs_fai",level:4},{value:"primary_contigs_list",id:"primary_contigs_list",level:4},{value:"contigs_header",id:"contigs_header",level:4},{value:"protein_coding_gtf",id:"protein_coding_gtf",level:4},{value:"reference_dict",id:"reference_dict",level:4},{value:"reference_fasta",id:"reference_fasta",level:4},{value:"reference_index",id:"reference_index",level:4},{value:"rmsk",id:"rmsk",level:4},{value:"segdups",id:"segdups",level:4},{value:"seed_cutoffs",id:"seed_cutoffs",level:4},{value:"single_sample_qc_definitions",id:"single_sample_qc_definitions",level:4},{value:"wgd_scoring_mask",id:"wgd_scoring_mask",level:4},{value:"wham_include_list_bed_file",id:"wham_include_list_bed_file",level:4},{value:"aou_recalibrate_gq_model_file",id:"aou_recalibrate_gq_model_file",level:4},{value:"hgdp_recalibrate_gq_model_file",id:"hgdp_recalibrate_gq_model_file",level:4},{value:"recalibrate_gq_genome_tracks",id:"recalibrate_gq_genome_tracks",level:4},{value:"Benchmarking datasets",id:"benchmarking-datasets",level:3},{value:"ccdg_abel_site_level_benchmarking_dataset",id:"ccdg_abel_site_level_benchmarking_dataset",level:4},{value:"gnomad_v2_collins_sample_level_benchmarking_dataset",id:"gnomad_v2_collins_sample_level_benchmarking_dataset",level:4},{value:"gnomad_v2_collins_site_level_benchmarking_dataset",id:"gnomad_v2_collins_site_level_benchmarking_dataset",level:4},{value:"hgsv_byrska_bishop_sample_level_benchmarking_dataset",id:"hgsv_byrska_bishop_sample_level_benchmarking_dataset",level:4},{value:"hgsv_byrska_bishop_sample_renaming_tsv",id:"hgsv_byrska_bishop_sample_renaming_tsv",level:4},{value:"hgsv_byrska_bishop_site_level_benchmarking_dataset",id:"hgsv_byrska_bishop_site_level_benchmarking_dataset",level:4},{value:"hgsv_ebert_sample_level_benchmarking_dataset",id:"hgsv_ebert_sample_level_benchmarking_dataset",level:4},{value:"ssc_belyeu_sample_level_benchmarking_dataset",id:"ssc_belyeu_sample_level_benchmarking_dataset",level:4},{value:"ssc_belyeu_site_level_benchmarking_dataset",id:"ssc_belyeu_site_level_benchmarking_dataset",level:4},{value:"ssc_sanders_sample_level_benchmarking_dataset",id:"ssc_sanders_sample_level_benchmarking_dataset",level:4},{value:"thousand_genomes_site_level_benchmarking_dataset",id:"thousand_genomes_site_level_benchmarking_dataset",level:4},{value:"asc_site_level_benchmarking_dataset",id:"asc_site_level_benchmarking_dataset",level:4},{value:"hgsv_site_level_benchmarking_dataset",id:"hgsv_site_level_benchmarking_dataset",level:4},{value:"collins_2017_sample_level_benchmarking_dataset",id:"collins_2017_sample_level_benchmarking_dataset",level:4},{value:"sanders_2015_sample_level_benchmarking_dataset",id:"sanders_2015_sample_level_benchmarking_dataset",level:4},{value:"werling_2018_sample_level_benchmarking_dataset",id:"werling_2018_sample_level_benchmarking_dataset",level:4}];function o(e){const n={a:"a",admonition:"admonition",code:"code",h3:"h3",h4:"h4",p:"p",strong:"strong",...(0,r.R)(),...e.components};return(0,l.jsxs)(l.Fragment,{children:[(0,l.jsxs)(n.p,{children:["This page contains descriptions of common resource files used in the pipeline. All required files are publicly available\nin Google Cloud Storage Buckets. URIs for these files are available in the ",(0,l.jsx)(n.a,{href:"https://github.com/broadinstitute/gatk-sv/blob/main/inputs/values/resources_hg38.json",children:"hg38 resources json"}),"."]}),"\n",(0,l.jsx)(n.admonition,{type:"info",children:(0,l.jsx)(n.p,{children:"Some resources contain sensitive data and are stored in a secure bucket for development purposes. These files are not\nrequired to run GATK-SV."})}),"\n",(0,l.jsx)(n.h3,{id:"reference-resources",children:"Reference resources"}),"\n",(0,l.jsx)(n.h4,{id:"allosome_file",children:"allosome_file"}),"\n",(0,l.jsxs)(n.p,{children:["Reference ",(0,l.jsx)(n.a,{href:"https://www.htslib.org/doc/faidx.html",children:"fasta index file"})," containing only allosomal contigs."]}),"\n",(0,l.jsx)(n.h4,{id:"autosome_file",children:"autosome_file"}),"\n",(0,l.jsxs)(n.p,{children:["Reference ",(0,l.jsx)(n.a,{href:"https://www.htslib.org/doc/faidx.html",children:"fasta index file"})," containing only autosomal contigs."]}),"\n",(0,l.jsx)(n.h4,{id:"bin_exclude",children:"bin_exclude"}),"\n",(0,l.jsxs)(n.p,{children:[(0,l.jsx)(n.a,{href:"https://www.htslib.org/doc/bgzip.html",children:"Block-compressed"})," ",(0,l.jsx)(n.a,{href:"https://genome.ucsc.edu/FAQ/FAQformat.html#format1",children:"bed file"}),"\nof intervals to exclude in the call set."]}),"\n",(0,l.jsx)(n.h4,{id:"cnmops_exclude_list",children:"cnmops_exclude_list"}),"\n",(0,l.jsx)(n.p,{children:"Plain text bed file of N-masked regions in the reference."}),"\n",(0,l.jsx)(n.h4,{id:"contig_ploidy_priors",children:"contig_ploidy_priors"}),"\n",(0,l.jsx)(n.p,{children:"Plain text TSV file of prior probabilities contig ploidies used by GATK-gCNV."}),"\n",(0,l.jsx)(n.h4,{id:"cytobands",children:"cytobands"}),"\n",(0,l.jsx)(n.p,{children:"Block-compressed bed file of cytoband intervals."}),"\n",(0,l.jsx)(n.h4,{id:"sd_locs_vcf",children:"sd_locs_vcf"}),"\n",(0,l.jsx)(n.p,{children:"Plain text VCF of SNP sites at which to collect site depth (SD) evidence."}),"\n",(0,l.jsx)(n.h4,{id:"depth_exclude_list",children:"depth_exclude_list"}),"\n",(0,l.jsx)(n.p,{children:"Block-compressed bed file of intervals over which to exclude overlapping depth-only calls."}),"\n",(0,l.jsx)(n.h4,{id:"empty_file",children:"empty_file"}),"\n",(0,l.jsx)(n.p,{children:"Empty file; used to satisfy some workflow code paths."}),"\n",(0,l.jsx)(n.h4,{id:"exclude_intervals_for_gcnv_filter_intervals",children:"exclude_intervals_for_gcnv_filter_intervals"}),"\n",(0,l.jsx)(n.p,{children:"Plain text bed file of intervals to exclude from GATK-gCNV."}),"\n",(0,l.jsx)(n.h4,{id:"external_af_ref_bed",children:"external_af_ref_bed"}),"\n",(0,l.jsx)(n.p,{children:"Block-compressed bed file of SV sites for external allele frequency annotation."}),"\n",(0,l.jsx)(n.h4,{id:"genome_file",children:"genome_file"}),"\n",(0,l.jsx)(n.p,{children:"Tab-delimited table of primary contigs with contig name in the first column and length in the second column."}),"\n",(0,l.jsx)(n.h4,{id:"manta_region_bed",children:"manta_region_bed"}),"\n",(0,l.jsx)(n.p,{children:"Block-compressed bed file of intervals to call with Manta."}),"\n",(0,l.jsx)(n.h4,{id:"mei_bed",children:"mei_bed"}),"\n",(0,l.jsx)(n.p,{children:"Block-compressed bed file of mobile element insertions in the reference genome."}),"\n",(0,l.jsx)(n.h4,{id:"melt_std_vcf_header",children:"melt_std_vcf_header"}),"\n",(0,l.jsx)(n.p,{children:"Text file containing the VCF header for raw MELT calls."}),"\n",(0,l.jsx)(n.h4,{id:"noncoding_bed",children:"noncoding_bed"}),"\n",(0,l.jsx)(n.p,{children:"Plain text bed file of non-coding elements in the reference genome."}),"\n",(0,l.jsx)(n.h4,{id:"par_bed",children:"par_bed"}),"\n",(0,l.jsx)(n.p,{children:"Plain text bed file of pseudoautosomal regions."}),"\n",(0,l.jsx)(n.h4,{id:"pesr_exclude_list",children:"pesr_exclude_list"}),"\n",(0,l.jsx)(n.p,{children:"Block-compressed bed file of intervals for filtering calls. Variants generated with non-CNV tools (Manta, MELT,\nScramble, Wham) that have either end in any of these intervals are hard-filtered."}),"\n",(0,l.jsx)(n.h4,{id:"preprocessed_intervals",children:"preprocessed_intervals"}),"\n",(0,l.jsx)(n.p,{children:"Intervals for read count collection and CNV calling."}),"\n",(0,l.jsx)(n.h4,{id:"primary_contigs_fai",children:"primary_contigs_fai"}),"\n",(0,l.jsxs)(n.p,{children:["Reference ",(0,l.jsx)(n.a,{href:"https://www.htslib.org/doc/faidx.html",children:"fasta index file"})," only containing the primary contigs, i.e. ",(0,l.jsx)(n.code,{children:"chr1"}),", ...,\n",(0,l.jsx)(n.code,{children:"chr22"}),", ",(0,l.jsx)(n.code,{children:"chrX"}),", and ",(0,l.jsx)(n.code,{children:"chrY"}),"."]}),"\n",(0,l.jsx)(n.h4,{id:"primary_contigs_list",children:"primary_contigs_list"}),"\n",(0,l.jsx)(n.p,{children:"Text file of primary contig names."}),"\n",(0,l.jsx)(n.h4,{id:"contigs_header",children:"contigs_header"}),"\n",(0,l.jsx)(n.p,{children:"Plain text VCF header section of primary contig sequences."}),"\n",(0,l.jsx)(n.h4,{id:"protein_coding_gtf",children:"protein_coding_gtf"}),"\n",(0,l.jsxs)(n.p,{children:["Protein-coding sequence definitions for functional annotation in ",(0,l.jsx)(n.a,{href:"https://www.ensembl.org/info/website/upload/gff.html",children:"General Transfer Format"}),". This GTF was created by subsetting the ",(0,l.jsx)(n.a,{href:"https://www.gencodegenes.org/human/releases.html",children:"GENCODE"})," GRCh38 basic gene annotation GTF with the script ",(0,l.jsx)(n.code,{children:"scripts/inputs/preprocess_gtf.py"}),". Transcripts annotated as either Ensembl canonical or MANE Select Plus Clinical, and as either protein-coding or from nonsense-mediated decay, were retained. The GENCODE version is included in the filename."]}),"\n",(0,l.jsx)(n.h4,{id:"reference_dict",children:"reference_dict"}),"\n",(0,l.jsxs)(n.p,{children:["Reference FASTA dictionary file (",(0,l.jsx)(n.code,{children:"*.dict"}),"). See ",(0,l.jsx)(n.a,{href:"https://gatk.broadinstitute.org/hc/en-us/articles/360035531652-FASTA-Reference-genome-format",children:"this article"})," for more information."]}),"\n",(0,l.jsx)(n.h4,{id:"reference_fasta",children:"reference_fasta"}),"\n",(0,l.jsxs)(n.p,{children:["Reference FASTA file (",(0,l.jsx)(n.code,{children:"*.fasta"}),"). See ",(0,l.jsx)(n.a,{href:"https://gatk.broadinstitute.org/hc/en-us/articles/360035531652-FASTA-Reference-genome-format",children:"this article"})," for more information."]}),"\n",(0,l.jsx)(n.h4,{id:"reference_index",children:"reference_index"}),"\n",(0,l.jsxs)(n.p,{children:["Reference FASTA index file (",(0,l.jsx)(n.code,{children:"*.fasta.fai"}),"). See ",(0,l.jsx)(n.a,{href:"https://gatk.broadinstitute.org/hc/en-us/articles/360035531652-FASTA-Reference-genome-format",children:"this article"})," for more information."]}),"\n",(0,l.jsx)(n.h4,{id:"rmsk",children:"rmsk"}),"\n",(0,l.jsxs)(n.p,{children:["Block-compressed bed file of ",(0,l.jsx)(n.a,{href:"https://genome.ucsc.edu/cgi-bin/hgTrackUi?g=rmsk",children:"RepeatMasker intervals"}),"."]}),"\n",(0,l.jsx)(n.h4,{id:"segdups",children:"segdups"}),"\n",(0,l.jsxs)(n.p,{children:["Block-compressed bed file of ",(0,l.jsx)(n.a,{href:"https://genome.ucsc.edu/cgi-bin/hgTrackUi?g=genomicSuperDups",children:"segmental duplication intervals"}),"."]}),"\n",(0,l.jsx)(n.h4,{id:"seed_cutoffs",children:"seed_cutoffs"}),"\n",(0,l.jsx)(n.p,{children:"TSV of cutoff priors for genotyping."}),"\n",(0,l.jsx)(n.h4,{id:"single_sample_qc_definitions",children:"single_sample_qc_definitions"}),"\n",(0,l.jsx)(n.p,{children:"TSV of recommended ranges for single-sample QC metrics."}),"\n",(0,l.jsx)(n.h4,{id:"wgd_scoring_mask",children:"wgd_scoring_mask"}),"\n",(0,l.jsx)(n.p,{children:"Plain text bed file of whole-genome dosage (WGD) score intervals over which to assess coverage bias."}),"\n",(0,l.jsx)(n.h4,{id:"wham_include_list_bed_file",children:"wham_include_list_bed_file"}),"\n",(0,l.jsx)(n.p,{children:"Plain text bed file of intervals to call with Wham."}),"\n",(0,l.jsx)(n.h4,{id:"aou_recalibrate_gq_model_file",children:"aou_recalibrate_gq_model_file"}),"\n",(0,l.jsx)(n.p,{children:"Genotype filtering model trained using data generated by All of Us."}),"\n",(0,l.jsx)(n.h4,{id:"hgdp_recalibrate_gq_model_file",children:"hgdp_recalibrate_gq_model_file"}),"\n",(0,l.jsx)(n.p,{children:"Genotype filtering model trained using data generated from the Human Genome Diversity Project."}),"\n",(0,l.jsx)(n.h4,{id:"recalibrate_gq_genome_tracks",children:"recalibrate_gq_genome_tracks"}),"\n",(0,l.jsx)(n.p,{children:"List of block-compressed bed files, each containing intervals from a separate genome track."}),"\n",(0,l.jsx)(n.h3,{id:"benchmarking-datasets",children:"Benchmarking datasets"}),"\n",(0,l.jsx)(n.h4,{id:"ccdg_abel_site_level_benchmarking_dataset",children:"ccdg_abel_site_level_benchmarking_dataset"}),"\n",(0,l.jsxs)(n.p,{children:["Benchmarking variant set from ",(0,l.jsx)(n.a,{href:"https://doi.org/10.1038/s41586-020-2371-0",children:"Abel et al. 2020"}),"."]}),"\n",(0,l.jsx)(n.h4,{id:"gnomad_v2_collins_sample_level_benchmarking_dataset",children:"gnomad_v2_collins_sample_level_benchmarking_dataset"}),"\n",(0,l.jsxs)(n.p,{children:["Benchmarking genotypes from gnomAD-SV-v2, see ",(0,l.jsx)(n.a,{href:"https://doi.org/10.1038/s41586-020-2287-8",children:"Collins et al. 2020"}),". ",(0,l.jsx)(n.strong,{children:"Not public data."})]}),"\n",(0,l.jsx)(n.h4,{id:"gnomad_v2_collins_site_level_benchmarking_dataset",children:"gnomad_v2_collins_site_level_benchmarking_dataset"}),"\n",(0,l.jsxs)(n.p,{children:["Benchmarking variant set from gnomAD-SV-v2, see ",(0,l.jsx)(n.a,{href:"https://doi.org/10.1038/s41586-020-2287-8",children:"Collins et al. 2020"}),"."]}),"\n",(0,l.jsx)(n.h4,{id:"hgsv_byrska_bishop_sample_level_benchmarking_dataset",children:"hgsv_byrska_bishop_sample_level_benchmarking_dataset"}),"\n",(0,l.jsxs)(n.p,{children:["Benchmarking genotypes from ",(0,l.jsx)(n.a,{href:"https://doi.org/10.1016/j.cell.2022.08.004",children:"Byrska-Bishop et al. 2022"}),"."]}),"\n",(0,l.jsx)(n.h4,{id:"hgsv_byrska_bishop_sample_renaming_tsv",children:"hgsv_byrska_bishop_sample_renaming_tsv"}),"\n",(0,l.jsx)(n.p,{children:"Sample renaming manifest for the Byrska-Bishop benchmarking genotypes."}),"\n",(0,l.jsx)(n.h4,{id:"hgsv_byrska_bishop_site_level_benchmarking_dataset",children:"hgsv_byrska_bishop_site_level_benchmarking_dataset"}),"\n",(0,l.jsxs)(n.p,{children:["Benchmarking variant set from ",(0,l.jsx)(n.a,{href:"https://doi.org/10.1016/j.cell.2022.08.004",children:"Byrska-Bishop et al. 2022"}),"."]}),"\n",(0,l.jsx)(n.h4,{id:"hgsv_ebert_sample_level_benchmarking_dataset",children:"hgsv_ebert_sample_level_benchmarking_dataset"}),"\n",(0,l.jsxs)(n.p,{children:["Benchmarking genotypes from ",(0,l.jsx)(n.a,{href:"https://doi.org/10.1126/science.abf7117",children:"Ebert et al. 2021"}),"."]}),"\n",(0,l.jsx)(n.h4,{id:"ssc_belyeu_sample_level_benchmarking_dataset",children:"ssc_belyeu_sample_level_benchmarking_dataset"}),"\n",(0,l.jsxs)(n.p,{children:["Benchmarking genotypes from the ",(0,l.jsx)(n.a,{href:"https://www.sfari.org/resource/simons-simplex-collection/",children:"Simons Simplex Collection"}),", derived from ",(0,l.jsx)(n.a,{href:"https://doi.org/10.1016/j.ajhg.2021.02.012",children:"Belyeu et al. 2021"}),". ",(0,l.jsx)(n.strong,{children:"Not public data."})]}),"\n",(0,l.jsx)(n.h4,{id:"ssc_belyeu_site_level_benchmarking_dataset",children:"ssc_belyeu_site_level_benchmarking_dataset"}),"\n",(0,l.jsxs)(n.p,{children:["Benchmarking variant set from the ",(0,l.jsx)(n.a,{href:"https://www.sfari.org/resource/simons-simplex-collection/",children:"Simons Simplex Collection"}),", derived from ",(0,l.jsx)(n.a,{href:"https://doi.org/10.1016/j.ajhg.2021.02.012",children:"Belyeu et al. 2021"}),". ",(0,l.jsx)(n.strong,{children:"Not public data."})]}),"\n",(0,l.jsx)(n.h4,{id:"ssc_sanders_sample_level_benchmarking_dataset",children:"ssc_sanders_sample_level_benchmarking_dataset"}),"\n",(0,l.jsxs)(n.p,{children:["Benchmarking genotypes from the ",(0,l.jsx)(n.a,{href:"https://www.sfari.org/resource/simons-simplex-collection/",children:"Simons Simplex Collection"}),", derived from ",(0,l.jsx)(n.a,{href:"https://doi.org/10.1016/j.neuron.2015.09.016",children:"Sanders et al. 2015"}),". ",(0,l.jsx)(n.strong,{children:"Not public data."})]}),"\n",(0,l.jsx)(n.h4,{id:"thousand_genomes_site_level_benchmarking_dataset",children:"thousand_genomes_site_level_benchmarking_dataset"}),"\n",(0,l.jsxs)(n.p,{children:["Benchmarking variant set from the 1000 Genomes Project Phase 3 SV call set, see ",(0,l.jsx)(n.a,{href:"https://doi.org/10.1038/nature15394",children:"Sudmant et al. 2015"}),"."]}),"\n",(0,l.jsx)(n.h4,{id:"asc_site_level_benchmarking_dataset",children:"asc_site_level_benchmarking_dataset"}),"\n",(0,l.jsxs)(n.p,{children:["Benchmarking variant set from the Autism Spectrum Consortium. ",(0,l.jsx)(n.strong,{children:"Not public data."})]}),"\n",(0,l.jsx)(n.h4,{id:"hgsv_site_level_benchmarking_dataset",children:"hgsv_site_level_benchmarking_dataset"}),"\n",(0,l.jsxs)(n.p,{children:["Benchmarking variant set from ",(0,l.jsx)(n.a,{href:"https://doi.org/10.1038%2Fs41588-018-0107-y",children:"Werling et al. 2018"}),". ",(0,l.jsx)(n.strong,{children:"Not public data."})]}),"\n",(0,l.jsx)(n.h4,{id:"collins_2017_sample_level_benchmarking_dataset",children:"collins_2017_sample_level_benchmarking_dataset"}),"\n",(0,l.jsxs)(n.p,{children:["Benchmarking genotypes from ",(0,l.jsx)(n.a,{href:"https://doi.org/10.1186/s13059-017-1158-6",children:"Collins et al. 2017"}),". ",(0,l.jsx)(n.strong,{children:"Not public data."})]}),"\n",(0,l.jsx)(n.h4,{id:"sanders_2015_sample_level_benchmarking_dataset",children:"sanders_2015_sample_level_benchmarking_dataset"}),"\n",(0,l.jsxs)(n.p,{children:["Benchmarking genotypes from ",(0,l.jsx)(n.a,{href:"https://doi.org/10.1016/j.neuron.2015.09.016",children:"Sanders et al. 2015"}),". ",(0,l.jsx)(n.strong,{children:"Not public data."})]}),"\n",(0,l.jsx)(n.h4,{id:"werling_2018_sample_level_benchmarking_dataset",children:"werling_2018_sample_level_benchmarking_dataset"}),"\n",(0,l.jsxs)(n.p,{children:["Benchmarking genotypes from ",(0,l.jsx)(n.a,{href:"https://doi.org/10.1038%2Fs41588-018-0107-y",children:"Werling et al. 2018"}),". ",(0,l.jsx)(n.strong,{children:"Not public data."})]})]})}function _(e={}){const{wrapper:n}={...(0,r.R)(),...e.components};return n?(0,l.jsx)(n,{...e,children:(0,l.jsx)(o,{...e})}):o(e)}},8453:(e,n,i)=>{i.d(n,{R:()=>a,x:()=>t});var s=i(6540);const l={},r=s.createContext(l);function a(e){const n=s.useContext(r);return s.useMemo(function(){return"function"==typeof e?e(n):{...n,...e}},[n,e])}function t(e){let n;return n=e.disableParentContext?"function"==typeof e.components?e.components(l):e.components||l:a(e.components),s.createElement(r.Provider,{value:n},e.children)}}}]);