"use strict";(globalThis.webpackChunkGATK_SV=globalThis.webpackChunkGATK_SV||[]).push([[4802],{3181(e,s,n){n.r(s),n.d(s,{assets:()=>o,contentTitle:()=>c,default:()=>u,frontMatter:()=>r,metadata:()=>t,toc:()=>d});const t=JSON.parse('{"id":"modules/join_raw_calls","title":"JoinRawCalls","description":"Clusters unfiltered variants across batches","source":"@site/docs/modules/join_raw_calls.md","sourceDirName":"modules","slug":"/modules/jrc","permalink":"/gatk-sv/docs/modules/jrc","draft":false,"unlisted":false,"editUrl":"https://github.com/broadinstitute/gatk-sv/tree/master/website/docs/modules/join_raw_calls.md","tags":[],"version":"current","sidebarPosition":17,"frontMatter":{"title":"JoinRawCalls","description":"Clusters unfiltered variants across batches","sidebar_position":17,"slug":"jrc"},"sidebar":"tutorialSidebar","previous":{"title":"RefineComplexVariants","permalink":"/gatk-sv/docs/modules/refcv"},"next":{"title":"SVConcordance","permalink":"/gatk-sv/docs/modules/svc"}}');var l=n(4848),i=n(8453);const r={title:"JoinRawCalls",description:"Clusters unfiltered variants across batches",sidebar_position:17,slug:"jrc"},c=void 0,o={},d=[{value:"Inputs",id:"inputs",level:3},{value:"<code>prefix</code>",id:"prefix",level:4},{value:"<code>clustered_*_vcfs</code>",id:"clustered__vcfs",level:4},{value:"<code>ped_file</code>",id:"ped_file",level:4},{value:"Outputs",id:"outputs",level:3},{value:"<code>joined_raw_calls_vcf</code>",id:"joined_raw_calls_vcf",level:4},{value:"<code>ploidy_table</code>",id:"ploidy_table",level:4}];function a(e){const s={a:"a",code:"code",h3:"h3",h4:"h4",mermaid:"mermaid",p:"p",...(0,i.R)(),...e.components};return(0,l.jsxs)(l.Fragment,{children:[(0,l.jsx)(s.p,{children:(0,l.jsx)(s.a,{href:"https://github.com/broadinstitute/gatk-sv/blob/main/wdl/JoinRawCalls.wdl",children:"WDL source code"})}),"\n",(0,l.jsxs)(s.p,{children:["This module clusters raw unfiltered variants from ",(0,l.jsx)(s.a,{href:"./cb",children:"ClusterBatch"})," across all batches. Concordance between these\ngenotypes and the joint call set usually can be indicative of variant quality and is used downstream for genotype\nfiltering."]}),"\n",(0,l.jsx)(s.p,{children:"The following diagram illustrates the recommended invocation order:"}),"\n",(0,l.jsx)(s.mermaid,{value:"\nstateDiagram\n  direction LR\n    \n  classDef inModules stroke-width:0px,fill:#caf0f8,color:#00509d\n  classDef thisModule font-weight:bold,stroke-width:0px,fill:#ff9900,color:white\n  classDef outModules stroke-width:0px,fill:#caf0f8,color:#00509d\n\n  cb: ClusterBatch\n  jrc: JoinRawCalls\n  svc: SVConcordance\n  cb --\x3e jrc\n  jrc --\x3e svc\n  \n  class jrc thisModule\n  class cb inModules\n  class svc outModules"}),"\n",(0,l.jsx)(s.h3,{id:"inputs",children:"Inputs"}),"\n",(0,l.jsx)(s.h4,{id:"prefix",children:(0,l.jsx)(s.code,{children:"prefix"})}),"\n",(0,l.jsxs)(s.p,{children:["Prefix for the output VCF, such as the cohort name. The guidelines outlined in the\n",(0,l.jsx)(s.a,{href:"/docs/gs/inputs#sampleids",children:"sample ID requirements"})," section apply here."]}),"\n",(0,l.jsx)(s.h4,{id:"clustered__vcfs",children:(0,l.jsx)(s.code,{children:"clustered_*_vcfs"})}),"\n",(0,l.jsxs)(s.p,{children:["Array of clustered VCFs for all batches, generated in ",(0,l.jsx)(s.a,{href:"./cb#clustered__vcf",children:"ClusterBatch"}),"."]}),"\n",(0,l.jsx)(s.h4,{id:"ped_file",children:(0,l.jsx)(s.code,{children:"ped_file"})}),"\n",(0,l.jsxs)(s.p,{children:["Family structures and sex assignments determined in ",(0,l.jsx)(s.a,{href:"./eqc",children:"EvidenceQC"}),". See ",(0,l.jsx)(s.a,{href:"/docs/gs/inputs#ped-format",children:"PED file format"}),"."]}),"\n",(0,l.jsx)(s.h3,{id:"outputs",children:"Outputs"}),"\n",(0,l.jsx)(s.h4,{id:"joined_raw_calls_vcf",children:(0,l.jsx)(s.code,{children:"joined_raw_calls_vcf"})}),"\n",(0,l.jsx)(s.p,{children:"VCF containing all raw calls in the cohort."}),"\n",(0,l.jsx)(s.h4,{id:"ploidy_table",children:(0,l.jsx)(s.code,{children:"ploidy_table"})}),"\n",(0,l.jsxs)(s.p,{children:["TSV of contig ploidies for all samples, assuming diploid autosome and sex assignments from the ",(0,l.jsx)(s.a,{href:"#ped_file",children:"ped_file"}),"."]})]})}function u(e={}){const{wrapper:s}={...(0,i.R)(),...e.components};return s?(0,l.jsx)(s,{...e,children:(0,l.jsx)(a,{...e})}):a(e)}},8453(e,s,n){n.d(s,{R:()=>r,x:()=>c});var t=n(6540);const l={},i=t.createContext(l);function r(e){const s=t.useContext(i);return t.useMemo(function(){return"function"==typeof e?e(s):{...s,...e}},[s,e])}function c(e){let s;return s=e.disableParentContext?"function"==typeof e.components?e.components(l):e.components||l:r(e.components),t.createElement(i.Provider,{value:s},e.children)}}}]);