"use strict";(self.webpackChunkGATK_SV=self.webpackChunkGATK_SV||[]).push([[3416],{1944:(e,n,t)=>{t.d(n,{$:()=>i});var o=t(4848);const i=({children:e})=>(0,o.jsx)("span",{style:{backgroundColor:"var(--highlight-optional-arg-background-color)",borderRadius:"2px",color:"var(--highlight-optional-arg-text-color)",padding:"0.2rem"},children:e})},3583:(e,n,t)=>{t.r(n),t.d(n,{assets:()=>d,contentTitle:()=>a,default:()=>p,frontMatter:()=>r,metadata:()=>o,toc:()=>c});const o=JSON.parse('{"id":"modules/train_gcnv","title":"TrainGCNV","description":"Train gCNV","source":"@site/docs/modules/train_gcnv.md","sourceDirName":"modules","slug":"/modules/gcnv","permalink":"/gatk-sv/docs/modules/gcnv","draft":false,"unlisted":false,"editUrl":"https://github.com/broadinstitute/gatk-sv/tree/master/website/docs/modules/train_gcnv.md","tags":[],"version":"current","sidebarPosition":3,"frontMatter":{"title":"TrainGCNV","description":"Train gCNV","sidebar_position":3,"slug":"gcnv"},"sidebar":"tutorialSidebar","previous":{"title":"EvidenceQC","permalink":"/gatk-sv/docs/modules/eqc"},"next":{"title":"GatherBatchEvidence","permalink":"/gatk-sv/docs/modules/gbe"}}');var i=t(4848),s=t(8453),l=t(1944);const r={title:"TrainGCNV",description:"Train gCNV",sidebar_position:3,slug:"gcnv"},a=void 0,d={},c=[{value:"Inputs",id:"inputs",level:2},{value:"<code>samples</code>",id:"samples",level:4},{value:"<code>count_files</code>",id:"count_files",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>n_samples_subsample</code>, <code>sample_ids_training_subset</code>",id:"optional-n_samples_subsample-sample_ids_training_subset",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>outlier_sample_ids</code>",id:"optional-outlier_sample_ids",level:4},{value:"Outputs",id:"outputs",level:2},{value:"<code>cohort_contig_ploidy_model_tar</code>",id:"cohort_contig_ploidy_model_tar",level:4},{value:"<code>cohort_gcnv_model_tars</code>",id:"cohort_gcnv_model_tars",level:4},{value:"<code>cohort_contig_ploidy_calls_tar</code>",id:"cohort_contig_ploidy_calls_tar",level:4},{value:"<code>cohort_gcnv_calls_tars</code>",id:"cohort_gcnv_calls_tars",level:4},{value:"<code>cohort_genotyped_segments_vcfs</code>",id:"cohort_genotyped_segments_vcfs",level:4},{value:"<code>cohort_gcnv_tracking_tars</code>",id:"cohort_gcnv_tracking_tars",level:4},{value:"<code>cohort_genotyped_intervals_vcfs</code>",id:"cohort_genotyped_intervals_vcfs",level:4},{value:"<code>cohort_denoised_copy_ratios</code>",id:"cohort_denoised_copy_ratios",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>annotated_intervals</code>",id:"optional-annotated_intervals",level:4},{value:"<HighlightOptionalArg>Optional</HighlightOptionalArg> <code>filtered_intervals_cnv</code>, <code>filtered_intervals_ploidy</code>",id:"optional-filtered_intervals_cnv-filtered_intervals_ploidy",level:4}];function h(e){const n={a:"a",admonition:"admonition",code:"code",h2:"h2",h4:"h4",mermaid:"mermaid",p:"p",...(0,s.R)(),...e.components};return(0,i.jsxs)(i.Fragment,{children:[(0,i.jsx)(n.p,{children:(0,i.jsx)(n.a,{href:"https://github.com/broadinstitute/gatk-sv/blob/main/wdl/TrainGCNV.wdl",children:"WDL source code"})}),"\n",(0,i.jsxs)(n.p,{children:[(0,i.jsx)(n.a,{href:"https://www.nature.com/articles/s41588-023-01449-0",children:"GATK-gCNV"}),"\nis a method for detecting rare germline copy number variants (CNVs)\nfrom short-read sequencing read-depth information.\nThe ",(0,i.jsx)(n.code,{children:"TrainGCNV"})," module trains a gCNV model for use in the ",(0,i.jsx)(n.a,{href:"./gbe",children:"GatherBatchEvidence"})," workflow."]}),"\n",(0,i.jsx)(n.p,{children:"The samples used for training should be homogeneous (concerning sequencing platform,\ncoverage, library preparation, etc.) and similar to the samples on which the model will be applied."}),"\n",(0,i.jsxs)(n.p,{children:["For small, relatively homogeneous cohorts, a single gCNV model is usually sufficient.\nHowever, for larger cohorts, especially those with multiple data sources,\nwe recommend training a separate model for each batch or group of batches (see\n",(0,i.jsx)(n.a,{href:"/docs/execution/joint/#batching",children:"batching section"})," for details).\nThe model can be trained on all or a subset of the samples to which it will be applied.\nA subset of 100 randomly selected samples from the batch is a reasonable\ninput size for training the model; when the ",(0,i.jsx)(n.code,{children:"n_samples_subsample"})," input is provided,\nthe ",(0,i.jsx)(n.code,{children:"TrainGCNV"})," workflow can automatically perform this random selection."]}),"\n",(0,i.jsx)(n.p,{children:"The following diagram illustrates the recommended invocation order:"}),"\n",(0,i.jsx)(n.mermaid,{value:"\nstateDiagram\n  direction LR\n  \n  classDef inModules stroke-width:0px,fill:#caf0f8,color:#00509d\n  classDef thisModule font-weight:bold,stroke-width:0px,fill:#ff9900,color:white\n  classDef outModules stroke-width:0px,fill:#caf0f8,color:#00509d\n\n  batching: Batching, sample QC, and sex assignment\n  t: TrainGCNV\n  gbe: GatherBatchEvidence\n  \n  batching --\x3e t\n  t --\x3e gbe \n  \n  class t thisModule\n  class batching inModules\n  class gbe outModules"}),"\n",(0,i.jsx)(n.h2,{id:"inputs",children:"Inputs"}),"\n",(0,i.jsxs)(n.p,{children:["The majority of the optional inputs of the workflow map to the optional arguments of the\ntool the workflow uses, ",(0,i.jsx)(n.code,{children:"GATK-GermlineCNVCaller"}),"; hence, you may refer to the\n",(0,i.jsx)(n.a,{href:"https://gatk.broadinstitute.org/hc/en-us/articles/360040097712-GermlineCNVCaller",children:"documentation"}),"\nof the tool for a description on these optional inputs.  We recommend that most users use the defaults."]}),"\n",(0,i.jsx)(n.admonition,{type:"info",children:(0,i.jsxs)(n.p,{children:["All array inputs of sample data must match in order. For example, the order of the ",(0,i.jsx)(n.code,{children:"samples"})," array should match that\nof the ",(0,i.jsx)(n.code,{children:"count_files"})," array."]})}),"\n",(0,i.jsx)(n.h4,{id:"samples",children:(0,i.jsx)(n.code,{children:"samples"})}),"\n",(0,i.jsx)(n.p,{children:"Sample IDs"}),"\n",(0,i.jsx)(n.h4,{id:"count_files",children:(0,i.jsx)(n.code,{children:"count_files"})}),"\n",(0,i.jsxs)(n.p,{children:["Per-sample binned read counts (",(0,i.jsx)(n.code,{children:"*.rd.txt.gz"}),") generated in the ",(0,i.jsx)(n.a,{href:"./gse#outputs",children:"GatherSampleEvidence"})," workflow."]}),"\n",(0,i.jsxs)(n.h4,{id:"optional-n_samples_subsample-sample_ids_training_subset",children:[(0,i.jsx)(l.$,{children:"Optional"})," ",(0,i.jsx)(n.code,{children:"n_samples_subsample"}),", ",(0,i.jsx)(n.code,{children:"sample_ids_training_subset"})]}),"\n",(0,i.jsxs)(n.p,{children:["Provide one of these inputs to subset the input batch. ",(0,i.jsx)(n.code,{children:"n_samples_subsample"})," will randomly subset, while\n",(0,i.jsx)(n.code,{children:"sample_ids_training_subset"})," is for defining a predetermined subset. These options are provided for convenience in Terra."]}),"\n",(0,i.jsxs)(n.h4,{id:"optional-outlier_sample_ids",children:[(0,i.jsx)(l.$,{children:"Optional"})," ",(0,i.jsx)(n.code,{children:"outlier_sample_ids"})]}),"\n",(0,i.jsxs)(n.p,{children:["Provide a file containing sample IDs, delimited by new lines, to exclude from the from the gCNV model training\nprocess - this should can be based on outlier samples identified following ",(0,i.jsx)(n.a,{href:"./eqc",children:"EvidenceQC"})," that are still\nretained in the cohort."]}),"\n",(0,i.jsx)(n.h2,{id:"outputs",children:"Outputs"}),"\n",(0,i.jsx)(n.h4,{id:"cohort_contig_ploidy_model_tar",children:(0,i.jsx)(n.code,{children:"cohort_contig_ploidy_model_tar"})}),"\n",(0,i.jsx)(n.p,{children:"Contig ploidy model tarball."}),"\n",(0,i.jsx)(n.h4,{id:"cohort_gcnv_model_tars",children:(0,i.jsx)(n.code,{children:"cohort_gcnv_model_tars"})}),"\n",(0,i.jsx)(n.p,{children:"CNV model tarballs scattered across genomic intervals."}),"\n",(0,i.jsx)(n.h4,{id:"cohort_contig_ploidy_calls_tar",children:(0,i.jsx)(n.code,{children:"cohort_contig_ploidy_calls_tar"})}),"\n",(0,i.jsx)(n.p,{children:"Contig ploidy calls for the submitted batch."}),"\n",(0,i.jsx)(n.h4,{id:"cohort_gcnv_calls_tars",children:(0,i.jsx)(n.code,{children:"cohort_gcnv_calls_tars"})}),"\n",(0,i.jsx)(n.p,{children:"CNV call tarballs scattered by sample and genomic region prior to segmentation."}),"\n",(0,i.jsx)(n.h4,{id:"cohort_genotyped_segments_vcfs",children:(0,i.jsx)(n.code,{children:"cohort_genotyped_segments_vcfs"})}),"\n",(0,i.jsx)(n.p,{children:"Single-sample VCFs of CNV calls for the submitted batch."}),"\n",(0,i.jsx)(n.h4,{id:"cohort_gcnv_tracking_tars",children:(0,i.jsx)(n.code,{children:"cohort_gcnv_tracking_tars"})}),"\n",(0,i.jsx)(n.p,{children:"Convergence tracking logs."}),"\n",(0,i.jsx)(n.h4,{id:"cohort_genotyped_intervals_vcfs",children:(0,i.jsx)(n.code,{children:"cohort_genotyped_intervals_vcfs"})}),"\n",(0,i.jsx)(n.p,{children:"Single-sample VCFs for the submitted batch containing per-interval genotypes prior to segmentation."}),"\n",(0,i.jsx)(n.h4,{id:"cohort_denoised_copy_ratios",children:(0,i.jsx)(n.code,{children:"cohort_denoised_copy_ratios"})}),"\n",(0,i.jsx)(n.p,{children:"TSV files containing denoised copy ratios in each sample."}),"\n",(0,i.jsxs)(n.h4,{id:"optional-annotated_intervals",children:[(0,i.jsx)(l.$,{children:"Optional"})," ",(0,i.jsx)(n.code,{children:"annotated_intervals"})]}),"\n",(0,i.jsxs)(n.p,{children:["The count files from ",(0,i.jsx)(n.a,{href:"./gse",children:"GatherSampleEvidence"})," with adjacent intervals combined into\nlocus-sorted ",(0,i.jsx)(n.code,{children:"DepthEvidence"})," files using ",(0,i.jsx)(n.code,{children:"GATK CondenseDepthEvidence"})," tool, which are\nannotated with GC content, mappability, and segmental-duplication content using\n",(0,i.jsx)(n.a,{href:"https://gatk.broadinstitute.org/hc/en-us/articles/360041416652-AnnotateIntervals",children:(0,i.jsx)(n.code,{children:"GATK-AnnotateIntervals"})}),"\ntool. This output is generated if ",(0,i.jsx)(n.code,{children:"do_explicit_gc_correction"})," is set to ",(0,i.jsx)(n.code,{children:"True"}),". Disabled by default."]}),"\n",(0,i.jsxs)(n.h4,{id:"optional-filtered_intervals_cnv-filtered_intervals_ploidy",children:[(0,i.jsx)(l.$,{children:"Optional"})," ",(0,i.jsx)(n.code,{children:"filtered_intervals_cnv"}),", ",(0,i.jsx)(n.code,{children:"filtered_intervals_ploidy"})]}),"\n",(0,i.jsxs)(n.p,{children:["Intervals of read count bins to be used for CNV and ploidy calling after filtering for problematic regions (e.g.\nhigh GC content). This output is generated if ",(0,i.jsx)(n.code,{children:"filter_intervals"})," is set to ",(0,i.jsx)(n.code,{children:"True"}),". Enabled by default."]})]})}function p(e={}){const{wrapper:n}={...(0,s.R)(),...e.components};return n?(0,i.jsx)(n,{...e,children:(0,i.jsx)(h,{...e})}):h(e)}},8453:(e,n,t)=>{t.d(n,{R:()=>l,x:()=>r});var o=t(6540);const i={},s=o.createContext(i);function l(e){const n=o.useContext(s);return o.useMemo(function(){return"function"==typeof e?e(n):{...n,...e}},[n,e])}function r(e){let n;return n=e.disableParentContext?"function"==typeof e.components?e.components(i):e.components||i:l(e.components),o.createElement(s.Provider,{value:n},e.children)}}}]);