<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper plugin-docs plugin-id-default docs-version-current docs-doc-page docs-doc-id-execution/single" data-has-hydrated="false">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v3.8.0">
<title data-rh="true">Single-sample | GATK-SV</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/gatk-sv/docs/execution/single"><meta data-rh="true" property="og:locale" content="en"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Single-sample | GATK-SV"><meta data-rh="true" name="description" content="Run the pipeline on a single sample"><meta data-rh="true" property="og:description" content="Run the pipeline on a single sample"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/gatk-sv/docs/execution/single"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/gatk-sv/docs/execution/single" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/gatk-sv/docs/execution/single" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://FPHFCFLYVM-dsn.algolia.net" crossorigin="anonymous"><script data-rh="true" type="application/ld+json">{"@context":"https://schema.org","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"name":"Execution","item":"https://broadinstitute.github.io/gatk-sv/docs/category/execution"},{"@type":"ListItem","position":2,"name":"Single-sample","item":"https://broadinstitute.github.io/gatk-sv/docs/execution/single"}]}</script><link rel="search" type="application/opensearchdescription+xml" title="GATK-SV" href="/gatk-sv/opensearch.xml"><link rel="stylesheet" href="/gatk-sv/assets/css/styles.04c1d369.css">
<script src="/gatk-sv/assets/js/runtime~main.863a676d.js" defer="defer"></script>
<script src="/gatk-sv/assets/js/main.33bd68b8.js" defer="defer"></script>
</head>
<body class="navigation-with-keyboard">
<svg xmlns="http://www.w3.org/2000/svg" style="display: none;"><defs>
<symbol id="theme-svg-external-link" viewBox="0 0 24 24"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"/></symbol>
</defs></svg>
<script>!function(){var t="light";var e=function(){try{return new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}}()||function(){try{return window.localStorage.getItem("theme")}catch(t){}}();document.documentElement.setAttribute("data-theme",e||t),document.documentElement.setAttribute("data-theme-choice",e||t)}(),function(){try{const c=new URLSearchParams(window.location.search).entries();for(var[t,e]of c)if(t.startsWith("docusaurus-data-")){var a=t.replace("docusaurus-data-","data-");document.documentElement.setAttribute(a,e)}}catch(t){}}()</script><div id="__docusaurus"><div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="theme-layout-navbar navbar navbar--fixed-top"><div class="navbar__inner"><div class="theme-layout-navbar-left navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/gatk-sv/"><b class="navbar__title text--truncate">GATK-SV</b></a></div><div class="theme-layout-navbar-right navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/gatk-sv/docs/intro">Documentation</a><a href="https://github.com/broadinstitute/gatk-sv/issues" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">Questions<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a><a href="https://github.com/broadinstitute/gatk-sv" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="system mode" aria-label="Switch between dark and light mode (currently system mode)"><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP systemToggleIcon_QzmC"><path fill="currentColor" d="m12 21c4.971 0 9-4.029 9-9s-4.029-9-9-9-9 4.029-9 9 4.029 9 9 9zm4.95-13.95c1.313 1.313 2.05 3.093 2.05 4.95s-0.738 3.637-2.05 4.95c-1.313 1.313-3.093 2.05-4.95 2.05v-14c1.857 0 3.637 0.737 4.95 2.05z"></path></svg></button></div><div class="navbarSearchContainer_Bca1"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search (Command+K)"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20" aria-hidden="true"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="theme-layout-main main-wrapper mainWrapper_z2l0"><div class="docsWrapper_hBAB"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docRoot_UBD9"><aside class="theme-doc-sidebar-container docSidebarContainer_YfHR"><div class="sidebarViewport_aRkj"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/intro">About</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/getting-started">Getting Started</a><button aria-label="Expand sidebar category &#x27;Getting Started&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/best_practices">Best Practices Guide</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" href="/gatk-sv/docs/category/execution">Execution</a><button aria-label="Collapse sidebar category &#x27;Execution&#x27;" aria-expanded="true" type="button" class="clean-btn menu__caret"></button></div><ul class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/execution/overview">Overview</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/gatk-sv/docs/execution/single">Single-sample</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/execution/joint">Joint calling</a></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/modules">Modules</a><button aria-label="Expand sidebar category &#x27;Modules&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/runtime_attr">Runtime attributes</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/resources">Resource files</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/troubleshooting">Troubleshooting</a><button aria-label="Expand sidebar category &#x27;Troubleshooting&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/advanced-guides">Advanced Guides</a><button aria-label="Expand sidebar category &#x27;Advanced Guides&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/acknowledgements">Acknowledgements</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/references">Citation and references</a></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_TBSr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/gatk-sv/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><a class="breadcrumbs__link" href="/gatk-sv/docs/category/execution"><span>Execution</span></a></li><li class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link">Single-sample</span></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><header><h1>Single-sample</h1></header><h2 class="anchor anchorWithStickyNavbar_LWe7" id="introduction">Introduction<a href="#introduction" class="hash-link" aria-label="Direct link to Introduction" title="Direct link to Introduction">​</a></h2>
<p><strong>Extending SV detection to small datasets</strong></p>
<p>The Single Sample pipeline is designed to facilitate running the methods developed for the cohort-mode GATK-SV pipeline on small data sets or in
clinical contexts where batching large numbers of samples is not an option. To do so, it uses precomputed data, SV calls,
and model parameters computed by the cohort pipeline on a reference panel composed of similar samples. The pipeline integrates this
precomputed information with signals extracted from the input CRAM file to produce a call set similar in quality to results
computed by the cohort pipeline in a computationally tractable and reproducible manner.</p>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="terra-workspace">Terra workspace<a href="#terra-workspace" class="hash-link" aria-label="Direct link to Terra workspace" title="Direct link to Terra workspace">​</a></h2>
<p>Users should clone the <a href="https://app.terra.bio/#workspaces/help-gatk/GATK-Structural-Variants-Single-Sample" target="_blank" rel="noopener noreferrer">Terra single-sample workspace</a>
which is configured with a demo sample to run on.
Refer to the following sections for instructions on how to run the pipeline on your data using this workspace.</p>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="data">Data<a href="#data" class="hash-link" aria-label="Direct link to Data" title="Direct link to Data">​</a></h2>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="case-sample">Case sample<a href="#case-sample" class="hash-link" aria-label="Direct link to Case sample" title="Direct link to Case sample">​</a></h3>
<p>The default workspace includes a NA12878 input WGS CRAM file configured in the workspace samples data table. This file is part of the
high coverage (30X) <a href="https://www.internationalgenome.org/data-portal/data-collection/30x-grch38" target="_blank" rel="noopener noreferrer">WGS data</a> for the 1000
Genomes Project samples generated by the New York Genome Center and hosted in
<a href="https://app.terra.bio/#workspaces/anvil-datastorage/1000G-high-coverage-2019" target="_blank" rel="noopener noreferrer">AnVIL</a>.</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="reference-panel">Reference panel<a href="#reference-panel" class="hash-link" aria-label="Direct link to Reference panel" title="Direct link to Reference panel">​</a></h3>
<p>The reference panel configured in the workspace consists of data and calls computed from 156 publicly available samples
chosen from the NYGC/AnVIL 1000 Genomes high coverage data linked above.</p>
<p>Inputs to the pipeline for the reference panel include:</p>
<ul>
<li>A precomputed SV callset VCF, and joint-called depth-based CNV call files</li>
<li>Raw calls for the reference panel samples from Manta and WHAM</li>
<li>Trained models for calling copy number variation in GATK gCNV case mode</li>
<li>Parameters learned by the cohort mode pipeline in training machine learning models on the reference panel samples.</li>
</ul>
<p>These resources are primarily configured in the &quot;Workspace Data&quot; for this workspace. However, several of the resources need
to be passed  to the workflow as large lists of files or strings. Due to Terra limitations on uploading data containing lists to the
workspace data table, these resources are specified directly in the workflow configuration.</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="reference-resources">Reference resources<a href="#reference-resources" class="hash-link" aria-label="Direct link to Reference resources" title="Direct link to Reference resources">​</a></h3>
<p>The pipeline uses a number of resource and data files computed for the hg38 reference:</p>
<ul>
<li>Reference sequences and indices</li>
<li>Genome annotation tracks such as segmental duplication and RepeatMasker tracks</li>
<li>Data used for annotation of called variants, including GenCode gene annotations and gnomAD site allele frequencies.</li>
</ul>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="single-sample-workflow">Single sample workflow<a href="#single-sample-workflow" class="hash-link" aria-label="Direct link to Single sample workflow" title="Direct link to Single sample workflow">​</a></h2>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="what-does-it-do">What does it do?<a href="#what-does-it-do" class="hash-link" aria-label="Direct link to What does it do?" title="Direct link to What does it do?">​</a></h3>
<p>The workflow <code>gatk-sv-single-sample</code> calls structural variations on a single input CRAM by running the GATK-SV Single Sample Pipeline end-to-end.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="what-does-it-require-as-input">What does it require as input?<a href="#what-does-it-require-as-input" class="hash-link" aria-label="Direct link to What does it require as input?" title="Direct link to What does it require as input?">​</a></h4>
<p>The workflow accepts a single CRAM or BAM file as input, configured in the following parameters:</p>
<table><thead><tr><th>Input Type</th><th>Input Name</th><th>Description</th></tr></thead><tbody><tr><td><code>String</code></td><td><code>sample_id</code></td><td>Case sample identifier</td></tr><tr><td><code>File</code></td><td><code>bam_or_cram_file</code></td><td>Path to the GCS location of the input CRAM or BAM file</td></tr><tr><td><code>String</code></td><td><code>batch</code></td><td>Arbitrary name to be assigned to the run</td></tr></tbody></table>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="additional-workspace-level-inputs">Additional workspace-level inputs<a href="#additional-workspace-level-inputs" class="hash-link" aria-label="Direct link to Additional workspace-level inputs" title="Direct link to Additional workspace-level inputs">​</a></h4>
<ul>
<li>Reference resources for hg38</li>
<li>Input data for the reference panel</li>
<li>The set of docker images used in the pipeline.</li>
</ul>
<p>Please contact GATK-SV developers if you are interested in customizing these
inputs beyond their defaults.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="what-does-it-return-as-output">What does it return as output?<a href="#what-does-it-return-as-output" class="hash-link" aria-label="Direct link to What does it return as output?" title="Direct link to What does it return as output?">​</a></h4>
<table><thead><tr><th>Output Type</th><th>Output Name</th><th>Description</th></tr></thead><tbody><tr><td><code>File</code></td><td><code>final_vcf</code></td><td>SV VCF output for the pipeline. Includes all sites genotyped as variant in the case sample and genotypes for the reference panel. Sites are annotated with overlap of functional genome elements and allele frequencies of matching variants in gnomAD</td></tr><tr><td><code>File</code></td><td><code>final_vcf_idx</code></td><td>Index file for <code>final_vcf</code></td></tr><tr><td><code>File</code></td><td><code>final_bed</code></td><td>Final output in BED format. Filter status, list of variant samples, and all VCF INFO fields are reported as additional columns.</td></tr><tr><td><code>File</code></td><td><code>metrics_file</code></td><td>Metrics computed from the input data and intermediate and final VCFs. Includes metrics on the SV evidence, and on the number of variants called, broken down by type and size range.</td></tr><tr><td><code>File</code></td><td><code>qc_file</code></td><td>Quality-control check file. This extracts several key metrics from the <code>metrics_file</code> and compares them to pre-specified threshold values. If any QC checks evaluate to FAIL, further diagnostics may be required.</td></tr><tr><td><code>File</code></td><td><code>ploidy_matrix</code></td><td>Matrix of contig ploidy estimates computed by GATK gCNV.</td></tr><tr><td><code>File</code></td><td><code>ploidy_plots</code></td><td>Plots of contig ploidy generated from <code>ploidy_matrix</code></td></tr><tr><td><code>File</code></td><td><code>non_genotyped_unique_depth_calls</code></td><td>This VCF file contains any depth based calls made in the case sample that did not pass genotyping checks and do not match a depth-based call from the reference panel. If very high sensitivity is desired, examine this file for additional large CNV calls.</td></tr><tr><td><code>File</code></td><td><code>non_genotyped_unique_depth_calls_idx</code></td><td>Index file for <code>non_genotyped_unique_depth_calls</code></td></tr><tr><td><code>File</code></td><td><code>pre_cleanup_vcf</code></td><td>VCF output in a representation used internally in the pipeline. This file is less compliant with the VCF spec and is intended for debugging purposes.</td></tr><tr><td><code>File</code></td><td><code>pre_cleanup_vcf_idx</code></td><td>Index file for <code>pre_cleanup_vcf</code></td></tr></tbody></table>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="example-time-and-cost-run-on-sample-data">Example time and cost run on sample data<a href="#example-time-and-cost-run-on-sample-data" class="hash-link" aria-label="Direct link to Example time and cost run on sample data" title="Direct link to Example time and cost run on sample data">​</a></h4>
<table><thead><tr><th>Sample Name</th><th>Sample Size</th><th>Time</th><th>Cost $</th></tr></thead><tbody><tr><td>NA12878</td><td>18.17 GiB</td><td>23hrs</td><td>~$7.34</td></tr></tbody></table>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="to-use-this-workflow-on-your-own-data">To use this workflow on your own data<a href="#to-use-this-workflow-on-your-own-data" class="hash-link" aria-label="Direct link to To use this workflow on your own data" title="Direct link to To use this workflow on your own data">​</a></h4>
<p>If you would like to run this workflow on your own samples (which must be medium-to-high coverage WGS data):</p>
<ul>
<li>Clone the <a href="https://app.terra.bio/#workspaces/help-gatk/GATK-Structural-Variants-Single-Sample" target="_blank" rel="noopener noreferrer">workspace</a> into a Terra project you have access to.
Select <code>us-central1</code> for the region. If you must use a different region, you will need to copy all GATK-SV docker images to the other region
before running the pipeline. See the <a href="/gatk-sv/docs/gs/dockers#regions-important">docker images section</a> for details.</li>
<li>In the cloned workspace, upload rows to the Sample and (optionally) the Participant Data Table that describe your samples.
Ensure that the rows you add to the Sample table contain the columns <code>sample_id</code> and <code>bam_or_cram_file</code> are populated appropriately.</li>
<li>There is no need to modify values in the workspace data or method configuration. If you are interested in modifying the reference
genome resources or reference panel, please contact the GATK team for support as listed below.</li>
<li>Launch the workflow from the &quot;Workflows&quot; tab, selecting your samples as the inputs.</li>
</ul>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="quality-control">Quality control<a href="#quality-control" class="hash-link" aria-label="Direct link to Quality control" title="Direct link to Quality control">​</a></h4>
<p>Please check the <code>qc_file</code> output to screen for data quality issues. This file provides a table of quality metrics and
suggested acceptable ranges.</p>
<div class="theme-admonition theme-admonition-warning admonition_xJq3 alert alert--warning"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 16 16"><path fill-rule="evenodd" d="M8.893 1.5c-.183-.31-.52-.5-.887-.5s-.703.19-.886.5L.138 13.499a.98.98 0 0 0 0 1.001c.193.31.53.501.886.501h13.964c.367 0 .704-.19.877-.5a1.03 1.03 0 0 0 .01-1.002L8.893 1.5zm.133 11.497H6.987v-2.003h2.039v2.003zm0-3.004H6.987V5.987h2.039v4.006z"></path></svg></span>warning</div><div class="admonitionContent_BuS1"><p>If a metric exceeds the recommended range, all variants will be automatically flagged with
a non-passing <code>FILTER</code> status in the output VCF.</p></div></div></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="row margin-top--sm theme-doc-footer-edit-meta-row"><div class="col"><a href="https://github.com/broadinstitute/gatk-sv/tree/master/website/docs/execution/single.md" target="_blank" rel="noopener noreferrer" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_JAkA"></div></div></footer></article><nav class="docusaurus-mt-lg pagination-nav" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/gatk-sv/docs/execution/overview"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Overview</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/gatk-sv/docs/execution/joint"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Joint calling</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#introduction" class="table-of-contents__link toc-highlight">Introduction</a></li><li><a href="#terra-workspace" class="table-of-contents__link toc-highlight">Terra workspace</a></li><li><a href="#data" class="table-of-contents__link toc-highlight">Data</a><ul><li><a href="#case-sample" class="table-of-contents__link toc-highlight">Case sample</a></li><li><a href="#reference-panel" class="table-of-contents__link toc-highlight">Reference panel</a></li><li><a href="#reference-resources" class="table-of-contents__link toc-highlight">Reference resources</a></li></ul></li><li><a href="#single-sample-workflow" class="table-of-contents__link toc-highlight">Single sample workflow</a><ul><li><a href="#what-does-it-do" class="table-of-contents__link toc-highlight">What does it do?</a></li></ul></li></ul></div></div></div></div></main></div></div></div><footer class="theme-layout-footer footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="theme-layout-footer-column col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/gatk-sv/docs/intro">About</a></li></ul></div><div class="theme-layout-footer-column col footer__col"><div class="footer__title">Community</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/broadinstitute/gatk-sv/discussions" target="_blank" rel="noopener noreferrer" class="footer__link-item">Github<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a></li></ul></div><div class="theme-layout-footer-column col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://talkowski.mgh.harvard.edu" target="_blank" rel="noopener noreferrer" class="footer__link-item">Talkowski lab</a></li><li class="footer__item"><a href="https://www.broadinstitute.org" target="_blank" rel="noopener noreferrer" class="footer__link-item">Broad Institute</a></li><li class="footer__item"><a href="https://www.massgeneral.org/research/cgm" target="_blank" rel="noopener noreferrer" class="footer__link-item">Center for Genomic Medicine</a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2025 BroadInstitute, Built with Docusaurus.</div></div></div></footer></div>
</body>
</html>