<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper plugin-docs plugin-id-default docs-version-current docs-doc-page docs-doc-id-execution/joint" data-has-hydrated="false">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v3.8.1">
<title data-rh="true">Joint calling | GATK-SV</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/gatk-sv/docs/execution/joint"><meta data-rh="true" property="og:locale" content="en"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Joint calling | GATK-SV"><meta data-rh="true" name="description" content="Run the pipeline on a cohort"><meta data-rh="true" property="og:description" content="Run the pipeline on a cohort"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/gatk-sv/docs/execution/joint"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/gatk-sv/docs/execution/joint" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/gatk-sv/docs/execution/joint" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://FPHFCFLYVM-dsn.algolia.net" crossorigin="anonymous"><script data-rh="true" type="application/ld+json">{"@context":"https://schema.org","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"name":"Execution","item":"https://broadinstitute.github.io/gatk-sv/docs/category/execution"},{"@type":"ListItem","position":2,"name":"Joint calling","item":"https://broadinstitute.github.io/gatk-sv/docs/execution/joint"}]}</script><link rel="preconnect" href="https://www.google-analytics.com">
<link rel="preconnect" href="https://www.googletagmanager.com">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-9DQEYKHD1M"></script>
<script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-9DQEYKHD1M",{anonymize_ip:!0})</script>



<link rel="search" type="application/opensearchdescription+xml" title="GATK-SV" href="/gatk-sv/opensearch.xml"><link rel="stylesheet" href="/gatk-sv/assets/css/styles.82e231fb.css">
<script src="/gatk-sv/assets/js/runtime~main.0333ddf4.js" defer="defer"></script>
<script src="/gatk-sv/assets/js/main.1ed6e5ce.js" defer="defer"></script>
</head>
<body class="navigation-with-keyboard">
<svg xmlns="http://www.w3.org/2000/svg" style="display: none;"><defs>
<symbol id="theme-svg-external-link" viewBox="0 0 24 24"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"/></symbol>
</defs></svg>
<script>!function(){var t="light";var e=function(){try{return new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}}()||function(){try{return window.localStorage.getItem("theme")}catch(t){}}();document.documentElement.setAttribute("data-theme",e||t),document.documentElement.setAttribute("data-theme-choice",e||t)}(),function(){try{const c=new URLSearchParams(window.location.search).entries();for(var[t,e]of c)if(t.startsWith("docusaurus-data-")){var a=t.replace("docusaurus-data-","data-");document.documentElement.setAttribute(a,e)}}catch(t){}}()</script><div id="__docusaurus"><div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="theme-layout-navbar navbar navbar--fixed-top"><div class="navbar__inner"><div class="theme-layout-navbar-left navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/gatk-sv/"><b class="navbar__title text--truncate">GATK-SV</b></a></div><div class="theme-layout-navbar-right navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/gatk-sv/docs/intro">Documentation</a><a href="https://github.com/broadinstitute/gatk-sv/issues" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">Questions<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a><a href="https://github.com/broadinstitute/gatk-sv" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="system mode" aria-label="Switch between dark and light mode (currently system mode)"><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP systemToggleIcon_QzmC"><path fill="currentColor" d="m12 21c4.971 0 9-4.029 9-9s-4.029-9-9-9-9 4.029-9 9 4.029 9 9 9zm4.95-13.95c1.313 1.313 2.05 3.093 2.05 4.95s-0.738 3.637-2.05 4.95c-1.313 1.313-3.093 2.05-4.95 2.05v-14c1.857 0 3.637 0.737 4.95 2.05z"></path></svg></button></div><div class="navbarSearchContainer_Bca1"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search (Command+K)"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20" aria-hidden="true"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="theme-layout-main main-wrapper mainWrapper_z2l0"><div class="docsWrapper_hBAB"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docRoot_UBD9"><aside class="theme-doc-sidebar-container docSidebarContainer_YfHR"><div class="sidebarViewport_aRkj"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/intro">About</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/getting-started">Getting Started</a><button aria-label="Expand sidebar category &#x27;Getting Started&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/best_practices">Best Practices Guide</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" href="/gatk-sv/docs/category/execution">Execution</a><button aria-label="Collapse sidebar category &#x27;Execution&#x27;" aria-expanded="true" type="button" class="clean-btn menu__caret"></button></div><ul class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/execution/overview">Overview</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/execution/single">Single-sample</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/gatk-sv/docs/execution/joint">Joint calling</a></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/modules">Modules</a><button aria-label="Expand sidebar category &#x27;Modules&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/runtime_attr">Runtime attributes</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/resources">Resource files</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/troubleshooting">Troubleshooting</a><button aria-label="Expand sidebar category &#x27;Troubleshooting&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/advanced-guides">Advanced Guides</a><button aria-label="Expand sidebar category &#x27;Advanced Guides&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/acknowledgements">Acknowledgements</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/references">Citation and references</a></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_TBSr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/gatk-sv/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><a class="breadcrumbs__link" href="/gatk-sv/docs/category/execution"><span>Execution</span></a></li><li class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link">Joint calling</span></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><header><h1>Joint calling</h1></header><h2 class="anchor anchorWithStickyNavbar_LWe7" id="terra-workspace">Terra workspace<a href="#terra-workspace" class="hash-link" aria-label="Direct link to Terra workspace" title="Direct link to Terra workspace">​</a></h2>
<p>Users should clone the <a href="https://app.terra.bio/#workspaces/broad-firecloud-dsde-methods/GATK-Structural-Variants-Joint-Calling" target="_blank" rel="noopener noreferrer">Terra joint calling workspace</a>
which is configured with a demo sample set.
Refer to the following sections for instructions on how to run the pipeline on your data using this workspace.</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="default-data">Default data<a href="#default-data" class="hash-link" aria-label="Direct link to Default data" title="Direct link to Default data">​</a></h3>
<p>The demonstration data in this workspace is 156 publicly-available 1000 Genomes Project samples from the
<a href="https://app.terra.bio/#workspaces/anvil-datastorage/1000G-high-coverage-2019" target="_blank" rel="noopener noreferrer">NYGC/AnVIL high coverage data set</a>.</p>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="pipeline-expectations">Pipeline Expectations<a href="#pipeline-expectations" class="hash-link" aria-label="Direct link to Pipeline Expectations" title="Direct link to Pipeline Expectations">​</a></h2>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="what-does-it-do">What does it do?<a href="#what-does-it-do" class="hash-link" aria-label="Direct link to What does it do?" title="Direct link to What does it do?">​</a></h3>
<p>This pipeline performs structural variation discovery from CRAMs, joint genotyping, and variant resolution on a cohort
of samples.</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="required-inputs">Required inputs<a href="#required-inputs" class="hash-link" aria-label="Direct link to Required inputs" title="Direct link to Required inputs">​</a></h3>
<p>Refer to the <a href="/gatk-sv/docs/gs/inputs">Input Data section</a> for details on file formats, sample QC, and sample ID restrictions.</p>
<p>The following inputs must be provided for each sample in the cohort, via the sample table described in <strong>Workspace
Setup</strong> step 2:</p>
<table><thead><tr><th>Input Type</th><th>Input Name</th><th>Description</th></tr></thead><tbody><tr><td><code>String</code></td><td><code>sample_id</code></td><td>Case sample identifier</td></tr><tr><td><code>File</code></td><td><code>bam_or_cram_file</code></td><td>Path to the GCS location of the input CRAM or BAM file.</td></tr></tbody></table>
<p>The following cohort-level or batch-level inputs are also required:</p>
<table><thead><tr><th>Input Type</th><th>Input Name</th><th>Description</th></tr></thead><tbody><tr><td><code>String</code></td><td><code>sample_set_id</code></td><td>Batch identifier</td></tr><tr><td><code>String</code></td><td><code>sample_set_set_id</code></td><td>Cohort identifier</td></tr><tr><td><code>File</code></td><td><code>cohort_ped_file</code></td><td>Path to the GCS location of a family structure definitions file in <a href="/gatk-sv/docs/gs/inputs#ped-format">PED format</a>.</td></tr></tbody></table>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="pipeline-outputs">Pipeline outputs<a href="#pipeline-outputs" class="hash-link" aria-label="Direct link to Pipeline outputs" title="Direct link to Pipeline outputs">​</a></h3>
<p>The following are the main pipeline outputs. For more information on the outputs of each module, refer to the
<a href="/gatk-sv/docs/category/modules">Modules section</a>.</p>
<table><thead><tr><th>Output Type</th><th>Output Name</th><th>Description</th></tr></thead><tbody><tr><td><code>File</code></td><td><code>annotated_vcf</code></td><td>Annotated SV VCF for the cohort***</td></tr><tr><td><code>File</code></td><td><code>annotated_vcf_idx</code></td><td>Index for <code>annotated_vcf</code></td></tr><tr><td><code>File</code></td><td><code>sv_vcf_qc_output</code></td><td>QC plots (bundled in a .tar.gz file)</td></tr></tbody></table>
<p>***Note that this VCF is not filtered</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="pipeline-overview">Pipeline overview<a href="#pipeline-overview" class="hash-link" aria-label="Direct link to Pipeline overview" title="Direct link to Pipeline overview">​</a></h3>
<p><img decoding="async" loading="lazy" src="https://media.githubusercontent.com/media/broadinstitute/gatk-sv/refs/tags/v1.0/terra_pipeline_diagram.jpg" alt="Pipeline Diagram" class="img_ev3q"></p>
<p>The following workflows and Jupyter notebooks are included in this workspace, to be executed in this order:</p>
<ol>
<li><code>01-GatherSampleEvidence</code>: Per-sample SV evidence collection, including calls from a configurable set of
algorithms (Manta, Scramble, and Wham), read depth (RD), split read positions (SR), and discordant pair positions (PE).</li>
<li><code>02-EvidenceQC</code>: Dosage bias scoring and ploidy estimation, run on preliminary batches</li>
<li>[Notebook] <code>SampleQC.ipynb</code>: Interactively perform sample QC and filtering using outputs from <code>02-EvidenceQC</code></li>
<li>[Notebook] <code>Batching.ipynb</code>: Create batches for subsequent steps. For cohorts &gt;500 samples or smaller heterogeneous cohorts</li>
<li><code>03-TrainGCNV</code>: Per-batch training of a gCNV model for use in <code>04-GatherBatchEvidence</code></li>
<li><code>04-GatherBatchEvidence</code>: Per-batch copy number variant calling using cn.MOPS and GATK gCNV; B-allele frequency (BAF)
generation; call and evidence aggregation</li>
<li><code>05-ClusterBatch</code>: Per-batch variant clustering</li>
<li><code>06-GenerateBatchMetrics</code>: Per-batch metric generation</li>
<li><code>07-FilterBatchSites</code>: Per-batch variant filtering and plot SV counts per sample per SV type to enable choice of IQR
cutoff for outlier filtration in <code>08-FilterBatchSamples</code></li>
<li><code>08-FilterBatchSamples</code>: Per-batch outlier sample filtration</li>
<li><code>09-MergeBatchSites</code>: Site merging of SVs discovered across batches, run on a cohort-level <code>sample_set_set</code></li>
<li><code>10-GenotypeBatch</code>: Per-batch genotyping of all sites in the cohort</li>
<li><code>11-RegenotypeCNVs</code>: Cohort-level genotype refinement of some depth calls</li>
<li><code>12-CombineBatches</code>: Cohort-level cross-batch integration and clustering</li>
<li><code>13-ResolveComplexVariants</code>: Complex variant resolution</li>
<li><code>14-GenotypeComplexVariants</code>: Complex variant re-genotyping</li>
<li><code>15-CleanVcf</code>: VCF cleanup</li>
<li><code>16-RefineComplexVariants</code>: Complex variant filtering and refinement</li>
<li><code>17-JoinRawCalls</code>: Raw call aggregation</li>
<li><code>18-SVConcordance</code>: Annotate genotype concordance with raw calls</li>
<li><code>19-FilterGenotypes</code>: Genotype filtering</li>
<li><code>20-AnnotateVcf</code>: Cohort VCF annotations, including functional annotation, allele frequency (AF) annotation, and
AF annotation with external population callsets</li>
</ol>
<p>Extra workflows (Not part of canonical pipeline, but included for your convenience. May require manual configuration):</p>
<ul>
<li><code>MainVcfQc</code>: Generate detailed call set QC plots</li>
<li><code>PlotSVCountsPerSample</code>: Plot SV counts per sample per SV type. Recommended to run before <code>FilterOutlierSamples</code>
(configured with the single VCF you want to filter) to enable IQR cutoff choice.</li>
<li><code>FilterOutlierSamples</code>: Filter outlier samples (in terms of SV counts) from a single VCF.</li>
<li><code>VisualizeCnvs</code>: Plot multi-sample depth profiles for CNVs</li>
</ul>
<p>For detailed instructions on running the pipeline in Terra, see <a href="#instructions">workflow instructions</a> below.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="what-is-the-maximum-number-of-samples-the-pipeline-can-handle">What is the maximum number of samples the pipeline can handle?<a href="#what-is-the-maximum-number-of-samples-the-pipeline-can-handle" class="hash-link" aria-label="Direct link to What is the maximum number of samples the pipeline can handle?" title="Direct link to What is the maximum number of samples the pipeline can handle?">​</a></h4>
<p>In Terra, we have tested batch sizes of up to 500 samples and cohort sizes (consisting of multiple batches) of up to
11,000 samples (and 98,000 samples with the final steps split by chromosome). On a dedicated Cromwell server, we have
tested the pipeline on cohorts of up to ~140,000 samples.</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="time-and-cost-estimates">Time and cost estimates<a href="#time-and-cost-estimates" class="hash-link" aria-label="Direct link to Time and cost estimates" title="Direct link to Time and cost estimates">​</a></h3>
<p>The following estimates pertain to the 1000 Genomes sample data in this workspace. They represent aggregated run time
and cost across modules for the whole pipeline. For workflows run multiple times (on each sample or on each batch),
the longest individual runtime was used. Call caching may affect some of this information.</p>
<table><thead><tr><th>Number of samples</th><th>Time</th><th>Total run cost</th><th>Per-sample run cost</th></tr></thead><tbody><tr><td>312</td><td>~76 hours</td><td>~$675</td><td>~$2.16/sample</td></tr></tbody></table>
<p>Please note that sample characteristics, cohort size, and level of filtering may influence pipeline compute costs,
with average costs ranging between $2-$3 per sample. For instance, PCR+ samples and samples with a high percentage
of improperly paired reads have been observed to cost more. Consider
<a href="/gatk-sv/docs/gs/inputs#sample-exclusion">excluding low-quality samples</a> prior to processing to keep costs low.</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="workspace-setup">Workspace setup<a href="#workspace-setup" class="hash-link" aria-label="Direct link to Workspace setup" title="Direct link to Workspace setup">​</a></h3>
<ol>
<li>
<p>Clone this workspace into a Terra project to which you have access. Select <code>us-central1</code> for the workspace region.
If you must use a different region, you will need to copy all GATK-SV docker images to the other region
before running the pipeline. See the <a href="/gatk-sv/docs/gs/dockers#regions-important">docker images section</a> for details.</p>
</li>
<li>
<p>In your new workspace, delete the example data. To do this, go to the <em>Data</em> tab of the workspace. Delete the data
tables in this order: <code>sample_set_set</code>, <code>sample_set</code>, and <code>sample</code>. For each table, click the 3 dots icon to the
right of the table name and click &quot;Delete table&quot;. Confirm when prompted.</p>
<img alt="deleting data tables" title="How to delete the sample data table" src="https://i.imgur.com/43M51WH.png" width="300" height="420">
</li>
<li>
<p>Create and upload a new sample data table for your samples. This should be a tab-separated file (.tsv) with one line
per sample, as well as a header (first) line. It should contain the columns <code>entity:sample_id</code> (first column) and
<code>bam_or_cram_file</code> at minimum. See the <strong>Required inputs</strong> section above for more information on these inputs. For
an example sample data table, refer to the sample data table for the 1000 Genomes samples in this workspace
<a href="https://github.com/broadinstitute/gatk-sv/blob/main/inputs/templates/terra_workspaces/cohort_mode/samples_1kgp_156.tsv.tmpl" target="_blank" rel="noopener noreferrer">here in the GATK-SV GitHub repository</a>.
To upload the TSV file, navigate to the <em>Data</em> tab of the workspace, click the <code>Import Data</code> button on the top left,
and select &quot;Upload TSV&quot;.</p>
<img alt="uploading a TSV data table" title="How to upload a TSV data table" src="https://i.imgur.com/1ZtwseH.png" width="300" height="250">
</li>
<li>
<p>Edit the <code>cohort_ped_file</code> item in the Workspace Data table (as shown in the screenshot below) to provide the Google
URI to the PED file for your cohort (make sure to share it with your Terra proxy account!).</p>
<img alt="editing cohort_ped_file" title="How to edit the cohort_ped_file attribute" src="https://i.imgur.com/IFwc0gs.png" width="800" height="400">
</li>
</ol>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="creating-sample_sets">Creating sample_sets<a href="#creating-sample_sets" class="hash-link" aria-label="Direct link to Creating sample_sets" title="Direct link to Creating sample_sets">​</a></h3>
<p>To create batches (in the <code>sample_set</code> table), we recommend using the <code>Batching.ipynb</code> notebook (see <a href="#batching">batching</a>).
To create batches manually, the easiest way is to upload a tab-separated sample set membership file.
This file should have one line per sample, plus a header (first) line. The first column should be
<code>membership:sample_set_id</code> (containing the <code>sample_set_id</code> for the sample in question), and the second should be
<code>sample</code> (containing the sample IDs). Recall that batch IDs (<code>sample_set_id</code>) should follow the
<a href="/gatk-sv/docs/gs/inputs#sampleids">sample ID requirements</a>. For an example sample set membership file, refer to
<a href="https://github.com/broadinstitute/gatk-sv/blob/main/inputs/templates/terra_workspaces/cohort_mode/sample_set_membership_1kgp.tsv.tmpl" target="_blank" rel="noopener noreferrer">this one in the GATK-SV GitHub repository</a>.</p>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="instructions">Workflow instructions<a href="#instructions" class="hash-link" aria-label="Direct link to Workflow instructions" title="Direct link to Workflow instructions">​</a></h2>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="general-recommendations">General recommendations<a href="#general-recommendations" class="hash-link" aria-label="Direct link to General recommendations" title="Direct link to General recommendations">​</a></h3>
<ul>
<li>It is recommended to run each workflow first on one sample/batch to check that the method is properly configured
before you attempt to process all of your data.</li>
<li>We recommend enabling call-caching (on by default in each workflow configuration).</li>
<li>We recommend enabling automatic intermediate file deletion by checking the box labeled &quot;Delete intermediate outputs&quot;
at the top of the workflow launch page every time you start a workflow. With this option enabled, intermediate files
(those not present in the Terra data table, and not needed for any further GATK-SV processing) will be deleted
automatically if the workflow succeeds. If the workflow fails, the outputs will be retained to enable a re-run to
pick up where it left off with call-caching. However, call-caching will not be possible for workflows that have
succeeded. For more information on this option, see
<a href="https://terra.bio/delete-intermediates-option-now-available-for-workflows-in-terra/" target="_blank" rel="noopener noreferrer">this article</a>. For guidance on
managing intermediate storage from failed workflows, or from workflows without the delete intermediates option enabled,
see the next bullet point.</li>
<li>There are cases when you may need to manage storage in other ways: for workflows that failed (only delete files from
a failed workflow after a version has succeeded, to avoid disabling call-caching), for workflows without intermediate
file deletion enabled, or once you are done processing and want to delete files from earlier steps in the pipeline
that you no longer need.<!-- -->
<ul>
<li>One option is to manually delete large files, or directories containing failed workflow intermediates (after
re-running the workflow successfully to take advantage of call-caching) with the command
<code>gsutil -m rm gs://path/to/workflow/directory/**file_extension_to_delete</code> to delete all files with the given extension
for that workflow, or <code>gsutil -m rm -r gs://path/to/workflow/directory/</code> to delete an entire workflow directory
(only after you are done with all the files!). Note that this can take a very long time for larger workflows, which
may contain thousands of files.</li>
<li>Another option is to use the <code>fiss mop</code> API call to delete all files that do not appear in one of the Terra data
tables (intermediate files). Always ensure that you are completely done with a step and you will not need to return
before using this option, as it will break call-caching. See
<a href="https://terra.bio/deleting-intermediate-workflow-outputs/" target="_blank" rel="noopener noreferrer">this blog post</a> for more details. This can also be done
<a href="https://github.com/broadinstitute/fiss/wiki/MOP:-reducing-your-cloud-storage-footprint" target="_blank" rel="noopener noreferrer">via the command line</a>.</li>
</ul>
</li>
<li>If your workflow fails, check the job manager for the error message. Most issues can be resolved by increasing the
memory or disk. Do not delete workflow log files until you are done troubleshooting. If call-caching is enabled, do not
delete any files from the failed workflow until you have run it successfully.</li>
<li>To display run costs, see <a href="https://support.terra.bio/hc/en-us/articles/360037862771#h_01EX5ED53HAZ59M29DRCG24CXY" target="_blank" rel="noopener noreferrer">this article</a>
for one-time setup instructions for non-Broad users.</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="01-gathersampleevidence">01-GatherSampleEvidence<a href="#01-gathersampleevidence" class="hash-link" aria-label="Direct link to 01-GatherSampleEvidence" title="Direct link to 01-GatherSampleEvidence">​</a></h3>
<p>Read the full GatherSampleEvidence documentation <a href="/gatk-sv/docs/modules/gse">here</a>.</p>
<ul>
<li>This workflow runs on a per-sample level, but you can launch many (a few hundred) samples at once, in arbitrary
partitions. Make sure to try just one sample first though!</li>
<li>Refer to the <a href="/gatk-sv/docs/gs/inputs">Input Data section</a> for details on file formats, sample QC, and sample ID restrictions.</li>
<li>It is normal for a few samples in a cohort to run out of memory during Wham SV calling, so we recommend enabling
auto-retry for out-of-memory errors for <code>01-GatherSampleEvidence</code> only. Before you launch the workflow, click the
checkbox reading &quot;Retry with more memory&quot; and set the memory retry factor to 2. This action must be performed each
time you launch a <code>01-GatherSampleEvidence</code> job.</li>
<li>If you enable &quot;Delete intermediate outputs&quot; whenever you launch this workflow (recommended), BAM files will be
deleted for successful runs; but BAM files will not be deleted if the run fails or if intermediate file deletion is
not enabled. Since BAM files are large, we recommend deleting them to save on storage costs, but only after fixing and
re-running the failed workflow, so that it will call-cache.</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="02-evidenceqc">02-EvidenceQC<a href="#02-evidenceqc" class="hash-link" aria-label="Direct link to 02-EvidenceQC" title="Direct link to 02-EvidenceQC">​</a></h3>
<p>Read the full EvidenceQC documentation <a href="/gatk-sv/docs/modules/eqc">here</a>.</p>
<ul>
<li><code>02-EvidenceQC</code> is run on arbitrary cohort partitions of up to 500 samples.</li>
<li>The outputs from <code>02-EvidenceQC</code> can be used for <a href="#sample-qc">sample QC</a> and
<a href="#batching">batching</a> before moving on to <a href="#traingcnv">TrainGCNV</a>.</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="sample-qc">Sample QC<a href="#sample-qc" class="hash-link" aria-label="Direct link to Sample QC" title="Direct link to Sample QC">​</a></h3>
<p>Read the documentation on preliminary sample QC <a href="/gatk-sv/docs/modules/eqc#preliminary-sample-qc">here</a>.
Follow the <code>SampleQC.ipynb</code> notebook step-by-step to evaluate sample data quality and remove low-quality samples as needed.
The notebook will produce a table of passing samples to use for <a href="#batching">batching</a>.</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="batching">Batching<a href="#batching" class="hash-link" aria-label="Direct link to Batching" title="Direct link to Batching">​</a></h3>
<p>Read the documentation on batching <a href="/gatk-sv/docs/modules/eqc#batching">here</a>.
If necessary, follow the <code>Batching.ipynb</code> notebook step-by-step to divide samples into batches
and create corresponding <code>sample_sets</code> for use in <code>03-TrainGCNV</code> and beyond.</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="traingcnv">03-TrainGCNV<a href="#traingcnv" class="hash-link" aria-label="Direct link to 03-TrainGCNV" title="Direct link to 03-TrainGCNV">​</a></h3>
<p>Read the full TrainGCNV documentation <a href="/gatk-sv/docs/modules/gcnv">here</a>.</p>
<ul>
<li>Before running this workflow, create the batches (~100-500 samples) you will use for the rest of the pipeline according
to the <a href="#batching">batching</a> instructions. These will likely not be the same as the batches you used for <code>02-EvidenceQC</code>.</li>
<li>By default, <code>03-TrainGCNV</code> is configured to be run once per <code>sample_set</code> on 100 randomly-chosen samples from that
set to create a gCNV model for each batch. To modify this behavior, you can set the <code>n_samples_subsample</code> parameter
to the number of samples to use for training.</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="04-gatherbatchevidence">04-GatherBatchEvidence<a href="#04-gatherbatchevidence" class="hash-link" aria-label="Direct link to 04-GatherBatchEvidence" title="Direct link to 04-GatherBatchEvidence">​</a></h3>
<p>Read the full GatherBatchEvidence documentation <a href="/gatk-sv/docs/modules/gbe">here</a>.</p>
<ul>
<li>Use the same <code>sample_set</code> definitions you used for <code>03-TrainGCNV</code>.</li>
<li>Before running this workflow, ensure that you have updated the <code>cohort_ped_file</code> attribute in Workspace Data with
your cohort&#x27;s PED file, with sex assignments updated based on ploidy detection from <code>02-EvidenceQC</code>.</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="steps-05-06">Steps 05-06<a href="#steps-05-06" class="hash-link" aria-label="Direct link to Steps 05-06" title="Direct link to Steps 05-06">​</a></h3>
<p>Read the full documentation for <a href="/gatk-sv/docs/modules/cb">ClusterBatch</a> and <a href="/gatk-sv/docs/modules/gbm">GenerateBatchMetrics</a>.</p>
<ul>
<li>Use the same <code>sample_set</code> definitions you used for <code>03-TrainGCNV</code> and <code>04-GatherBatchEvidence</code>.</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="steps-07-08">Steps 07-08<a href="#steps-07-08" class="hash-link" aria-label="Direct link to Steps 07-08" title="Direct link to Steps 07-08">​</a></h3>
<p>These two workflows make up FilterBatch; they are subdivided in this workspace to enable tuning of outlier filtration
cutoffs. Read the full FilterBatch documentation <a href="/gatk-sv/docs/modules/fb">here</a>.</p>
<ul>
<li>Use the same <code>sample_set</code> definitions you used for <code>03-TrainGCNV</code> through <code>06-GenerateBatchMetrics</code>.</li>
<li><code>07-FilterBatchSites</code> produces SV count plots and files, as well as a preview of the outlier samples to be filtered.
The input <code>N_IQR_cutoff_plotting</code> is used to visualize filtration thresholds on the SV count plots and preview the
samples to be filtered; the default value is set to 6. You can adjust this value depending on your needs, and you can
re-run the workflow with new <code>N_IQR_cutoff_plotting</code> values until the plots and outlier sample lists suit the purposes
of your study. Once you have chosen an IQR cutoff, provide it to the <code>N_IQR_cutoff</code> input in <code>08-FilterBatchSamples</code> to
filter the VCFs using the chosen cutoff.</li>
<li><code>08-FilterBatchSamples</code> performs outlier sample filtration, removing samples with an abnormal number of SV calls of
at least one SV type. To tune the filtering threshold to your needs, edit the <code>N_IQR_cutoff</code> input value based on the
plots and outlier sample preview lists from <code>07-FilterBatchSites</code>. The default value for <code>N_IQR_cutoff</code> in this step
is 10000, which essentially means that no samples are filtered.</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="09-mergebatchsites">09-MergeBatchSites<a href="#09-mergebatchsites" class="hash-link" aria-label="Direct link to 09-MergeBatchSites" title="Direct link to 09-MergeBatchSites">​</a></h3>
<p>Read the full MergeBatchSites documentation <a href="/gatk-sv/docs/modules/msites">here</a>.</p>
<ul>
<li><code>09-MergeBatchSites</code> is a cohort-level workflow, so it is run on a <code>sample_set_set</code> containing all the batches
in the cohort. Navigate to the Data tab of your workspace. If there is no <code>sample_set_set</code> data table, you will need
to create it. To do this, select the <code>sample_set</code> data table, then select (with the checkboxes) all the batches
(<code>sample_set</code>) in your cohort. These should be the <code>sample_sets</code> that you used to run steps <code>03-TrainGCNV</code> through
<code>08-FilterBatchSamples</code>. Then click the &quot;Edit&quot; icon above the table and choose &quot;Save selection as set.&quot; Enter a name
that follows the <strong>Sample ID requirements</strong>. This will create a new <code>sample_set_set</code> containing all of the <code>sample_sets</code>
in your cohort. When you launch MergeBatchSites, you can now select this <code>sample_set_set</code>.</li>
</ul>
<img alt="selecting batches" title="Selecting sample_sets in the data table" src="https://i.imgur.com/E5x3qqk.png" width="400" height="200">
<img alt="creating a new set" title="Creating a new sample_set_set" src="https://i.imgur.com/pizOtX9.png" width="400" height="200">
<ul>
<li>If there is already a <code>sample_set_set</code> data table in your workspace, you can create this <code>sample_set_set</code> while you
are launching the <code>09-MergeBatchSites</code> workflow: click &quot;Select Data&quot;, choose &quot;Create new sample_set_set [...]&quot;, check
all the batches to include (all the ones used in <code>03-TrainGCNV</code> through <code>08-FilterBatchSamples</code>), and give it a name
that follows the <a href="/gatk-sv/docs/gs/inputs#sampleids">sample ID requirements</a>.</li>
</ul>
<img alt="creating a cohort sample_set_set" title="How to create a cohort sample_set_set" src="https://i.imgur.com/zKEtSbe.png" width="377" height="363">
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="10-genotypebatch">10-GenotypeBatch<a href="#10-genotypebatch" class="hash-link" aria-label="Direct link to 10-GenotypeBatch" title="Direct link to 10-GenotypeBatch">​</a></h3>
<p>Read the full GenotypeBatch documentation <a href="/gatk-sv/docs/modules/gb">here</a>.</p>
<ul>
<li>Use the same <code>sample_set</code> definitions you used for <code>03-TrainGCNV</code> through <code>08-FilterBatchSamples</code>.</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="steps-11-20">Steps 11-20<a href="#steps-11-20" class="hash-link" aria-label="Direct link to Steps 11-20" title="Direct link to Steps 11-20">​</a></h3>
<p>Read the full documentation for <a href="/gatk-sv/docs/modules/rgcnvs">RegenotypeCNVs</a>, <a href="/gatk-sv/docs/modules/cmb">CombineBatches</a>,
<a href="/gatk-sv/docs/modules/rcv">ResolveComplexVariants</a>, <a href="/gatk-sv/docs/modules/gcv">GenotypeComplexVariants</a>, <a href="/gatk-sv/docs/modules/cvcf">CleanVcf</a>,
<a href="/gatk-sv/docs/modules/refcv">RefineComplexVariants</a>, <a href="/gatk-sv/docs/modules/jrc">JoinRawCalls</a>, <a href="/gatk-sv/docs/modules/svc">SVConcordance</a>,
<a href="/gatk-sv/docs/modules/fg">FilterGenotypes</a>, and <a href="/gatk-sv/docs/modules/av">AnnotateVcf</a>.</p>
<ul>
<li>Use the same cohort <code>sample_set_set</code> you created and used for <code>09-MergeBatchSites</code>.</li>
</ul></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="row margin-top--sm theme-doc-footer-edit-meta-row"><div class="col"><a href="https://github.com/broadinstitute/gatk-sv/tree/master/website/docs/execution/joint.md" target="_blank" rel="noopener noreferrer" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_JAkA"></div></div></footer></article><nav class="docusaurus-mt-lg pagination-nav" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/gatk-sv/docs/execution/single"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Single-sample</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/gatk-sv/docs/category/modules"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Modules</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#terra-workspace" class="table-of-contents__link toc-highlight">Terra workspace</a><ul><li><a href="#default-data" class="table-of-contents__link toc-highlight">Default data</a></li></ul></li><li><a href="#pipeline-expectations" class="table-of-contents__link toc-highlight">Pipeline Expectations</a><ul><li><a href="#what-does-it-do" class="table-of-contents__link toc-highlight">What does it do?</a></li><li><a href="#required-inputs" class="table-of-contents__link toc-highlight">Required inputs</a></li><li><a href="#pipeline-outputs" class="table-of-contents__link toc-highlight">Pipeline outputs</a></li><li><a href="#pipeline-overview" class="table-of-contents__link toc-highlight">Pipeline overview</a></li><li><a href="#time-and-cost-estimates" class="table-of-contents__link toc-highlight">Time and cost estimates</a></li><li><a href="#workspace-setup" class="table-of-contents__link toc-highlight">Workspace setup</a></li><li><a href="#creating-sample_sets" class="table-of-contents__link toc-highlight">Creating sample_sets</a></li></ul></li><li><a href="#instructions" class="table-of-contents__link toc-highlight">Workflow instructions</a><ul><li><a href="#general-recommendations" class="table-of-contents__link toc-highlight">General recommendations</a></li><li><a href="#01-gathersampleevidence" class="table-of-contents__link toc-highlight">01-GatherSampleEvidence</a></li><li><a href="#02-evidenceqc" class="table-of-contents__link toc-highlight">02-EvidenceQC</a></li><li><a href="#sample-qc" class="table-of-contents__link toc-highlight">Sample QC</a></li><li><a href="#batching" class="table-of-contents__link toc-highlight">Batching</a></li><li><a href="#traingcnv" class="table-of-contents__link toc-highlight">03-TrainGCNV</a></li><li><a href="#04-gatherbatchevidence" class="table-of-contents__link toc-highlight">04-GatherBatchEvidence</a></li><li><a href="#steps-05-06" class="table-of-contents__link toc-highlight">Steps 05-06</a></li><li><a href="#steps-07-08" class="table-of-contents__link toc-highlight">Steps 07-08</a></li><li><a href="#09-mergebatchsites" class="table-of-contents__link toc-highlight">09-MergeBatchSites</a></li><li><a href="#10-genotypebatch" class="table-of-contents__link toc-highlight">10-GenotypeBatch</a></li><li><a href="#steps-11-20" class="table-of-contents__link toc-highlight">Steps 11-20</a></li></ul></li></ul></div></div></div></div></main></div></div></div><footer class="theme-layout-footer footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="theme-layout-footer-column col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/gatk-sv/docs/intro">About</a></li></ul></div><div class="theme-layout-footer-column col footer__col"><div class="footer__title">Community</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/broadinstitute/gatk-sv/discussions" target="_blank" rel="noopener noreferrer" class="footer__link-item">Github<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a></li></ul></div><div class="theme-layout-footer-column col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://talkowski.mgh.harvard.edu" target="_blank" rel="noopener noreferrer" class="footer__link-item">Talkowski lab</a></li><li class="footer__item"><a href="https://www.broadinstitute.org" target="_blank" rel="noopener noreferrer" class="footer__link-item">Broad Institute</a></li><li class="footer__item"><a href="https://www.massgeneral.org/research/cgm" target="_blank" rel="noopener noreferrer" class="footer__link-item">Center for Genomic Medicine</a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2025 BroadInstitute, Built with Docusaurus.</div></div></div></footer></div>
</body>
</html>