---
title: Resource files
description: Resource files manifest
sidebar_position: 7
---

This page contains descriptions of common resource files used in the pipeline. All required files are publicly available 
in Google Cloud Storage Buckets. URIs for these files are available in the [hg38 resources json](https://github.com/broadinstitute/gatk-sv/blob/main/inputs/values/resources_hg38.json).

:::info
Some resources contain sensitive data and are stored in a secure bucket for development purposes. These files are not 
required to run GATK-SV.
:::

### Reference resources

#### allosome_file
Reference [fasta index file](https://www.htslib.org/doc/faidx.html) containing only allosomal contigs.

#### autosome_file
Reference [fasta index file](https://www.htslib.org/doc/faidx.html) containing only autosomal contigs.

#### bin_exclude
[Block-compressed](https://www.htslib.org/doc/bgzip.html) [bed file](https://genome.ucsc.edu/FAQ/FAQformat.html#format1) 
of intervals to exclude in the call set.

#### cnmops_exclude_list
Plain text bed file of N-masked regions in the reference.

#### contig_ploidy_priors
Plain text TSV file of prior probabilities contig ploidies used by GATK-gCNV.

#### cytobands
Block-compressed bed file of cytoband intervals.

#### sd_locs_vcf
Plain text VCF of SNP sites at which to collect site depth (SD) evidence.

#### depth_exclude_list
Block-compressed bed file of intervals over which to exclude overlapping depth-only calls.

#### empty_file
Empty file; used to satisfy some workflow code paths.

#### exclude_intervals_for_gcnv_filter_intervals
Plain text bed file of intervals to exclude from GATK-gCNV.

#### external_af_ref_bed
Block-compressed bed file of SV sites for external allele frequency annotation.

#### genome_file
Tab-delimited table of primary contigs with contig name in the first column and length in the second column.

#### manta_region_bed
Block-compressed bed file of intervals to call with Manta.

#### mei_bed
Block-compressed bed file of mobile element insertions in the reference genome.

#### melt_std_vcf_header
Text file containing the VCF header for raw MELT calls.

#### noncoding_bed
Plain text bed file of non-coding elements in the reference genome.

#### par_bed
Plain text bed file of pseudoautosomal regions.

#### pesr_exclude_list
Block-compressed bed file of intervals for filtering calls. Variants generated with non-CNV tools (Dragen, Manta, MELT, 
Scramble, Wham) that have either end in any of these intervals are hard-filtered.

#### preprocessed_intervals
Intervals for read count collection and CNV calling.

#### primary_contigs_fai
Reference [fasta index file](https://www.htslib.org/doc/faidx.html) only containing the primary contigs, i.e. `chr1`, ..., 
`chr22`, `chrX`, and `chrY`.

#### primary_contigs_list
Text file of primary contig names.

#### contigs_header
Plain text VCF header section of primary contig sequences.

#### protein_coding_gtf
Protein-coding sequence definitions for functional annotation in [General Transfer Format](https://www.ensembl.org/info/website/upload/gff.html). This GTF was created by subsetting the [GENCODE](https://www.gencodegenes.org/human/releases.html) GRCh38 basic gene annotation GTF with the script `scripts/inputs/preprocess_gtf.py`. Transcripts annotated as either Ensembl canonical or MANE Select Plus Clinical, and as either protein-coding or from nonsense-mediated decay, were retained. The GENCODE version is included in the filename.

#### reference_dict
Reference FASTA dictionary file (`*.dict`). See [this article](https://gatk.broadinstitute.org/hc/en-us/articles/360035531652-FASTA-Reference-genome-format) for more information.

#### reference_fasta
Reference FASTA file (`*.fasta`). See [this article](https://gatk.broadinstitute.org/hc/en-us/articles/360035531652-FASTA-Reference-genome-format) for more information.

#### reference_index
Reference FASTA index file (`*.fasta.fai`). See [this article](https://gatk.broadinstitute.org/hc/en-us/articles/360035531652-FASTA-Reference-genome-format) for more information.

#### rmsk
Block-compressed bed file of [RepeatMasker intervals](https://genome.ucsc.edu/cgi-bin/hgTrackUi?g=rmsk).

#### segdups
Block-compressed bed file of [segmental duplication intervals](https://genome.ucsc.edu/cgi-bin/hgTrackUi?g=genomicSuperDups).

#### seed_cutoffs
TSV of cutoff priors for genotyping.

#### single_sample_qc_definitions
TSV of recommended ranges for single-sample QC metrics.

#### wgd_scoring_mask
Plain text bed file of whole-genome dosage (WGD) score intervals over which to assess coverage bias.

#### wham_include_list_bed_file
Plain text bed file of intervals to call with Wham.

#### aou_recalibrate_gq_model_file
Genotype filtering model trained using data generated by All of Us.

#### hgdp_recalibrate_gq_model_file
Genotype filtering model trained using data generated from the Human Genome Diversity Project.

#### recalibrate_gq_genome_tracks
List of block-compressed bed files, each containing intervals from a separate genome track.

### Benchmarking datasets

#### ccdg_abel_site_level_benchmarking_dataset
Benchmarking variant set from [Abel et al. 2020](https://doi.org/10.1038/s41586-020-2371-0).

#### gnomad_v2_collins_sample_level_benchmarking_dataset
Benchmarking genotypes from gnomAD-SV-v2, see [Collins et al. 2020](https://doi.org/10.1038/s41586-020-2287-8). **Not public data.**

#### gnomad_v2_collins_site_level_benchmarking_dataset
Benchmarking variant set from gnomAD-SV-v2, see [Collins et al. 2020](https://doi.org/10.1038/s41586-020-2287-8).

#### hgsv_byrska_bishop_sample_level_benchmarking_dataset
Benchmarking genotypes from [Byrska-Bishop et al. 2022](https://doi.org/10.1016/j.cell.2022.08.004).

#### hgsv_byrska_bishop_sample_renaming_tsv
Sample renaming manifest for the Byrska-Bishop benchmarking genotypes.

#### hgsv_byrska_bishop_site_level_benchmarking_dataset
Benchmarking variant set from [Byrska-Bishop et al. 2022](https://doi.org/10.1016/j.cell.2022.08.004).

#### hgsv_ebert_sample_level_benchmarking_dataset
Benchmarking genotypes from [Ebert et al. 2021](https://doi.org/10.1126/science.abf7117).

#### ssc_belyeu_sample_level_benchmarking_dataset
Benchmarking genotypes from the [Simons Simplex Collection](https://www.sfari.org/resource/simons-simplex-collection/), derived from [Belyeu et al. 2021](https://doi.org/10.1016/j.ajhg.2021.02.012). **Not public data.**

#### ssc_belyeu_site_level_benchmarking_dataset
Benchmarking variant set from the [Simons Simplex Collection](https://www.sfari.org/resource/simons-simplex-collection/), derived from [Belyeu et al. 2021](https://doi.org/10.1016/j.ajhg.2021.02.012). **Not public data.**

#### ssc_sanders_sample_level_benchmarking_dataset
Benchmarking genotypes from the [Simons Simplex Collection](https://www.sfari.org/resource/simons-simplex-collection/), derived from [Sanders et al. 2015](https://doi.org/10.1016/j.neuron.2015.09.016). **Not public data.**

#### thousand_genomes_site_level_benchmarking_dataset
Benchmarking variant set from the 1000 Genomes Project Phase 3 SV call set, see [Sudmant et al. 2015](https://doi.org/10.1038/nature15394).

#### asc_site_level_benchmarking_dataset
Benchmarking variant set from the Autism Spectrum Consortium. **Not public data.**

#### hgsv_site_level_benchmarking_dataset
Benchmarking variant set from [Werling et al. 2018](https://doi.org/10.1038%2Fs41588-018-0107-y). **Not public data.**

#### collins_2017_sample_level_benchmarking_dataset
Benchmarking genotypes from [Collins et al. 2017](https://doi.org/10.1186/s13059-017-1158-6). **Not public data.**

#### sanders_2015_sample_level_benchmarking_dataset
Benchmarking genotypes from [Sanders et al. 2015](https://doi.org/10.1016/j.neuron.2015.09.016). **Not public data.**

#### werling_2018_sample_level_benchmarking_dataset
Benchmarking genotypes from [Werling et al. 2018](https://doi.org/10.1038%2Fs41588-018-0107-y). **Not public data.**

