<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper plugin-docs plugin-id-default docs-version-current docs-doc-page docs-doc-id-modules/main_vcf_qc" data-has-hydrated="false">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v3.8.0">
<title data-rh="true">MainVcfQC | GATK-SV</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/gatk-sv/docs/modules/mvqc"><meta data-rh="true" property="og:locale" content="en"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="MainVcfQC | GATK-SV"><meta data-rh="true" name="description" content="VCF QC plotting"><meta data-rh="true" property="og:description" content="VCF QC plotting"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/gatk-sv/docs/modules/mvqc"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/gatk-sv/docs/modules/mvqc" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/gatk-sv/docs/modules/mvqc" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://FPHFCFLYVM-dsn.algolia.net" crossorigin="anonymous"><script data-rh="true" type="application/ld+json">{"@context":"https://schema.org","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"name":"Modules","item":"https://broadinstitute.github.io/gatk-sv/docs/category/modules"},{"@type":"ListItem","position":2,"name":"MainVcfQC","item":"https://broadinstitute.github.io/gatk-sv/docs/modules/mvqc"}]}</script><link rel="search" type="application/opensearchdescription+xml" title="GATK-SV" href="/gatk-sv/opensearch.xml"><link rel="stylesheet" href="/gatk-sv/assets/css/styles.04c1d369.css">
<script src="/gatk-sv/assets/js/runtime~main.863a676d.js" defer="defer"></script>
<script src="/gatk-sv/assets/js/main.33bd68b8.js" defer="defer"></script>
</head>
<body class="navigation-with-keyboard">
<svg xmlns="http://www.w3.org/2000/svg" style="display: none;"><defs>
<symbol id="theme-svg-external-link" viewBox="0 0 24 24"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"/></symbol>
</defs></svg>
<script>!function(){var t="light";var e=function(){try{return new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}}()||function(){try{return window.localStorage.getItem("theme")}catch(t){}}();document.documentElement.setAttribute("data-theme",e||t),document.documentElement.setAttribute("data-theme-choice",e||t)}(),function(){try{const c=new URLSearchParams(window.location.search).entries();for(var[t,e]of c)if(t.startsWith("docusaurus-data-")){var a=t.replace("docusaurus-data-","data-");document.documentElement.setAttribute(a,e)}}catch(t){}}()</script><div id="__docusaurus"><div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="theme-layout-navbar navbar navbar--fixed-top"><div class="navbar__inner"><div class="theme-layout-navbar-left navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/gatk-sv/"><b class="navbar__title text--truncate">GATK-SV</b></a></div><div class="theme-layout-navbar-right navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/gatk-sv/docs/intro">Documentation</a><a href="https://github.com/broadinstitute/gatk-sv/issues" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">Questions<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a><a href="https://github.com/broadinstitute/gatk-sv" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="system mode" aria-label="Switch between dark and light mode (currently system mode)"><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP systemToggleIcon_QzmC"><path fill="currentColor" d="m12 21c4.971 0 9-4.029 9-9s-4.029-9-9-9-9 4.029-9 9 4.029 9 9 9zm4.95-13.95c1.313 1.313 2.05 3.093 2.05 4.95s-0.738 3.637-2.05 4.95c-1.313 1.313-3.093 2.05-4.95 2.05v-14c1.857 0 3.637 0.737 4.95 2.05z"></path></svg></button></div><div class="navbarSearchContainer_Bca1"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search (Command+K)"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20" aria-hidden="true"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="theme-layout-main main-wrapper mainWrapper_z2l0"><div class="docsWrapper_hBAB"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docRoot_UBD9"><aside class="theme-doc-sidebar-container docSidebarContainer_YfHR"><div class="sidebarViewport_aRkj"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/intro">About</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/getting-started">Getting Started</a><button aria-label="Expand sidebar category &#x27;Getting Started&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/best_practices">Best Practices Guide</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/execution">Execution</a><button aria-label="Expand sidebar category &#x27;Execution&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" href="/gatk-sv/docs/category/modules">Modules</a><button aria-label="Collapse sidebar category &#x27;Modules&#x27;" aria-expanded="true" type="button" class="clean-btn menu__caret"></button></div><ul class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/">Overview</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/gse">GatherSampleEvidence</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/eqc">EvidenceQC</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/gcnv">TrainGCNV</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/gbe">GatherBatchEvidence</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/cb">ClusterBatch</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/gbm">GenerateBatchMetrics</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/fb">FilterBatch</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/msites">MergeBatchSites</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/gb">GenotypeBatch</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/rgcnvs">RegenotypeCNVs</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/cmb">CombineBatches</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/rcv">ResolveComplexVariants</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/gcv">GenotypeComplexVariants</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/cvcf">CleanVcf</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/refcv">RefineComplexVariants</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/jrc">JoinRawCalls</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/svc">SVConcordance</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/fg">FilterGenotypes</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/av">AnnotateVcf</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/gatk-sv/docs/modules/mvqc">MainVcfQC</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/gatk-sv/docs/modules/viz">VisualizeCnvs</a></li></ul></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/runtime_attr">Runtime attributes</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/resources">Resource files</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/troubleshooting">Troubleshooting</a><button aria-label="Expand sidebar category &#x27;Troubleshooting&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/advanced-guides">Advanced Guides</a><button aria-label="Expand sidebar category &#x27;Advanced Guides&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/acknowledgements">Acknowledgements</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/references">Citation and references</a></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_TBSr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/gatk-sv/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><a class="breadcrumbs__link" href="/gatk-sv/docs/category/modules"><span>Modules</span></a></li><li class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link">MainVcfQC</span></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><header><h1>MainVcfQC</h1></header><p><a href="https://github.com/broadinstitute/gatk-sv/blob/main/wdl/MainVcfQC.wdl" target="_blank" rel="noopener noreferrer">WDL source code</a></p>
<p>Creates plots for call set analysis and quality control (QC). This module is run by default with the
<a href="/gatk-sv/docs/modules/fg">FilterGenotypes</a> module. Note, however, that the module is a stand-alone workflow and can be used with any batch-
or cohort-level (i.e. multi-sample) SV VCF, including outputs from <a href="/gatk-sv/docs/modules/cb">ClusterBatch</a> and onward.</p>
<p>Users can optionally provide external benchmarking datasets with the
<a href="#optional-site_level_comparison_datasets">site_level_comparison_datasets</a> and
<a href="#optional-sample_level_comparison_datasets">sample_level_comparison_datasets</a> parameters.</p>
<p>The following sections provide guidance on interpreting these plots for VCFs that have been run with the recommended
pipeline through <a href="/gatk-sv/docs/modules/fg">FilterGenotypes</a>.</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="output-structure">Output structure<a href="#output-structure" class="hash-link" aria-label="Direct link to Output structure" title="Direct link to Output structure">​</a></h3>
<p>The tarball output has the following directory structure:</p>
<ul>
<li><code>/plots/main_plots</code> : main QC plots, described in detail <a href="#plots-description">below</a>.</li>
<li><code>/plots/supplementary_plots</code> : comprehensive set of QC plots with greater detail, some of which are used as panels
in <code>main_plots</code>.</li>
<li><code>/data</code> : various datafiles from the QC process such as family structures, a list of the samples
used for plotting, variant statistics, and lists of overlapping variants against any external call sets provided.</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="plots-description">Plots description<a href="#plots-description" class="hash-link" aria-label="Direct link to Plots description" title="Direct link to Plots description">​</a></h3>
<p>This section summarizes the most important plots used for overall call set QC found in the <code>/plots/main_plots</code>
directory.</p>
<p>Example plots are provided from a high-quality dataset consisting of 161 samples from the 1000 Genomes Project aligned
with DRAGEN v3.7.8. Note that QC metrics and distributions will vary among data sets, particularly as sample size
increases, and these examples are intended to provide a simple baseline for acceptable call set quality. Please see
the <a href="#recommendations">Recommendations</a> section for prescribed quality criteria.</p>
<div class="theme-admonition theme-admonition-note admonition_xJq3 alert alert--secondary"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 14 16"><path fill-rule="evenodd" d="M6.3 5.69a.942.942 0 0 1-.28-.7c0-.28.09-.52.28-.7.19-.18.42-.28.7-.28.28 0 .52.09.7.28.18.19.28.42.28.7 0 .28-.09.52-.28.7a1 1 0 0 1-.7.3c-.28 0-.52-.11-.7-.3zM8 7.99c-.02-.25-.11-.48-.31-.69-.2-.19-.42-.3-.69-.31H6c-.27.02-.48.13-.69.31-.2.2-.3.44-.31.69h1v3c.02.27.11.5.31.69.2.2.42.31.69.31h1c.27 0 .48-.11.69-.31.2-.19.3-.42.31-.69H8V7.98v.01zM7 2.3c-3.14 0-5.7 2.54-5.7 5.68 0 3.14 2.56 5.7 5.7 5.7s5.7-2.55 5.7-5.7c0-3.15-2.56-5.69-5.7-5.69v.01zM7 .98c3.86 0 7 3.14 7 7s-3.14 7-7 7-7-3.12-7-7 3.14-7 7-7z"></path></svg></span>note</div><div class="admonitionContent_BuS1"><p>The following plots are of variants passing all filters (i.e. with the <code>FILTER</code> status set to <code>PASS</code> or <code>MULTIALLELIC</code>). This is the
default behavior of the QC plots generated in <a href="/gatk-sv/docs/modules/fg">FilterGenotypes</a>.</p><p>When running <code>MainVcfQc</code> as a standalone workflow, users may set the
<a href="#optional--bcftools_preprocessing_options">bcftools_preprocessing_options</a> argument to limit plotted variants based on <code>FILTER</code> status.
For example, to limit to <code>PASS</code> and <code>MULTIALLELIC</code> variants for a VCF generated from <a href="/gatk-sv/docs/modules/fg">FilterGenotypes</a> use:</p><div class="language-text codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_QJqH"><pre tabindex="0" class="prism-code language-text codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">&quot;bcftools_preprocessing_options&quot;: &quot;-i &#x27;FILTER=\&quot;PASS\&quot; || FILTER=\&quot;MULTIALLELIC\&quot;&#x27;&quot;</span><br></span></code></pre></div></div></div></div>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="sv-per-genome">SV per genome<a href="#sv-per-genome" class="hash-link" aria-label="Direct link to SV per genome" title="Direct link to SV per genome">​</a></h4>
<p>Shows per-genome SV count distributions broken down by SV type. The top row is a site-level analysis, and
the bottom row provides counts of individual alleles. Distributions of genotype quality (<code>GQ</code>), SV size, and
frequency are also shown.</p>
<p><img decoding="async" loading="lazy" alt="SV counts" src="/gatk-sv/assets/images/VcfQcSvPerGenome-28e9963e6f919ffe84e949061681863c.png" width="3300" height="1500" class="img_ev3q"></p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="sv-counts">SV counts<a href="#sv-counts" class="hash-link" aria-label="Direct link to SV counts" title="Direct link to SV counts">​</a></h4>
<p>Summarizes total variant counts by type, frequency, and size.</p>
<p><img decoding="async" loading="lazy" alt="SV counts" src="/gatk-sv/assets/images/VcfQcSvCountsMerged-10c85b5f3fda832a686c3faaf740d069.png" width="648" height="360" class="img_ev3q"></p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="genotype-distributions">Genotype distributions<a href="#genotype-distributions" class="hash-link" aria-label="Direct link to Genotype distributions" title="Direct link to Genotype distributions">​</a></h4>
<p>Distributions of genotypes across the call set.</p>
<p>The left panel plots carrier frequency against allele frequency and can be used to visualize genotyping bias. In this
plot, the rolling mean should not lie on the lower or upper extremes (all het. or all hom.).</p>
<p>The middle panel plots the Hardy-Weinberg (HW) distribution of the cohort and provides estimates for the proportions of
variants in HW equilibrium (or nominally in HW equilibrium before a multiple testing correction). The right-hand
plots stratify HW by SV size.</p>
<div class="theme-admonition theme-admonition-note admonition_xJq3 alert alert--secondary"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 14 16"><path fill-rule="evenodd" d="M6.3 5.69a.942.942 0 0 1-.28-.7c0-.28.09-.52.28-.7.19-.18.42-.28.7-.28.28 0 .52.09.7.28.18.19.28.42.28.7 0 .28-.09.52-.28.7a1 1 0 0 1-.7.3c-.28 0-.52-.11-.7-.3zM8 7.99c-.02-.25-.11-.48-.31-.69-.2-.19-.42-.3-.69-.31H6c-.27.02-.48.13-.69.31-.2.2-.3.44-.31.69h1v3c.02.27.11.5.31.69.2.2.42.31.69.31h1c.27 0 .48-.11.69-.31.2-.19.3-.42.31-.69H8V7.98v.01zM7 2.3c-3.14 0-5.7 2.54-5.7 5.68 0 3.14 2.56 5.7 5.7 5.7s5.7-2.55 5.7-5.7c0-3.15-2.56-5.69-5.7-5.69v.01zM7 .98c3.86 0 7 3.14 7 7s-3.14 7-7 7-7-3.12-7-7 3.14-7 7-7z"></path></svg></span>note</div><div class="admonitionContent_BuS1"><p>The horizontal line bisecting the vertical axis in the AF vs. carrier frequency plot was fixed in.
<a href="https://github.com/broadinstitute/gatk-sv/pull/625" target="_blank" rel="noopener noreferrer">#625</a>).</p></div></div>
<p><img decoding="async" loading="lazy" alt="SV counts" src="/gatk-sv/assets/images/VcfQcGenotypeDistributions-4e7e7f0f44798d578d0cb7dc5621c458.png" width="4200" height="1200" class="img_ev3q"></p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="size-distributions">Size distributions<a href="#size-distributions" class="hash-link" aria-label="Direct link to Size distributions" title="Direct link to Size distributions">​</a></h4>
<p>Contains plots of SV length stratified by class and frequency.</p>
<p><img decoding="async" loading="lazy" alt="SV counts" src="/gatk-sv/assets/images/VcfQcSizeDistributionsMerged-48e89ca021aeeb152523ec291255e612.png" width="720" height="432" class="img_ev3q"></p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="frequency-distributions">Frequency distributions<a href="#frequency-distributions" class="hash-link" aria-label="Direct link to Frequency distributions" title="Direct link to Frequency distributions">​</a></h4>
<p>Contains plots of allele frequency stratified by class and size.</p>
<p><img decoding="async" loading="lazy" alt="SV counts" src="/gatk-sv/assets/images/VcfQqFreqDistributionsMerged-1c0f5750fe3aa50caabab8c7467a8579.png" width="720" height="432" class="img_ev3q"></p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-sv-trio-inheritance"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> SV trio inheritance<a href="#optional-sv-trio-inheritance" class="hash-link" aria-label="Direct link to optional-sv-trio-inheritance" title="Direct link to optional-sv-trio-inheritance">​</a></h4>
<p>Inheritance analysis of sites and alleles, including <em>de novo</em> rate distributions by SV class, size, frequency, and
genotype quality. This plot is only generated if the cohort contains trios defined in the <a href="#optional-ped_file">PED file</a>.</p>
<p><img decoding="async" loading="lazy" alt="SV counts" src="/gatk-sv/assets/images/VcfQcSvTrioInheritance-b3d297861b28000cb75a2583acb0cc6e.png" width="864" height="360" class="img_ev3q"></p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-call-set-benchmarking"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> Call set benchmarking<a href="#optional-call-set-benchmarking" class="hash-link" aria-label="Direct link to optional-call-set-benchmarking" title="Direct link to optional-call-set-benchmarking">​</a></h4>
<p>This plot will be generated for each external call set provided. The left-hand panels show distributions of overlapping
variants in the input VCF by SV class, size, and frequency. Colors depict proportions of variants meeting each overlap
criterion. The right-hand panels show distributions among variants overlapping between the input and external call sets
by SV class, size, and frequency.</p>
<div class="theme-admonition theme-admonition-note admonition_xJq3 alert alert--secondary"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 14 16"><path fill-rule="evenodd" d="M6.3 5.69a.942.942 0 0 1-.28-.7c0-.28.09-.52.28-.7.19-.18.42-.28.7-.28.28 0 .52.09.7.28.18.19.28.42.28.7 0 .28-.09.52-.28.7a1 1 0 0 1-.7.3c-.28 0-.52-.11-.7-.3zM8 7.99c-.02-.25-.11-.48-.31-.69-.2-.19-.42-.3-.69-.31H6c-.27.02-.48.13-.69.31-.2.2-.3.44-.31.69h1v3c.02.27.11.5.31.69.2.2.42.31.69.31h1c.27 0 .48-.11.69-.31.2-.19.3-.42.31-.69H8V7.98v.01zM7 2.3c-3.14 0-5.7 2.54-5.7 5.68 0 3.14 2.56 5.7 5.7 5.7s5.7-2.55 5.7-5.7c0-3.15-2.56-5.69-5.7-5.69v.01zM7 .98c3.86 0 7 3.14 7 7s-3.14 7-7 7-7-3.12-7-7 3.14-7 7-7z"></path></svg></span>note</div><div class="admonitionContent_BuS1"><p>Note the horizontal line bisecting the vertical axis in the AF vs. AF plot has been fixed in
<a href="https://github.com/broadinstitute/gatk-sv/pull/625" target="_blank" rel="noopener noreferrer">#625</a>).</p></div></div>
<p><img decoding="async" loading="lazy" alt="SV counts" src="/gatk-sv/assets/images/VcfQcGnomADv2CollinsSVAllCallsetBenchmarking-f8e004899e2a7366fa9ec288af1bc36a.png" width="4200" height="1800" class="img_ev3q"></p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="recommendations">Recommendations<a href="#recommendations" class="hash-link" aria-label="Direct link to Recommendations" title="Direct link to Recommendations">​</a></h3>
<p>We suggest users observe the following basic criteria to assess the overall quality of the final call set:</p>
<ul>
<li>Number of PASS variants per genome (which excludes <code>BND</code>) between 7,000 and 11,000.</li>
<li>Relative overall proportions of SV types similar to those shown above.</li>
<li>Similar size and frequency distributions to those shown above.</li>
<li>At least 70% of variants in HW equilibrium. Note that this may be lower depending on how well
the underlying cohort population structure adheres to the assumptions of the HW model. Note that HW equilibrium
usually improves after genotype filtering.</li>
<li>Strong allele frequency correlation against an external benchmarking dataset such as gnomAD-v2.</li>
<li>Low allele <em>de novo</em> inheritance rate (if trios are present), typically below 10%.</li>
</ul>
<p>These are intended as general guidelines and not strict limits on call set quality.</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="inputs">Inputs<a href="#inputs" class="hash-link" aria-label="Direct link to Inputs" title="Direct link to Inputs">​</a></h3>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="vcfs"><code>vcfs</code><a href="#vcfs" class="hash-link" aria-label="Direct link to vcfs" title="Direct link to vcfs">​</a></h4>
<p>Array of one or more input VCFs. Each VCF must be indexed. This input is structured as an <code>Array</code> type for convenience
when VCFs are sharded by contig. All VCFs in the array are concatenated before processing.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional--vcf_format_has_cn"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span>  <code>vcf_format_has_cn</code><a href="#optional--vcf_format_has_cn" class="hash-link" aria-label="Direct link to optional--vcf_format_has_cn" title="Direct link to optional--vcf_format_has_cn">​</a></h4>
<p>Default: <code>true</code>. This must be set to <code>false</code> for VCFs from modules before <a href="/gatk-sv/docs/modules/cvcf">CleanVcf</a> (i.e. without <code>CN</code> fields).</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional--bcftools_preprocessing_options"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span>  <code>bcftools_preprocessing_options</code><a href="#optional--bcftools_preprocessing_options" class="hash-link" aria-label="Direct link to optional--bcftools_preprocessing_options" title="Direct link to optional--bcftools_preprocessing_options">​</a></h4>
<p>Additional arguments to pass to <a href="https://samtools.github.io/bcftools/bcftools.html" target="_blank" rel="noopener noreferrer">bcftools view</a> before running
the VCF through QC tools.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-ped_file"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>ped_file</code><a href="#optional-ped_file" class="hash-link" aria-label="Direct link to optional-ped_file" title="Direct link to optional-ped_file">​</a></h4>
<p>Family structures and sex assignments determined in <a href="/gatk-sv/docs/modules/eqc">EvidenceQC</a>. See <a href="/gatk-sv/docs/gs/inputs#ped-format">PED file format</a>.
If provided and there are trios present then <a href="#optional-sv-trio-inheritance">SV trio plots</a> will be generated.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-list_of_samples_to_include"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>list_of_samples_to_include</code><a href="#optional-list_of_samples_to_include" class="hash-link" aria-label="Direct link to optional-list_of_samples_to_include" title="Direct link to optional-list_of_samples_to_include">​</a></h4>
<p>If provided, the input VCF(s) will be subset to samples in the list.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-sample_renaming_tsv"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>sample_renaming_tsv</code><a href="#optional-sample_renaming_tsv" class="hash-link" aria-label="Direct link to optional-sample_renaming_tsv" title="Direct link to optional-sample_renaming_tsv">​</a></h4>
<p>File with mapping to rename per-sample benchmark sample IDs for compatibility with cohort (see
<a href="/gatk-sv/docs/resources#hgsv_byrska_bishop_sample_renaming_tsv">here</a> for example).</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-max_trios"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>max_trios</code><a href="#optional-max_trios" class="hash-link" aria-label="Direct link to optional-max_trios" title="Direct link to optional-max_trios">​</a></h4>
<p>Default: <code>1000</code>. Upper limit on the number of trios to use for inheritance analysis.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="prefix"><code>prefix</code><a href="#prefix" class="hash-link" aria-label="Direct link to prefix" title="Direct link to prefix">​</a></h4>
<p>Output prefix, such as cohort name. May be alphanumeric with underscores.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="sv_per_shard"><code>sv_per_shard</code><a href="#sv_per_shard" class="hash-link" aria-label="Direct link to sv_per_shard" title="Direct link to sv_per_shard">​</a></h4>
<p>Records per shard for parallel processing. This parameter may be reduced if the workflow is running slowly.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="samples_per_shard"><code>samples_per_shard</code><a href="#samples_per_shard" class="hash-link" aria-label="Direct link to samples_per_shard" title="Direct link to samples_per_shard">​</a></h4>
<p>Samples per shard for per-sample QC processing. This parameter may be reduced if the workflow is running slowly. Only
has an effect when <a href="#optional-do_per_sample_qc">do_per_sample_qc</a> is enabled.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-do_per_sample_qc"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>do_per_sample_qc</code><a href="#optional-do_per_sample_qc" class="hash-link" aria-label="Direct link to optional-do_per_sample_qc" title="Direct link to optional-do_per_sample_qc">​</a></h4>
<p>Default: <code>true</code>. If enabled, performs per-sample SV count plots, family analysis, and site-level benchmarking.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-site_level_comparison_datasets"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>site_level_comparison_datasets</code><a href="#optional-site_level_comparison_datasets" class="hash-link" aria-label="Direct link to optional-site_level_comparison_datasets" title="Direct link to optional-site_level_comparison_datasets">​</a></h4>
<p>Array of two-element arrays, one per dataset, each a tuple containing <code>[prefix, bed_uri_path]</code>, where the latter is
a directory containing multiple bed files containing the call set stratified by subpopulation. For example:</p>
<div class="language-text codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_QJqH"><pre tabindex="0" class="prism-code language-text codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">[&quot;gnomAD_v2_Collins&quot;, &quot;gs://gatk-sv-resources-public/hg38/v0/sv-resources/resources/v1/gnomAD_v2_Collins&quot;]</span><br></span></code></pre></div></div>
<p>where the bucket directory contains the following files:</p>
<div class="language-text codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_QJqH"><pre tabindex="0" class="prism-code language-text codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">gs://gatk-sv-resources-public/hg38/v0/sv-resources/resources/v1/gnomAD_v2_Collins/gnomAD_v2_Collins.SV.AFR.bed.gz</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gs://gatk-sv-resources-public/hg38/v0/sv-resources/resources/v1/gnomAD_v2_Collins/gnomAD_v2_Collins.SV.ALL.bed.gz</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gs://gatk-sv-resources-public/hg38/v0/sv-resources/resources/v1/gnomAD_v2_Collins/gnomAD_v2_Collins.SV.AMR.bed.gz</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gs://gatk-sv-resources-public/hg38/v0/sv-resources/resources/v1/gnomAD_v2_Collins/gnomAD_v2_Collins.SV.EAS.bed.gz</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gs://gatk-sv-resources-public/hg38/v0/sv-resources/resources/v1/gnomAD_v2_Collins/gnomAD_v2_Collins.SV.EUR.bed.gz</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gs://gatk-sv-resources-public/hg38/v0/sv-resources/resources/v1/gnomAD_v2_Collins/gnomAD_v2_Collins.SV.OTH.bed.gz</span><br></span></code></pre></div></div>
<p>Users can examine the above files as an example of how to format custom benchmarking datasets.</p>
<p>See the <code>*_site_level_benchmarking_dataset</code> entries in the <a href="/gatk-sv/docs/resources">Resource files</a> section for available
benchmarking call sets.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-sample_level_comparison_datasets"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>sample_level_comparison_datasets</code><a href="#optional-sample_level_comparison_datasets" class="hash-link" aria-label="Direct link to optional-sample_level_comparison_datasets" title="Direct link to optional-sample_level_comparison_datasets">​</a></h4>
<p>Array of two-element arrays, one per dataset, each a tuple containing <code>[prefix, tarball_uri]</code>, where the latter is
a tarball containing sample-level benchmarking data. For example:</p>
<div class="language-text codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_QJqH"><pre tabindex="0" class="prism-code language-text codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">[[&quot;HGSV_ByrskaBishop&quot;, &quot;gs://gatk-sv-resources-public/hg38/v0/sv-resources/resources/v1/HGSV_ByrskaBishop_GATKSV_perSample.tar.gz&quot;]]</span><br></span></code></pre></div></div>
<p>Users can examine the above file as an example of how to format custom benchmarking datasets.</p>
<p>See the <code>*_sample_level_benchmarking_dataset</code> entries in the <a href="/gatk-sv/docs/resources#benchmarking-datasets">Resource files</a>
section for available benchmarking call sets.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-random_seed"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>random_seed</code><a href="#optional-random_seed" class="hash-link" aria-label="Direct link to optional-random_seed" title="Direct link to optional-random_seed">​</a></h4>
<p>Default: <code>0</code>. Random seed for sample subsetting in external call set comparisons.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-max_gq"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>max_gq</code><a href="#optional-max_gq" class="hash-link" aria-label="Direct link to optional-max_gq" title="Direct link to optional-max_gq">​</a></h4>
<p>Default: <code>99</code>. Max value to define range for <code>GQ</code> plotting. For modules prior to <code>CombineBatches</code>, use <code>999</code>.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="optional-downsample_qc_per_sample"><span style="background-color:var(--highlight-optional-arg-background-color);border-radius:2px;color:var(--highlight-optional-arg-text-color);padding:0.2rem">Optional</span> <code>downsample_qc_per_sample</code><a href="#optional-downsample_qc_per_sample" class="hash-link" aria-label="Direct link to optional-downsample_qc_per_sample" title="Direct link to optional-downsample_qc_per_sample">​</a></h4>
<p>Default: <code>1000</code>. Number of samples to subset to for per-sample QC.</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="outputs">Outputs<a href="#outputs" class="hash-link" aria-label="Direct link to Outputs" title="Direct link to Outputs">​</a></h3>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="sv_vcf_qc_output"><code>sv_vcf_qc_output</code><a href="#sv_vcf_qc_output" class="hash-link" aria-label="Direct link to sv_vcf_qc_output" title="Direct link to sv_vcf_qc_output">​</a></h4>
<p>Tarball of QC plots and data tables.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="vcf2bed_output"><code>vcf2bed_output</code><a href="#vcf2bed_output" class="hash-link" aria-label="Direct link to vcf2bed_output" title="Direct link to vcf2bed_output">​</a></h4>
<p>Bed file containing all input variants.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="row margin-top--sm theme-doc-footer-edit-meta-row"><div class="col"><a href="https://github.com/broadinstitute/gatk-sv/tree/master/website/docs/modules/main_vcf_qc.md" target="_blank" rel="noopener noreferrer" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_JAkA"></div></div></footer></article><nav class="docusaurus-mt-lg pagination-nav" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/gatk-sv/docs/modules/av"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">AnnotateVcf</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/gatk-sv/docs/modules/viz"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">VisualizeCnvs</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#output-structure" class="table-of-contents__link toc-highlight">Output structure</a></li><li><a href="#plots-description" class="table-of-contents__link toc-highlight">Plots description</a></li><li><a href="#recommendations" class="table-of-contents__link toc-highlight">Recommendations</a></li><li><a href="#inputs" class="table-of-contents__link toc-highlight">Inputs</a></li><li><a href="#outputs" class="table-of-contents__link toc-highlight">Outputs</a></li></ul></div></div></div></div></main></div></div></div><footer class="theme-layout-footer footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="theme-layout-footer-column col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/gatk-sv/docs/intro">About</a></li></ul></div><div class="theme-layout-footer-column col footer__col"><div class="footer__title">Community</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/broadinstitute/gatk-sv/discussions" target="_blank" rel="noopener noreferrer" class="footer__link-item">Github<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a></li></ul></div><div class="theme-layout-footer-column col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://talkowski.mgh.harvard.edu" target="_blank" rel="noopener noreferrer" class="footer__link-item">Talkowski lab</a></li><li class="footer__item"><a href="https://www.broadinstitute.org" target="_blank" rel="noopener noreferrer" class="footer__link-item">Broad Institute</a></li><li class="footer__item"><a href="https://www.massgeneral.org/research/cgm" target="_blank" rel="noopener noreferrer" class="footer__link-item">Center for Genomic Medicine</a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2025 BroadInstitute, Built with Docusaurus.</div></div></div></footer></div>
</body>
</html>