<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper plugin-docs plugin-id-default docs-version-current docs-doc-page docs-doc-id-resources" data-has-hydrated="false">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v3.8.1">
<title data-rh="true">Resource files | GATK-SV</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://broadinstitute.github.io/gatk-sv/docs/resources"><meta data-rh="true" property="og:locale" content="en"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Resource files | GATK-SV"><meta data-rh="true" name="description" content="Resource files manifest"><meta data-rh="true" property="og:description" content="Resource files manifest"><link data-rh="true" rel="canonical" href="https://broadinstitute.github.io/gatk-sv/docs/resources"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/gatk-sv/docs/resources" hreflang="en"><link data-rh="true" rel="alternate" href="https://broadinstitute.github.io/gatk-sv/docs/resources" hreflang="x-default"><link data-rh="true" rel="preconnect" href="https://FPHFCFLYVM-dsn.algolia.net" crossorigin="anonymous"><script data-rh="true" type="application/ld+json">{"@context":"https://schema.org","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"name":"Resource files","item":"https://broadinstitute.github.io/gatk-sv/docs/resources"}]}</script><link rel="preconnect" href="https://www.google-analytics.com">
<link rel="preconnect" href="https://www.googletagmanager.com">
<script async src="https://www.googletagmanager.com/gtag/js?id=G-9DQEYKHD1M"></script>
<script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-9DQEYKHD1M",{anonymize_ip:!0})</script>



<link rel="search" type="application/opensearchdescription+xml" title="GATK-SV" href="/gatk-sv/opensearch.xml"><link rel="stylesheet" href="/gatk-sv/assets/css/styles.82e231fb.css">
<script src="/gatk-sv/assets/js/runtime~main.6abd9fa4.js" defer="defer"></script>
<script src="/gatk-sv/assets/js/main.97c5ebd9.js" defer="defer"></script>
</head>
<body class="navigation-with-keyboard">
<svg xmlns="http://www.w3.org/2000/svg" style="display: none;"><defs>
<symbol id="theme-svg-external-link" viewBox="0 0 24 24"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"/></symbol>
</defs></svg>
<script>!function(){var t="light";var e=function(){try{return new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}}()||function(){try{return window.localStorage.getItem("theme")}catch(t){}}();document.documentElement.setAttribute("data-theme",e||t),document.documentElement.setAttribute("data-theme-choice",e||t)}(),function(){try{const c=new URLSearchParams(window.location.search).entries();for(var[t,e]of c)if(t.startsWith("docusaurus-data-")){var a=t.replace("docusaurus-data-","data-");document.documentElement.setAttribute(a,e)}}catch(t){}}()</script><div id="__docusaurus"><div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="theme-layout-navbar navbar navbar--fixed-top"><div class="navbar__inner"><div class="theme-layout-navbar-left navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/gatk-sv/"><b class="navbar__title text--truncate">GATK-SV</b></a></div><div class="theme-layout-navbar-right navbar__items navbar__items--right"><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/gatk-sv/docs/intro">Documentation</a><a href="https://github.com/broadinstitute/gatk-sv/issues" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">Questions<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a><a href="https://github.com/broadinstitute/gatk-sv" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="system mode" aria-label="Switch between dark and light mode (currently system mode)"><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP systemToggleIcon_QzmC"><path fill="currentColor" d="m12 21c4.971 0 9-4.029 9-9s-4.029-9-9-9-9 4.029-9 9 4.029 9 9 9zm4.95-13.95c1.313 1.313 2.05 3.093 2.05 4.95s-0.738 3.637-2.05 4.95c-1.313 1.313-3.093 2.05-4.95 2.05v-14c1.857 0 3.637 0.737 4.95 2.05z"></path></svg></button></div><div class="navbarSearchContainer_Bca1"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search (Command+K)"><span class="DocSearch-Button-Container"><svg width="20" height="20" class="DocSearch-Search-Icon" viewBox="0 0 20 20" aria-hidden="true"><path d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z" stroke="currentColor" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"></span></button></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="theme-layout-main main-wrapper mainWrapper_z2l0"><div class="docsWrapper_hBAB"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docRoot_UBD9"><aside class="theme-doc-sidebar-container docSidebarContainer_YfHR"><div class="sidebarViewport_aRkj"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/intro">About</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/getting-started">Getting Started</a><button aria-label="Expand sidebar category &#x27;Getting Started&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/best_practices">Best Practices Guide</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/execution">Execution</a><button aria-label="Expand sidebar category &#x27;Execution&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/modules">Modules</a><button aria-label="Expand sidebar category &#x27;Modules&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/runtime_attr">Runtime attributes</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" href="/gatk-sv/docs/resources">Resource files</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/troubleshooting">Troubleshooting</a><button aria-label="Expand sidebar category &#x27;Troubleshooting&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/gatk-sv/docs/category/advanced-guides">Advanced Guides</a><button aria-label="Expand sidebar category &#x27;Advanced Guides&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/acknowledgements">Acknowledgements</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/gatk-sv/docs/references">Citation and references</a></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_TBSr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/gatk-sv/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link">Resource files</span></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><header><h1>Resource files</h1></header><p>This page contains descriptions of common resource files used in the pipeline. All required files are publicly available
in Google Cloud Storage Buckets. URIs for these files are available in the <a href="https://github.com/broadinstitute/gatk-sv/blob/main/inputs/values/resources_hg38.json" target="_blank" rel="noopener noreferrer">hg38 resources json</a>.</p>
<div class="theme-admonition theme-admonition-info admonition_xJq3 alert alert--info"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 14 16"><path fill-rule="evenodd" d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg></span>info</div><div class="admonitionContent_BuS1"><p>Some resources contain sensitive data and are stored in a secure bucket for development purposes. These files are not
required to run GATK-SV.</p></div></div>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="reference-resources">Reference resources<a href="#reference-resources" class="hash-link" aria-label="Direct link to Reference resources" title="Direct link to Reference resources">​</a></h3>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="allosome_file">allosome_file<a href="#allosome_file" class="hash-link" aria-label="Direct link to allosome_file" title="Direct link to allosome_file">​</a></h4>
<p>Reference <a href="https://www.htslib.org/doc/faidx.html" target="_blank" rel="noopener noreferrer">fasta index file</a> containing only allosomal contigs.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="autosome_file">autosome_file<a href="#autosome_file" class="hash-link" aria-label="Direct link to autosome_file" title="Direct link to autosome_file">​</a></h4>
<p>Reference <a href="https://www.htslib.org/doc/faidx.html" target="_blank" rel="noopener noreferrer">fasta index file</a> containing only autosomal contigs.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="bin_exclude">bin_exclude<a href="#bin_exclude" class="hash-link" aria-label="Direct link to bin_exclude" title="Direct link to bin_exclude">​</a></h4>
<p><a href="https://www.htslib.org/doc/bgzip.html" target="_blank" rel="noopener noreferrer">Block-compressed</a> <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1" target="_blank" rel="noopener noreferrer">bed file</a>
of intervals to exclude in the call set.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="cnmops_exclude_list">cnmops_exclude_list<a href="#cnmops_exclude_list" class="hash-link" aria-label="Direct link to cnmops_exclude_list" title="Direct link to cnmops_exclude_list">​</a></h4>
<p>Plain text bed file of N-masked regions in the reference.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="contig_ploidy_priors">contig_ploidy_priors<a href="#contig_ploidy_priors" class="hash-link" aria-label="Direct link to contig_ploidy_priors" title="Direct link to contig_ploidy_priors">​</a></h4>
<p>Plain text TSV file of prior probabilities contig ploidies used by GATK-gCNV.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="cytobands">cytobands<a href="#cytobands" class="hash-link" aria-label="Direct link to cytobands" title="Direct link to cytobands">​</a></h4>
<p>Block-compressed bed file of cytoband intervals.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="sd_locs_vcf">sd_locs_vcf<a href="#sd_locs_vcf" class="hash-link" aria-label="Direct link to sd_locs_vcf" title="Direct link to sd_locs_vcf">​</a></h4>
<p>Plain text VCF of SNP sites at which to collect site depth (SD) evidence.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="depth_exclude_list">depth_exclude_list<a href="#depth_exclude_list" class="hash-link" aria-label="Direct link to depth_exclude_list" title="Direct link to depth_exclude_list">​</a></h4>
<p>Block-compressed bed file of intervals over which to exclude overlapping depth-only calls.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="empty_file">empty_file<a href="#empty_file" class="hash-link" aria-label="Direct link to empty_file" title="Direct link to empty_file">​</a></h4>
<p>Empty file; used to satisfy some workflow code paths.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="exclude_intervals_for_gcnv_filter_intervals">exclude_intervals_for_gcnv_filter_intervals<a href="#exclude_intervals_for_gcnv_filter_intervals" class="hash-link" aria-label="Direct link to exclude_intervals_for_gcnv_filter_intervals" title="Direct link to exclude_intervals_for_gcnv_filter_intervals">​</a></h4>
<p>Plain text bed file of intervals to exclude from GATK-gCNV.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="external_af_ref_bed">external_af_ref_bed<a href="#external_af_ref_bed" class="hash-link" aria-label="Direct link to external_af_ref_bed" title="Direct link to external_af_ref_bed">​</a></h4>
<p>Block-compressed bed file of SV sites for external allele frequency annotation.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="genome_file">genome_file<a href="#genome_file" class="hash-link" aria-label="Direct link to genome_file" title="Direct link to genome_file">​</a></h4>
<p>Tab-delimited table of primary contigs with contig name in the first column and length in the second column.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="manta_region_bed">manta_region_bed<a href="#manta_region_bed" class="hash-link" aria-label="Direct link to manta_region_bed" title="Direct link to manta_region_bed">​</a></h4>
<p>Block-compressed bed file of intervals to call with Manta.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="mei_bed">mei_bed<a href="#mei_bed" class="hash-link" aria-label="Direct link to mei_bed" title="Direct link to mei_bed">​</a></h4>
<p>Block-compressed bed file of mobile element insertions in the reference genome.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="melt_std_vcf_header">melt_std_vcf_header<a href="#melt_std_vcf_header" class="hash-link" aria-label="Direct link to melt_std_vcf_header" title="Direct link to melt_std_vcf_header">​</a></h4>
<p>Text file containing the VCF header for raw MELT calls.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="noncoding_bed">noncoding_bed<a href="#noncoding_bed" class="hash-link" aria-label="Direct link to noncoding_bed" title="Direct link to noncoding_bed">​</a></h4>
<p>Plain text bed file of non-coding elements in the reference genome.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="par_bed">par_bed<a href="#par_bed" class="hash-link" aria-label="Direct link to par_bed" title="Direct link to par_bed">​</a></h4>
<p>Plain text bed file of pseudoautosomal regions.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="pesr_exclude_list">pesr_exclude_list<a href="#pesr_exclude_list" class="hash-link" aria-label="Direct link to pesr_exclude_list" title="Direct link to pesr_exclude_list">​</a></h4>
<p>Block-compressed bed file of intervals for filtering calls. Variants generated with non-CNV tools (Dragen, Manta, MELT,
Scramble, Wham) that have either end in any of these intervals are hard-filtered.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="preprocessed_intervals">preprocessed_intervals<a href="#preprocessed_intervals" class="hash-link" aria-label="Direct link to preprocessed_intervals" title="Direct link to preprocessed_intervals">​</a></h4>
<p>Intervals for read count collection and CNV calling.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="primary_contigs_fai">primary_contigs_fai<a href="#primary_contigs_fai" class="hash-link" aria-label="Direct link to primary_contigs_fai" title="Direct link to primary_contigs_fai">​</a></h4>
<p>Reference <a href="https://www.htslib.org/doc/faidx.html" target="_blank" rel="noopener noreferrer">fasta index file</a> only containing the primary contigs, i.e. <code>chr1</code>, ...,
<code>chr22</code>, <code>chrX</code>, and <code>chrY</code>.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="primary_contigs_list">primary_contigs_list<a href="#primary_contigs_list" class="hash-link" aria-label="Direct link to primary_contigs_list" title="Direct link to primary_contigs_list">​</a></h4>
<p>Text file of primary contig names.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="contigs_header">contigs_header<a href="#contigs_header" class="hash-link" aria-label="Direct link to contigs_header" title="Direct link to contigs_header">​</a></h4>
<p>Plain text VCF header section of primary contig sequences.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="protein_coding_gtf">protein_coding_gtf<a href="#protein_coding_gtf" class="hash-link" aria-label="Direct link to protein_coding_gtf" title="Direct link to protein_coding_gtf">​</a></h4>
<p>Protein-coding sequence definitions for functional annotation in <a href="https://www.ensembl.org/info/website/upload/gff.html" target="_blank" rel="noopener noreferrer">General Transfer Format</a>. This GTF was created by subsetting the <a href="https://www.gencodegenes.org/human/releases.html" target="_blank" rel="noopener noreferrer">GENCODE</a> GRCh38 basic gene annotation GTF with the script <code>scripts/inputs/preprocess_gtf.py</code>. Transcripts annotated as either Ensembl canonical or MANE Select Plus Clinical, and as either protein-coding or from nonsense-mediated decay, were retained. The GENCODE version is included in the filename.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="reference_dict">reference_dict<a href="#reference_dict" class="hash-link" aria-label="Direct link to reference_dict" title="Direct link to reference_dict">​</a></h4>
<p>Reference FASTA dictionary file (<code>*.dict</code>). See <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531652-FASTA-Reference-genome-format" target="_blank" rel="noopener noreferrer">this article</a> for more information.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="reference_fasta">reference_fasta<a href="#reference_fasta" class="hash-link" aria-label="Direct link to reference_fasta" title="Direct link to reference_fasta">​</a></h4>
<p>Reference FASTA file (<code>*.fasta</code>). See <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531652-FASTA-Reference-genome-format" target="_blank" rel="noopener noreferrer">this article</a> for more information.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="reference_index">reference_index<a href="#reference_index" class="hash-link" aria-label="Direct link to reference_index" title="Direct link to reference_index">​</a></h4>
<p>Reference FASTA index file (<code>*.fasta.fai</code>). See <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531652-FASTA-Reference-genome-format" target="_blank" rel="noopener noreferrer">this article</a> for more information.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="rmsk">rmsk<a href="#rmsk" class="hash-link" aria-label="Direct link to rmsk" title="Direct link to rmsk">​</a></h4>
<p>Block-compressed bed file of <a href="https://genome.ucsc.edu/cgi-bin/hgTrackUi?g=rmsk" target="_blank" rel="noopener noreferrer">RepeatMasker intervals</a>.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="segdups">segdups<a href="#segdups" class="hash-link" aria-label="Direct link to segdups" title="Direct link to segdups">​</a></h4>
<p>Block-compressed bed file of <a href="https://genome.ucsc.edu/cgi-bin/hgTrackUi?g=genomicSuperDups" target="_blank" rel="noopener noreferrer">segmental duplication intervals</a>.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="seed_cutoffs">seed_cutoffs<a href="#seed_cutoffs" class="hash-link" aria-label="Direct link to seed_cutoffs" title="Direct link to seed_cutoffs">​</a></h4>
<p>TSV of cutoff priors for genotyping.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="single_sample_qc_definitions">single_sample_qc_definitions<a href="#single_sample_qc_definitions" class="hash-link" aria-label="Direct link to single_sample_qc_definitions" title="Direct link to single_sample_qc_definitions">​</a></h4>
<p>TSV of recommended ranges for single-sample QC metrics.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="wgd_scoring_mask">wgd_scoring_mask<a href="#wgd_scoring_mask" class="hash-link" aria-label="Direct link to wgd_scoring_mask" title="Direct link to wgd_scoring_mask">​</a></h4>
<p>Plain text bed file of whole-genome dosage (WGD) score intervals over which to assess coverage bias.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="wham_include_list_bed_file">wham_include_list_bed_file<a href="#wham_include_list_bed_file" class="hash-link" aria-label="Direct link to wham_include_list_bed_file" title="Direct link to wham_include_list_bed_file">​</a></h4>
<p>Plain text bed file of intervals to call with Wham.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="sl_cutoff_table">sl_cutoff_table<a href="#sl_cutoff_table" class="hash-link" aria-label="Direct link to sl_cutoff_table" title="Direct link to sl_cutoff_table">​</a></h4>
<p>Cutoffs used in the genotype filtering model trained using data generated by All of Us.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="aou_recalibrate_gq_model_file">aou_recalibrate_gq_model_file<a href="#aou_recalibrate_gq_model_file" class="hash-link" aria-label="Direct link to aou_recalibrate_gq_model_file" title="Direct link to aou_recalibrate_gq_model_file">​</a></h4>
<p>Genotype filtering model trained using data generated by All of Us.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="hgdp_recalibrate_gq_model_file">hgdp_recalibrate_gq_model_file<a href="#hgdp_recalibrate_gq_model_file" class="hash-link" aria-label="Direct link to hgdp_recalibrate_gq_model_file" title="Direct link to hgdp_recalibrate_gq_model_file">​</a></h4>
<p>Genotype filtering model trained using data generated from the Human Genome Diversity Project.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="recalibrate_gq_genome_tracks">recalibrate_gq_genome_tracks<a href="#recalibrate_gq_genome_tracks" class="hash-link" aria-label="Direct link to recalibrate_gq_genome_tracks" title="Direct link to recalibrate_gq_genome_tracks">​</a></h4>
<p>List of block-compressed bed files, each containing intervals from a separate genome track.</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="benchmarking-datasets">Benchmarking datasets<a href="#benchmarking-datasets" class="hash-link" aria-label="Direct link to Benchmarking datasets" title="Direct link to Benchmarking datasets">​</a></h3>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="ccdg_abel_site_level_benchmarking_dataset">ccdg_abel_site_level_benchmarking_dataset<a href="#ccdg_abel_site_level_benchmarking_dataset" class="hash-link" aria-label="Direct link to ccdg_abel_site_level_benchmarking_dataset" title="Direct link to ccdg_abel_site_level_benchmarking_dataset">​</a></h4>
<p>Benchmarking variant set from <a href="https://doi.org/10.1038/s41586-020-2371-0" target="_blank" rel="noopener noreferrer">Abel et al. 2020</a>.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="gnomad_v2_collins_sample_level_benchmarking_dataset">gnomad_v2_collins_sample_level_benchmarking_dataset<a href="#gnomad_v2_collins_sample_level_benchmarking_dataset" class="hash-link" aria-label="Direct link to gnomad_v2_collins_sample_level_benchmarking_dataset" title="Direct link to gnomad_v2_collins_sample_level_benchmarking_dataset">​</a></h4>
<p>Benchmarking genotypes from gnomAD-SV-v2, see <a href="https://doi.org/10.1038/s41586-020-2287-8" target="_blank" rel="noopener noreferrer">Collins et al. 2020</a>. <strong>Not public data.</strong></p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="gnomad_v2_collins_site_level_benchmarking_dataset">gnomad_v2_collins_site_level_benchmarking_dataset<a href="#gnomad_v2_collins_site_level_benchmarking_dataset" class="hash-link" aria-label="Direct link to gnomad_v2_collins_site_level_benchmarking_dataset" title="Direct link to gnomad_v2_collins_site_level_benchmarking_dataset">​</a></h4>
<p>Benchmarking variant set from gnomAD-SV-v2, see <a href="https://doi.org/10.1038/s41586-020-2287-8" target="_blank" rel="noopener noreferrer">Collins et al. 2020</a>.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="hgsv_byrska_bishop_sample_level_benchmarking_dataset">hgsv_byrska_bishop_sample_level_benchmarking_dataset<a href="#hgsv_byrska_bishop_sample_level_benchmarking_dataset" class="hash-link" aria-label="Direct link to hgsv_byrska_bishop_sample_level_benchmarking_dataset" title="Direct link to hgsv_byrska_bishop_sample_level_benchmarking_dataset">​</a></h4>
<p>Benchmarking genotypes from <a href="https://doi.org/10.1016/j.cell.2022.08.004" target="_blank" rel="noopener noreferrer">Byrska-Bishop et al. 2022</a>.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="hgsv_byrska_bishop_sample_renaming_tsv">hgsv_byrska_bishop_sample_renaming_tsv<a href="#hgsv_byrska_bishop_sample_renaming_tsv" class="hash-link" aria-label="Direct link to hgsv_byrska_bishop_sample_renaming_tsv" title="Direct link to hgsv_byrska_bishop_sample_renaming_tsv">​</a></h4>
<p>Sample renaming manifest for the Byrska-Bishop benchmarking genotypes.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="hgsv_byrska_bishop_site_level_benchmarking_dataset">hgsv_byrska_bishop_site_level_benchmarking_dataset<a href="#hgsv_byrska_bishop_site_level_benchmarking_dataset" class="hash-link" aria-label="Direct link to hgsv_byrska_bishop_site_level_benchmarking_dataset" title="Direct link to hgsv_byrska_bishop_site_level_benchmarking_dataset">​</a></h4>
<p>Benchmarking variant set from <a href="https://doi.org/10.1016/j.cell.2022.08.004" target="_blank" rel="noopener noreferrer">Byrska-Bishop et al. 2022</a>.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="hgsv_ebert_sample_level_benchmarking_dataset">hgsv_ebert_sample_level_benchmarking_dataset<a href="#hgsv_ebert_sample_level_benchmarking_dataset" class="hash-link" aria-label="Direct link to hgsv_ebert_sample_level_benchmarking_dataset" title="Direct link to hgsv_ebert_sample_level_benchmarking_dataset">​</a></h4>
<p>Benchmarking genotypes from <a href="https://doi.org/10.1126/science.abf7117" target="_blank" rel="noopener noreferrer">Ebert et al. 2021</a>.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="ssc_belyeu_sample_level_benchmarking_dataset">ssc_belyeu_sample_level_benchmarking_dataset<a href="#ssc_belyeu_sample_level_benchmarking_dataset" class="hash-link" aria-label="Direct link to ssc_belyeu_sample_level_benchmarking_dataset" title="Direct link to ssc_belyeu_sample_level_benchmarking_dataset">​</a></h4>
<p>Benchmarking genotypes from the <a href="https://www.sfari.org/resource/simons-simplex-collection/" target="_blank" rel="noopener noreferrer">Simons Simplex Collection</a>, derived from <a href="https://doi.org/10.1016/j.ajhg.2021.02.012" target="_blank" rel="noopener noreferrer">Belyeu et al. 2021</a>. <strong>Not public data.</strong></p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="ssc_belyeu_site_level_benchmarking_dataset">ssc_belyeu_site_level_benchmarking_dataset<a href="#ssc_belyeu_site_level_benchmarking_dataset" class="hash-link" aria-label="Direct link to ssc_belyeu_site_level_benchmarking_dataset" title="Direct link to ssc_belyeu_site_level_benchmarking_dataset">​</a></h4>
<p>Benchmarking variant set from the <a href="https://www.sfari.org/resource/simons-simplex-collection/" target="_blank" rel="noopener noreferrer">Simons Simplex Collection</a>, derived from <a href="https://doi.org/10.1016/j.ajhg.2021.02.012" target="_blank" rel="noopener noreferrer">Belyeu et al. 2021</a>. <strong>Not public data.</strong></p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="ssc_sanders_sample_level_benchmarking_dataset">ssc_sanders_sample_level_benchmarking_dataset<a href="#ssc_sanders_sample_level_benchmarking_dataset" class="hash-link" aria-label="Direct link to ssc_sanders_sample_level_benchmarking_dataset" title="Direct link to ssc_sanders_sample_level_benchmarking_dataset">​</a></h4>
<p>Benchmarking genotypes from the <a href="https://www.sfari.org/resource/simons-simplex-collection/" target="_blank" rel="noopener noreferrer">Simons Simplex Collection</a>, derived from <a href="https://doi.org/10.1016/j.neuron.2015.09.016" target="_blank" rel="noopener noreferrer">Sanders et al. 2015</a>. <strong>Not public data.</strong></p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="thousand_genomes_site_level_benchmarking_dataset">thousand_genomes_site_level_benchmarking_dataset<a href="#thousand_genomes_site_level_benchmarking_dataset" class="hash-link" aria-label="Direct link to thousand_genomes_site_level_benchmarking_dataset" title="Direct link to thousand_genomes_site_level_benchmarking_dataset">​</a></h4>
<p>Benchmarking variant set from the 1000 Genomes Project Phase 3 SV call set, see <a href="https://doi.org/10.1038/nature15394" target="_blank" rel="noopener noreferrer">Sudmant et al. 2015</a>.</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="asc_site_level_benchmarking_dataset">asc_site_level_benchmarking_dataset<a href="#asc_site_level_benchmarking_dataset" class="hash-link" aria-label="Direct link to asc_site_level_benchmarking_dataset" title="Direct link to asc_site_level_benchmarking_dataset">​</a></h4>
<p>Benchmarking variant set from the Autism Spectrum Consortium. <strong>Not public data.</strong></p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="hgsv_site_level_benchmarking_dataset">hgsv_site_level_benchmarking_dataset<a href="#hgsv_site_level_benchmarking_dataset" class="hash-link" aria-label="Direct link to hgsv_site_level_benchmarking_dataset" title="Direct link to hgsv_site_level_benchmarking_dataset">​</a></h4>
<p>Benchmarking variant set from <a href="https://doi.org/10.1038%2Fs41588-018-0107-y" target="_blank" rel="noopener noreferrer">Werling et al. 2018</a>. <strong>Not public data.</strong></p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="collins_2017_sample_level_benchmarking_dataset">collins_2017_sample_level_benchmarking_dataset<a href="#collins_2017_sample_level_benchmarking_dataset" class="hash-link" aria-label="Direct link to collins_2017_sample_level_benchmarking_dataset" title="Direct link to collins_2017_sample_level_benchmarking_dataset">​</a></h4>
<p>Benchmarking genotypes from <a href="https://doi.org/10.1186/s13059-017-1158-6" target="_blank" rel="noopener noreferrer">Collins et al. 2017</a>. <strong>Not public data.</strong></p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="sanders_2015_sample_level_benchmarking_dataset">sanders_2015_sample_level_benchmarking_dataset<a href="#sanders_2015_sample_level_benchmarking_dataset" class="hash-link" aria-label="Direct link to sanders_2015_sample_level_benchmarking_dataset" title="Direct link to sanders_2015_sample_level_benchmarking_dataset">​</a></h4>
<p>Benchmarking genotypes from <a href="https://doi.org/10.1016/j.neuron.2015.09.016" target="_blank" rel="noopener noreferrer">Sanders et al. 2015</a>. <strong>Not public data.</strong></p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="werling_2018_sample_level_benchmarking_dataset">werling_2018_sample_level_benchmarking_dataset<a href="#werling_2018_sample_level_benchmarking_dataset" class="hash-link" aria-label="Direct link to werling_2018_sample_level_benchmarking_dataset" title="Direct link to werling_2018_sample_level_benchmarking_dataset">​</a></h4>
<p>Benchmarking genotypes from <a href="https://doi.org/10.1038%2Fs41588-018-0107-y" target="_blank" rel="noopener noreferrer">Werling et al. 2018</a>. <strong>Not public data.</strong></p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="row margin-top--sm theme-doc-footer-edit-meta-row"><div class="col"><a href="https://github.com/broadinstitute/gatk-sv/tree/master/website/docs/resources.md" target="_blank" rel="noopener noreferrer" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_JAkA"></div></div></footer></article><nav class="docusaurus-mt-lg pagination-nav" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/gatk-sv/docs/runtime_attr"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Runtime attributes</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/gatk-sv/docs/category/troubleshooting"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Troubleshooting</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#reference-resources" class="table-of-contents__link toc-highlight">Reference resources</a></li><li><a href="#benchmarking-datasets" class="table-of-contents__link toc-highlight">Benchmarking datasets</a></li></ul></div></div></div></div></main></div></div></div><footer class="theme-layout-footer footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="theme-layout-footer-column col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/gatk-sv/docs/intro">About</a></li></ul></div><div class="theme-layout-footer-column col footer__col"><div class="footer__title">Community</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/broadinstitute/gatk-sv/discussions" target="_blank" rel="noopener noreferrer" class="footer__link-item">Github<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a></li></ul></div><div class="theme-layout-footer-column col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://talkowski.mgh.harvard.edu" target="_blank" rel="noopener noreferrer" class="footer__link-item">Talkowski lab</a></li><li class="footer__item"><a href="https://www.broadinstitute.org" target="_blank" rel="noopener noreferrer" class="footer__link-item">Broad Institute</a></li><li class="footer__item"><a href="https://www.massgeneral.org/research/cgm" target="_blank" rel="noopener noreferrer" class="footer__link-item">Center for Genomic Medicine</a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2026 BroadInstitute, Built with Docusaurus.</div></div></div></footer></div>
</body>
</html>